import pandas as pd
import re
import os

configfile: "config/config.yaml"

basedir = config['base_dir']

# Sample metadata
sample_metadata_df = pd.read_excel(config['sample_metadata'], header=0)

treatments = ['Vedolizumab', 'Ustekinumab']
comparisons = ['T1RvNR', 'T2RvNR', 'RvNR', 'RT2vT1', 'RvNRvT2vT1']

# Illumina HumanMethylation EPIC BeadChip

# Rules
include: "rules/epic.smk"
include: "rules/rnaseq.smk"

wildcard_constraints:
  treatment='|'.join([re.escape(x) for x in treatments]),
  comparison='|'.join([re.escape(x) for x in comparisons]),

rule all:
  input:
    ## Illumina HumanMethylation EPIC BeadChip
    #expand("{basedir}/output/epic/import/rgset.Rds", basedir=basedir),
    #expand("{basedir}/output/epic/qc/rgset_qc.Rds", basedir=basedir),
    #expand("{basedir}/output/epic/subset/gmset_{treatment}.Rds", basedir=basedir, treatment=treatments),
    #expand("{basedir}/output/epic/dmp/dmp_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/dmp/dmp_{treatment}_annotated.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/dmp/dmp_continuous_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/dmp/dmp_T1RvNR_wconfounders_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/dmp/dmp_T1CRP_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/dmp/dmp_T1FCP_{treatment}.csv", basedir=basedir, treatment=treatments),
    # expand("{basedir}/output/epic/dmp/ora/dmp_{comparison}_{treatment}_ora_go.csv", basedir=basedir, comparison=comparisons, treatment=treatments),
    # expand("{basedir}/output/epic/dmg/dmg_{comparison}_{treatment}.csv", basedir=basedir, treatment=treatments, comparison=comparisons),
    expand("{basedir}/output/epic/dmr/dmrcate_{comparison}_{treatment}.Rds", basedir=basedir, treatment=treatments, comparison=comparisons),
    expand("{basedir}/output/epic/icc/icc_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/training/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/validation/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/multibiological_validation/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/timepoint2_validation/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/ancillary_validation/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/validation_regressedconfounders/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/horaizon/crossdrug_validation/X_{treatment}.csv", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/epic/metadata/priorantitnf_glmmodel_{treatment}_discovery.Rds", basedir=basedir, treatment=treatments),
    # expand("{basedir}/output/epic/horaizon/ora/predictor_{treatment}_ora_go.csv", basedir=basedir, treatment=treatments),
    ## RNA-sequencing
    # expand("output/rnaseq/fastqc/{readfile}/{readfile}_fastqc.zip", readfile=rnaseq_readfiles),
    # expand("output/rnaseq/bam/{rnaseq_sampleID}/{rnaseq_sampleID}_filtered.bam.bai", rnaseq_sampleID=rnaseq_sampleIDs),
    # "output/rnaseq/bam/STAR.log",
    # "output/rnaseq/counts/counts.txt",
    # "output/rnaseq/multiqc",
    # expand("{basedir}/output/rnaseq/r/dds.Rds", basedir=basedir),
    # expand("{basedir}/output/rnaseq/r/dds_{treatment}.Rds", basedir=basedir, treatment=treatments),
    expand("{basedir}/output/rnaseq/degs/degs_{treatment}.csv", basedir=basedir, treatment=treatments),
    # expand("{basedir}/output/rnaseq/degs/fgsea_go_{treatment}_fgseaobj_list.Rds", basedir=basedir, treatment=treatments),
