---
title: "Integrative DNA methylation and Transcriptomic Analyses"
author: "Andrew Y.F. Li Yim"
date: '2022-09-21'
output: html_document
editor_options: 
  chunk_output_type: console
---

In this document, we seek to understand whether the observations at the level of DNA methylation and the observations at the level of transcriptomics overlap.

```{r setup, include=FALSE}
library(minfi)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(DESeq2)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
library(ChIPpeakAnno)
library(ggpubr)
```

```{r import epic}

```

```{r import horaizon}
epic_anno_gr <- GenomicRanges::makeGRangesFromDataFrame(minfi::getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19), keep.extra.columns = T, start.field = "pos", end.field = "pos", seqnames.field = "chr")

pchic_gr_bait <- GenomicRanges::makeGRangesFromDataFrame(read.csv("resources/pchic/ActivePromoterEnhancerLinks.tsv", sep = "\t"), keep.extra.columns = T, start.field = "baitSt", end.field = "baitEnd", seqnames.field = "baitChr")

pchic_gr_anno_bait <- annotatePeakInBatch(myPeakList = pchic_gr, AnnotationData=genes(TxDb.Hsapiens.UCSC.hg19.knownGene))

pchic_gr_anno_bait <- addGeneIDs(annotatedPeak=pchic_gr_anno,
                                 orgAnn="org.Hs.eg.db",
                                 feature_id_type="entrez_id",
                                 IDs2Add="symbol")

pchic_gr_anno_oe <- data.frame(pchic_gr_anno_bait) %>%
  dplyr::rename(baitChr = seqnames,
                baitSt = start,
                baitEnd = end) %>%
  makeGRangesFromDataFrame(keep.extra.columns = T, seqnames.field = "oeChr", start.field = "oeSt", end.field = "oeEnd")

predictor_cpgs <- readxl::read_excel("resources/epic/horaizon/train_test_split/feature_importances.xlsx") %>%
  dplyr::filter(FeatName != "Random_variable") %>%
  dplyr::left_join(data.frame(epic_anno_gr), by = c("FeatName" = "Name")) %>%
  makeGRangesFromDataFrame(keep.extra.columns = T, seqnames.field = "seqnames", start.field = "start", end.field = "end")

predictor_cpgs$Enhancer_gene <- NA

pchic_anno_oe_predictor_cpg_overlap <- findOverlaps(pchic_gr_anno_oe, predictor_cpgs)

predictor_cpgs$Enhancer_gene[subjectHits(pchic_anno_oe_predictor_cpg_overlap)] <- pchic_gr_anno_oe[queryHits(pchic_anno_oe_predictor_cpg_overlap),]$symbol

genes_of_interest <- paste0(predictor_cpgs$GencodeBasicV12_NAME, ";", predictor_cpgs$Enhancer_gene)
genes_of_interest_unique <- unique(unlist(strsplit(genes_of_interest, ";")))
genes_of_interest_unique <- genes_of_interest_unique[!genes_of_interest_unique %in% c("", "NA")]
```

```{r import rnaseq}
dds <- readRDS("output/rnaseq/deseq2/dds.Rds")
rld <- readRDS("output/rnaseq/deseq2/rld.Rds")
degs_T1_RvNR <- read.csv("output/rnaseq/deseq2/T1_RvNR.csv")[,-1]
degs_T2_RvNR <- read.csv("output/rnaseq/deseq2/T2_RvNR.csv")[,-1]
degs_T2vT1_RvNR <- read.csv("output/rnaseq/deseq2/T2vT1vRvNR.csv")[,-1]

degs_genes_of_interest <- degs_T1_RvNR %>%
  dplyr::filter(HGNC %in% genes_of_interest_unique) %>%
  dplyr::select(ENSGv, ENSG, HGNC, pvalue, padj) %>%
  dplyr::rename(T1RvNRpvalue = pvalue, 
                T1RvNRpadj = padj) %>%
  dplyr::left_join(degs_T2_RvNR %>%
                     dplyr::select(ENSGv, pvalue, padj) %>%
                     dplyr::rename(T2RvNRpvalue = pvalue, 
                                   T2RvNRpadj = padj)) %>%
  dplyr::left_join(degs_T2vT1_RvNR %>%
                     dplyr::select(ENSGv, pvalue, padj) %>%
                     dplyr::rename(T2vT1vRvNRpvalue = pvalue, 
                                   T2vT1vRvNRpadj = padj))

for(i in 1:nrow(degs_genes_of_interest)){
  ggplotobj_T1T2 <- data.frame(Expr = assay(rld)[degs_genes_of_interest[i,"ENSGv"],],
                               Response = colData(rld)$Response,
                               Timepoint = colData(rld)$Timepoint,
                               Donor = colData(rld)$DonorID) %>%
    ggplot(aes(x = Timepoint, y = Expr, col = Response)) +
    stat_summary(fun.data = mean_se, geom = "crossbar",  aes(color = Response)) +
    geom_point(aes(group = Donor, shape = Response), alpha = 0.45, show.legend = F, position = position_dodge(0.1)) +
    geom_line(aes(group = Donor, alpha = 0.45), linetype = "dotted", show.legend = F, position = position_dodge(0.1)) +
    labs(subtitle = paste0("p-value (T2vT1vRvNR) = ", 
                           ifelse(degs_genes_of_interest[i,"T2vT1vRvNRpvalue"]<0.01, 
                                  formatC(x = degs_genes_of_interest[i,"T2vT1vRvNRpvalue"], digits = 3, format = "e"),
                                  round(degs_genes_of_interest[i,"T2vT1vRvNRpvalue"], digits = 3))), 
         y = bquote(log[2]~"(counts+1)")) +
    theme_bw() +
    theme(legend.title = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          legend.position = "bottom",
          text = element_text(size = 12))
  
  ggplotobj_T1 <- data.frame(Expr = assay(rld)[degs_genes_of_interest[i,"ENSGv"],],
                             Response = colData(rld)$Response,
                             Timepoint = colData(rld)$Timepoint,
                             Donor = colData(rld)$DonorID) %>%
    dplyr::filter(Timepoint == "T1") %>%
    ggplot(aes(x = Response, y = Expr, col = Response)) +
    stat_summary(fun.data = mean_se, geom = "crossbar",  aes(color = Response)) +
    geom_jitter(alpha = 0.45, show.legend = F) +
    labs(subtitle = paste0("p-value (T1:RvNR) = ",
                           ifelse(degs_genes_of_interest[i,"T1RvNRpvalue"]<0.01, 
                                  formatC(x = degs_genes_of_interest[i,"T1RvNRpvalue"], digits = 3, format = "e"),
                                  round(degs_genes_of_interest[i,"T1RvNRpvalue"], digits = 3))), 
         y = bquote(log[2]~"(counts+1)")) +
    theme_bw() +
    theme(legend.title = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          legend.position = "bottom",
          text = element_text(size = 12))
  
  ggplotobj_T2 <- data.frame(Expr = assay(rld)[degs_genes_of_interest[i,"ENSGv"],],
                             Response = colData(rld)$Response,
                             Timepoint = colData(rld)$Timepoint,
                             Donor = colData(rld)$DonorID) %>%
    dplyr::filter(Timepoint == "T2") %>%
    ggplot(aes(x = Response, y = Expr, col = Response)) +
    stat_summary(fun.data = mean_se, geom = "crossbar",  aes(color = Response)) +
    geom_jitter(alpha = 0.45, show.legend = F) +
    labs(subtitle = paste0("p-value (T2:RvNR) = ",
                           ifelse(degs_genes_of_interest[i,"T2RvNRpvalue"]<0.01, 
                                  formatC(x = degs_genes_of_interest[i,"T2RvNRpvalue"], digits = 3, format = "e"),
                                  round(degs_genes_of_interest[i,"T2RvNRpvalue"], digits = 3))), 
         y = bquote(log[2]~"(counts+1)")) +
    theme_bw() +
    theme(legend.title = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          legend.position = "bottom",
          text = element_text(size = 12))
  
  ggplotobj <- ggarrange(ggplotobj_T1, ggplotobj_T2, ggplotobj_T1T2, nrow = 1, ncol = 3, align = "hv")
  
  ggplotobj <- annotate_figure(ggplotobj, 
                               top =  text_grob(paste0(ifelse(is.na(degs_genes_of_interest[i,"HGNC"]), "", degs_genes_of_interest[i,"HGNC"]), " (", predictor_cpg_degs_T1_RvNR[i,"ENSGv"], ")"), face = "bold"))
  
  svglite::svglite(filename = paste0(degs_genes_of_interest[i,"ENSG"], ".svg"), width = 15, height = 5)
  print(ggplotobj)
  dev.off()
}
```




# Multi-omics machine learning

To this end, we will utilize a tool developed by [Ding et al. 2022](https://www.pnas.org/doi/10.1073/pnas.2202113119) called multiview cooperative learning. 

```{r}

```

.