---
title: "DM analyses vedolizumab response"
author: "Andrew Y.F. Li Yim"
date: "11 November 2017"
output: html_document
---

In this report we will perform the differential methylation analyses on responders and non-responders towards the α4β7 agonist vedolizumab

From 00_Phenosheet_analysis.Rmd we had already found that there was a slight issue with the genders (i.e. we had female responders and non-responders, but no male non-responders).

Import the relevant data
```{r setup, echo = FALSE}
source("src/dmr_plot.R")
require(minfi)
samplesDir <- file.path("data", "samples")
commondataDir <- file.path("/home", "ayliyim", "archive", "common_data")
outputDir <- file.path("output","02_DM_analyses_vedolizumab")
dir.create(outputDir)

samplesheet <- read.metharray.sheet(base = samplesDir, pattern = "samplesheet_PRJ0000008_CDTNFRESP_V1.csv")
samplesheet$Basename <- samplesheet$Path

samplesheet_vedo <- samplesheet[samplesheet$Drug == "Vedolizumab",]

rgset_vedo <- read.metharray.exp(targets = samplesheet_vedo)
pData(rgset_vedo)$resptime <- with(pData(rgset_vedo), paste0(Response, "_", Timepoint))
```

```{r Quality control, echo = FALSE}
edaDir <- file.path(outputDir, "eda")
dir.create(edaDir)

require(MethylAid)
methylaid_vedo <- MethylAid::summarize(samplesheet_vedo)
visualize(methylaid_vedo)

require(shinyMethyl)
shinymethyl_vedo <- shinyMethyl::shinySummarize(rgset_vedo)
runShinyMethyl(shinymethyl_vedo)

require(ewastools)
ewastools_vedo <- read_idats(samplesheet_vedo$Basename)
ewastools_vedo_betas <- dont_normalize(ewastools_vedo)
ewastools_vedo_snps <- call_genotypes(ewastools_vedo_betas[ewastools_vedo$manifest[probe_type == 'rs', index], ], learn = F)
check_snp_agreement(genotypes = ewastools_vedo_snps$gamma, weights = 1-ewastools_vedo_snps$outliers, donor_ids = samplesheet_vedo$Donor_ID, sample_ids = samplesheet_vedo$Sample_ID)
```

According to MethylAid, all samples pass the quality control. Preprocessing is performed using the preprocessFunnorm due to its usage of noob for background correction and subsequently the control probes for correction. Additionally, the ewastools package was used to confirm that the samples were obtained from the same origin.
As the sample pertains peripheral blood, we can utilize the estimateCellCounts function from minfi to see whether we observe any differences in estimated cell populations between the various groups.

```{r Blood cell distribution}
require("FlowSorted.Blood.EPIC")
require(ExperimentHub)
hub <- ExperimentHub()
FlowSorted.Blood.EPIC <- hub[["EH1136"]]

vedo_cellCounts <- estimateCellCounts(rgset_vedo, 
                                     cellTypes = c("CD8T","CD4T","NK", "Bcell","Mono"),
                                     compositeCellType = "Blood",
                                     processMethod = "auto", 
                                     probeSelect = "auto",
                                     referencePlatform = c("IlluminaHumanMethylationEPIC"))
vedo_cellCounts <- data.frame(vedo_cellCounts, resptime = pData(rgset_vedo)$resptime, array = rownames(vedo_cellCounts))

vedo_cellCounts_melt <- reshape2::melt(vedo_cellCounts, id.vars = c("resptime", "array"), id.vals = c("Response"))

vedo_cellCounts_plot <- ggplot(vedo_cellCounts_melt, aes(x = resptime, y = value)) +
  geom_boxplot() +
  geom_jitter() +
  facet_wrap(~variable) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

Cairo(width = 1000, height = 1000, type = "png", file = file.path(edaDir, "vedo_estimatedCellCounts.png"), bg = "white", units = "px", dpi = 120)
print(vedo_cellCounts_plot)
dev.off()
```

The next step is to remove probes associated to SNPs (MAF > 0.01), sex chromosomes, and those that are promiscuous.

```{r Preprocessing}
gmset_vedo <- preprocessFunnorm(rgSet = rgset_vedo)
gmset_vedo_nosnpcpg <- gmset_vedo[with(getSnpInfo(gmset_vedo), which(CpG_maf < 0.01 | is.na(CpG_maf))),]
gmset_vedo_nosnpsbe <- gmset_vedo_nosnpcpg[with(getSnpInfo(gmset_vedo_nosnpcpg), which(SBE_maf < 0.01 | is.na(SBE_maf))),]
gmset_vedo_nosex <- gmset_vedo_nosnpsbe[-which(getAnnotation(gmset_vedo_nosnpsbe)$chr %in% c("chrX", "chrY")),]

pmprobes <- read.csv(file.path(commondataDir, "IlluminaHumanMethylationBeadChipEPIC", "Non-specific-probes-illuminaEPIC.csv"), stringsAsFactors = F, header = F)[,1]
gmset_vedo_filtered <- gmset_vedo_nosex[!names(gmset_vedo_nosex) %in% pmprobes,]

betas_vedo <- getBeta(gmset_vedo_filtered)
mvals_vedo <- getM(gmset_vedo_filtered)
```

```{r Annotated and predicted sex}
require(ggplot2)
require(dplyr)
require(ggrepel)

pData(gmset_vedo) %>% 
  data.frame %>% 
  select(Sample_ID, Sex, predictedSex, xMed, yMed) %>%
  mutate(Sample_label = ifelse(Sex == predictedSex, "", Sample_ID)) %>% 
  ggplot(aes(x = xMed, y = yMed, col = Sex, label = Sample_label)) +
  geom_text_repel(show.legend = F) +
  geom_point(size = 4) +
  theme_bw() +
  ggtitle("Sex chromosome probe median intensities") +
  xlab("chrX") +
  ylab("chrY") +
  theme(text = element_text(size=20))
```

Four samples derived from 2 donors appear to be mixed up, namely F097 and F096 appear to be labeled incorrectly. After having discussed with the professor in question, it appeared the labels have been mixed up for these two donors.

```{r Mislabeled samples}
mislabels <- pData(gmset_vedo_filtered)[pData(gmset_vedo_filtered)$Donor_ID %in% c("F096", "F097"),]
variables <- c("Sample_ID", "Donor_ID", "Timepoint", "Drug", "Response", "Sex", "Disease", "Age")

pData(gmset_vedo_filtered)["201532190023_R07C01", variables] <- mislabels["201532190023_R08C01", c("Sample_ID", "Donor_ID", "Timepoint", "Drug", "Response", "Sex", "Disease", "Age")]
pData(gmset_vedo_filtered)["201532190023_R08C01", variables] <- mislabels["201532190023_R07C01", c("Sample_ID", "Donor_ID", "Timepoint", "Drug", "Response", "Sex", "Disease", "Age")]
pData(gmset_vedo_filtered)["201506820100_R03C01", variables] <- mislabels["201506820100_R04C01", c("Sample_ID", "Donor_ID", "Timepoint", "Drug", "Response", "Sex", "Disease", "Age")]
pData(gmset_vedo_filtered)["201506820100_R04C01", variables] <- mislabels["201506820100_R03C01", c("Sample_ID", "Donor_ID", "Timepoint", "Drug", "Response", "Sex", "Disease", "Age")]

as.data.frame(pData(gmset_vedo_filtered)[, c("Sex", "predictedSex")])

saveRDS(gmset_vedo_filtered, file.path(outputDir, "gmset_vedo_filtered.Rds"))
```

```{r Sample-sample correlation}
betas_vedo_cor <- cor(betas_vedo)

require(pheatmap)
pheatmap(betas_vedo_cor)
```

In general, the correlation is high (>0.98), which suggests that the methylation of all samples is reasonably similar. We expect that samples derived from the same patient would show the highest pairwise correlation, but given that the correlation is already so high, it would not be strange that we might miss some.

```{r PCA}
require(plotly)

mvals_vedo_dm <- mvals_vedo - rowMeans(mvals_vedo)
mvals_vedo_svd <- svd(t(mvals_vedo_dm))

mvals_svd_df <- data.frame(Sample_ID = pData(gmset_vedo_filtered)$Sample_ID,
                           Origin = pData(gmset_vedo_filtered)$Donor_ID,
                           Timepoint = pData(gmset_vedo_filtered)$Timepoint,
                           Response = pData(gmset_vedo_filtered)$Response,
                           Sex = pData(gmset_vedo_filtered)$predictedSex,
                           PC1 = mvals_vedo_svd$u[,1],
                           PC2 = mvals_vedo_svd$u[,2])

mvals_svd_plotobj <- ggplot(mvals_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Origin = Origin, Timepoint = Timepoint, Response = Response, Sex = Sex)) +
  geom_point(aes(col = Timepoint, shape = Response)) +
  theme_bw() +
  theme(text = element_text(size=20))

ggplotly(mvals_svd_plotobj)
```

There appears to be two samples from one donor that behave distinct from the rest: F139

```{r DM analyses preparation}
require(limma)

design_vedo <- model.matrix(~0 + resptime + Sex + Age, data = pData(gmset_vedo_filtered))
colnames(design_vedo) <- c("T0_NR", "T14_NR", "T0_R", "T14_R", "Male", "Age")
mvals_lmfit_vedo <- lmFit(mvals_vedo, design = design_vedo)
betas_lmfit_vedo <- lmFit(betas_vedo, design = design_vedo)

contrast_mat_vedo <- makeContrasts(
  RvNR = (T0_R+T14_R)/2-(T0_NR+T14_NR)/2,
  T14vT0 = (T14_R+T14_NR)/2-(T0_R+T0_NR)/2,
  T14RvT0R = T14_R-T0_R,
  T14NRvT0NR = T14_NR-T0_NR,
  TR = (T14_R-T0_R)-(T14_NR-T0_NR),
  levels = design_vedo
)

mvals_contrastfit_vedo <- contrasts.fit(mvals_lmfit_vedo, contrast_mat_vedo)
mvals_ebayes_vedo <- eBayes(mvals_contrastfit_vedo)

betas_contrastfit_vedo <- contrasts.fit(betas_lmfit_vedo, contrast_mat_vedo)
betas_ebayes_vedo <- eBayes(betas_contrastfit_vedo)
```

For the first analyses we will investigate whether we can find any differentially methylated positions (DMPs)
```{r DMP analysis, echo = F}
vedodmpDir <- file.path(outputDir, "dmps")
dir.create(vedodmpDir)

gmset_vedo_anno <- getAnnotation(gmset_vedo_filtered)
gmset_vedo_anno_gr <- makeGRangesFromDataFrame(gmset_vedo_anno, keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")

RvNR_vedo_top <- topTable(mvals_ebayes_vedo, coef = "RvNR", number = "Inf", adjust.method = "BH")
RvNR_vedo_top <- cbind(RvNR_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(RvNR_vedo_top), "RvNR"], gmset_vedo_anno[rownames(RvNR_vedo_top), ])
T14vT0_vedo_top <- topTable(mvals_ebayes_vedo, coef = "T14vT0", number = "Inf", adjust.method = "BH")
T14vT0_vedo_top <- cbind(T14vT0_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(T14vT0_vedo_top), "T14vT0"], gmset_vedo_anno[rownames(T14vT0_vedo_top), ])
T14RvT0R_vedo_top <- topTable(mvals_ebayes_vedo, coef = "T14RvT0R", number = "Inf", adjust.method = "BH")
T14RvT0R_vedo_top <- cbind(T14RvT0R_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(T14RvT0R_vedo_top), "T14RvT0R"], gmset_vedo_anno[rownames(T14RvT0R_vedo_top), ])
T14NRvT0NR_vedo_top <- topTable(mvals_ebayes_vedo, coef = "T14NRvT0NR", number = "Inf", adjust.method = "BH")
T14NRvT0NR_vedo_top <- cbind(T14NRvT0NR_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(T14NRvT0NR_vedo_top), "T14NRvT0NR"], gmset_vedo_anno[rownames(T14NRvT0NR_vedo_top), ])
TR_vedo_top <- topTable(mvals_ebayes_vedo, coef = "TR", number = "Inf", adjust.method = "BH")
TR_vedo_top <- cbind(TR_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(TR_vedo_top), "TR"], gmset_vedo_anno[rownames(TR_vedo_top), ])

write.csv(RvNR_vedo_top, file.path(vedodmpDir, "RvNR_vedo.csv"))
write.csv(T14vT0_vedo_top, file.path(vedodmpDir, "T14vT0_vedo.csv"))
write.csv(T14RvT0R_vedo_top, file.path(vedodmpDir, "T14RvT0R_vedo.csv"))
write.csv(T14NRvT0NR_vedo_top, file.path(vedodmpDir, "T14NRvT0NR_vedo.csv"))
write.csv(TR_vedo_top, file.path(vedodmpDir, "TR_vedo.csv"))

RvNR_vedo_sig <- RvNR_vedo_top[RvNR_vedo_top$adj.P.Val < 0.05,]
write.csv(RvNR_vedo_sig, file.path(vedodmpDir, "RvNR_vedo_sig.csv"))
```

Only the responders vs non-responders comparison appears to have yielded interesting results. 

```{r Vedo sig DMP enrichment}
require(missMethyl)

vedo_dmp_enrichment <- gometh(rownames(RvNR_vedo_sig), all.cpg = NULL, collection = "KEGG", array.type = "EPIC", plot.bias = FALSE, prior.prob = TRUE)
vedo_dmp_enrichment <- vedo_dmp_enrichment[order(vedo_dmp_enrichment$P.DE),]
write.csv(vedo_dmp_enrichment, file.path(vedodmpDir, "vedo_dmp_enrichment.csv"))
```

```{r Plot the top 10 RvNR}
require(Cairo)

vedorvnrt10Dir <- file.path(vedodmpDir, "top10")
dir.create(vedorvnrt10Dir)

for(i in 1:10){
  cpg_num <- rownames(RvNR_vedo_top)[i]
  
  #Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(vedorvnrt10Dir, paste0(i, "_", cpg_num, ".pdf")), bg = "white", units = "px", dpi = 120)
  print(NDlib::cpg_strip_plot(cpg_num = cpg_num, betas = betas_vedo, factor_interest = pData(gmset_vedo_filtered)$resptime) + theme(legend.position = "none"))
  #dev.off()
  
  #Cairo(width = 400, height = 400, type = "png", file = file.path(vedorvnrt10Dir, paste0(i, "_", cpg_num, ".png")), bg = "white", units = "px")
  #print(NDlib::cpg_strip_plot(cpg_num = cpg_num, betas = betas_vedo, factor_interest = pData(gmset_vedo_filtered)$resptime) + theme(legend.position = "none"))
  #dev.off()
}
```

One sample in particular appears to display a signal different from the other non-responders when looking at the top RvNR DMP (cg01110291). In fact, it looks like a responder (based on this CpG probe alone).

```{r cg01110291 beta}
require(dplyr)
cg01110291_beta <- data.frame(betas = betas_vedo[rownames(RvNR_vedo_top)[1],], resptime = pData(gmset_vedo_filtered)$resptime, sample = pData(gmset_vedo_filtered)$Sample_ID)

cg01110291_beta %>% filter(grepl("Nonresponder", resptime))
```

It appears to be two samples from donor F096

```{r F096 pdata}
pData(gmset_vedo_filtered) %>% data.frame() %>% filter(Donor_ID == "F096")
```

Nothing peculiar suggests that this donor is different from the rest.

```{r cg24051749 surrounding CpGs}
gmset_vedo_anno_gr[(which(names(gmset_vedo_anno_gr) == "cg24051749")-2):(which(names(gmset_vedo_anno_gr) == "cg24051749")+2),]
```

```{r cg24051749 region plot}
print(dmr_plot(dmr_gr = makeGRangesFromDataFrame(df = RvNR_vedo_top[4,], seqnames.field = "chr", start.field = "pos", end.field = "pos"),
               flanks = 500,
               meth_data = betas_vedo, 
               anno_gr = gmset_vedo_anno_gr, 
               meth_groups = pData(gmset_vedo_filtered)$Response, 
               title = "cg24051749 100bps"))
```

```{r Genes with multiple vedoresp-DMPs}
vedorvnrmdmpsDir <- file.path(vedodmpDir, "mdmps")
dir.create(vedorvnrmdmpsDir)

vedosig_genes <- do.call(c, lapply(strsplit(x = RvNR_vedo_sig$UCSC_RefGene_Name, ";"), unique))
vedosig_genes_freq <- sort(table(vedosig_genes))
vedosig_genes_mDMPs <- names(which(vedosig_genes_freq > 1))

vedosig_genes_mDMPs_min <- do.call(rbind, lapply(vedosig_genes_mDMPs, function(i) {
  df <- RvNR_vedo_sig[grep(i, RvNR_vedo_sig$UCSC_RefGene_Name),]
  df_min <- with(df, data.frame(seqnames = unique(chr),
                                start = min(pos),
                                end = max(pos),
                                gene = i))
  
  return(df_min)
}))

vedosig_genes_mDMPs_gr <- makeGRangesFromDataFrame(vedosig_genes_mDMPs_min, keep.extra.columns = T)

for(i in 1:length(vedosig_genes_mDMPs_gr)){
  Cairo(width = 2000, height = 2000, type = "pdf", file = file.path(vedorvnrmdmpsDir, paste0(vedosig_genes_mDMPs_gr$gene[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(dmr_plot(dmr_gr = vedosig_genes_mDMPs_gr[i], meth_data = betas_vedo, anno_gr = gmset_vedo_anno_gr, meth_groups = pData(gmset_vedo_filtered)$Response, title = as.character(vedosig_genes_mDMPs_gr$gene[i])))
  dev.off()
  
    Cairo(width = 500, height = 500, type = "png", file = file.path(vedorvnrmdmpsDir, paste0(vedosig_genes_mDMPs_gr$gene[i], ".png")), bg = "white", units = "px")
  print(dmr_plot(dmr_gr = vedosig_genes_mDMPs_gr[i], meth_data = betas_vedo, anno_gr = gmset_vedo_anno_gr, meth_groups = pData(gmset_vedo_filtered)$Response, title = as.character(vedosig_genes_mDMPs_gr$gene[i])))
  dev.off()
}
```

```{r dmrs}
vedodmrDir <- file.path(outputDir, "dmrs")
dir.create(vedodmrDir)

require(DMRcate)
dmr_anno_rvnr_vedo <- cpg.annotate(object = mvals_vedo,
                                  datatype = "array",
                                  arraytype = "EPIC",
                                  what = "M",
                                  analysis.type = "differential",
                                  design = design_vedo,
                                  contrasts = T,
                                  cont.matrix = contrast_mat_vedo,
                                  coef = "RvNR")
dmr_rvnr_vedo <- dmrcate(dmr_anno_rvnr_vedo, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
dmr_rvnr_vedo_gr <- extractRanges(dmr_rvnr_vedo, genome = "hg19")

write.csv(as.data.frame(dmr_rvnr_vedo_gr), file.path(vedodmrDir, "dmrs_vedo.csv"))

dmr_plot(dmr_gr = dmr_rvnr_vedo_gr[1], meth_data = betas_vedo, anno_gr = gmset_vedo_anno_gr, meth_groups = pData(gmset_vedo_filtered)$Response)
```