---
title: "01_candidate_sample_identification"
author: "Andrew Y.F. Li Yim"
date: "5/7/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

We previously observed drug response-associated differentially methylated positions (DMPs). The goal of this subproject is to validate whether the methylation can be observed using a different technology as well as identify whether the observed difference inmethylation is simply an artefact caused by differences in genetic backgrounds. Many of the reported DMPs display a pattern that suggests that a genetic variant underlies the observed signal. To this end, we seek to use the Nanopore to proverbially "kill two birds with one stone", as the Nanopore reportedly can elucidate the sequence as well as the methylation status without the need for a bisulfite conversion. However, to verify that the Nanopore is accurate, we will also perform Sanger sequencing and bisulfite sequencing to identify the underlying DNA as well as the methylation signal, respectively. 

This workbook is part of a small pilot project where we seek to investigate the methylation signal for a subset of the drug response-associated DMPs (cg26913836, cg01379846, cg15152149, cg13679303, cg13653328, cg09163005, and cg11214014). As the actual samples are scarce and precious, we will use a set of samples from a different study where material is more abundant. The goal of this notebook is therefore to identify 3 samples from PRJ0000021_12QDIAGPILOT suitable for validating the Nanopore validation method.

Most drug response-associated DMPs display the dual/triple clustering pattern typical of underlying genetic variants. Preferably, we can identify 3 samples, each of which belong to the different groups. 

```{r setup}
require(minfi)
require(dplyr)

dataDir <- file.path("data", "technical_validation", "pilot")
outDir <- file.path("output", "02_DM_validation")

samples <- read.csv(file.path(dataDir, "samples_SETD1B.csv"), stringsAsFactors = F, sep = "\t")
samples$Basename <- paste0("~/archive/", samples$Path, samples$Plate, "/", samples$Plate, "_", samples$Pos)
samples$ID <- paste0(samples$Plate, "_", samples$Pos)
samples <- samples %>%
  filter(Sample_Group == "C")

rgset <- read.metharray.exp(targets = samples)
gmset <- preprocessFunnorm(rgSet = rgset)
betas <- getBeta(gmset)
mannotations <- getAnnotation(gmset)
```

```{r cpgs of interest}
require(ggplot2)

cpgs_of_interest <- read.csv("/home/ayliyim/archive/projects/PRJ0000008_CDMEDRESP/NANOPORE_PILOT/sanger_validation/nanopore_targets.csv", stringsAsFactors = F)
betas_of_interest <- betas[cpgs_of_interest$CpG,]

betas_of_interest_melt <- reshape2::melt(betas_of_interest)
colnames(betas_of_interest_melt) <- c("CpG", "Sample", "Beta")

beta_distribution_pilot_cpgs <- ggplot(betas_of_interest_melt, aes(x = CpG, y = Beta)) +
  geom_jitter() +
  theme_bw() +
  ylim(0,1)
  #facet_grid(.~CpG, scales = "free_x")

Cairo(file = file.path(outDir, "beta_distribution_pilot_cpgs.pdf"), type = "pdf", units = "px", width = 1500, height = 500, dpi = 90, bg = "white")
print(beta_distribution_pilot_cpgs)
dev.off()
```

We see roughly the same pattern as was observed for the drug response-associated methylation with the exception of cg01379846, cg13679303 and cg09163005, which are the DMPs that are most likely not influenced by SNPs. We would like to find three samples that span the different clusters we observe.  

```{r visualizing cpgs of interest}
meth_cpgs <- c("cg01379846", "cg13679303", "cg09163005")

betas_of_interest_melt_filtered <- betas_of_interest_melt %>%
  filter(!CpG %in% meth_cpgs)

beta_distribution_pilot_cpgs_filtered <- ggplot(betas_of_interest_melt_filtered, aes(x = CpG, y = Beta, group = Sample, col = Sample)) +
  geom_jitter() +
  #geom_line() +
  theme_bw() +
  ylim(0,1) +
  theme(legend.pos = "none")

Cairo(file = file.path(outDir, "beta_distribution_pilot_cpgs_filtered.pdf"), type = "pdf", units = "px", width = 1250, height = 500, dpi = 90, bg = "white")
print(beta_distribution_pilot_cpgs_filtered)
dev.off()

ggplotly(beta_distribution_pilot_cpgs_filtered)
```

One approach of identifying such samples would be to find samples that are distinct from one another. This distinctness can somewhat be visualized through principal component analysis (PCA).

```{r pca cpgs of interest}
require(plotly)

betas_of_interest_filtered <- betas_of_interest[!rownames(betas_of_interest) %in% meth_cpgs,]

betas_of_interest_dm <- betas_of_interest_filtered - rowMeans(betas_of_interest_filtered)
betas_of_interest_svd <- svd(t(betas_of_interest_dm))

svd_df <- data.frame(PC1 = betas_of_interest_svd$u[,1],
                     PC2 = betas_of_interest_svd$u[,2],
                     samples = colnames(betas_of_interest_dm))

svd_plot <- ggplot(svd_df, aes(x = PC1, y = PC2, ID = samples, col = samples)) +
  geom_point() +
  theme_bw()

ggplotly(svd_plot)
```

There appear to be 6 separate clusters on PC1 and PC2. As we want samples that span all the different clusters observed for methylation, we make the assumption that the most distant samples on the PCA would be proper representatives. Samples to choose from:
- 201503590101_R04C01 (16D3898)
- 201524770078_R03C01 (15D6787)
- 202128330112_R07C01 (15D3563)

```{r samples of interest}
samples_of_interest <- c("201503590101_R04C01", "201524770078_R03C01", "202128330112_R07C01")

betas_of_interest_melt_filtered$alpha <- 0.25
betas_of_interest_melt_filtered$alpha[betas_of_interest_melt_filtered$Sample %in% samples_of_interest] <- 1

selected_lineplot <- ggplot(betas_of_interest_melt_filtered, aes(x = CpG, y = Beta, group = Sample, col = Sample, alpha = alpha)) +
  geom_point() +
  geom_line(size = 2) +
  theme_bw() +
  ylim(0,1) +
  theme(legend.pos = "none")

Cairo(file = file.path(outDir, "selected_samples.pdf"), type = "pdf", units = "px", width = 800, height = 800, dpi = 90, bg = "white")
print(selected_lineplot)
dev.off()

samples_of_interest_df <- samples[samples$ID %in% samples_of_interest,]
write.csv(samples_of_interest_df, file.path(outDir, "samples_of_interest.csv"))
```

```{r Samples of interest only}
betas_of_interest_melt_samples_of_interest <- betas_of_interest_melt_filtered %>% 
  filter(alpha == 1) %>%
  mutate(Sample = as.character(Sample)) %>%
  mutate(Sample = replace(Sample, Sample == "201524770078_R03C01", "16D3898")) %>%
  mutate(Sample = replace(Sample, Sample == "201503590101_R04C01", "15D6787")) %>%
  mutate(Sample = replace(Sample, Sample == "202128330112_R07C01", "15D3563"))

selected_lineplot_soi <- ggplot(betas_of_interest_melt_samples_of_interest, aes(x = CpG, y = Beta, group = Sample, col = Sample)) +
  geom_point() +
  geom_line(size = 2) +
  theme_bw() +
  ylim(0,1) +
  theme(legend.pos = "bottom")

Cairo(file = file.path(outDir, "selected_samples_only.pdf"), type = "pdf", units = "px", width = 800, height = 800, dpi = 90, bg = "white")
selected_lineplot_soi
dev.off()
```

```{r}
boi_soi <- betas_of_interest[!rownames(betas_of_interest) %in% meth_cpgs, samples_of_interest]
boi_soi <- data.frame(boi_soi, mannotations[rownames(boi_soi),c("CpG_rs", "CpG_maf", "chr", "pos")])
colnames(boi_soi)[1:3] <- c("15D6787", "16D3898", "15D3563")

write.csv(boi_soi, file.path(outDir, "betas_samples_of_interest.csv"))
```

