---
title: "01_DM_validation"
author: "Andrew Y.F. Li Yim"
date: "9/18/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r packages}
require(minfi)
require(limma)
require(MethylAid)
require(shinyMethyl)
require(ewastools)
require(FlowSorted.Blood.EPIC)
require(ExperimentHub)
require(ggrastr)
```

```{r setup, include=FALSE}
response_cols <- c(R = "#E69F00", NR = "#56B4E9")

commondataDir <- file.path("~", "lkgres", "common_data")
samplesDir <- file.path("data", "samples", "02_validation")
outputDir <- file.path("output", "02_DM_validation")
rdsDir <- file.path(outputDir, "rds")
```

```{r import methylation data}
samplesheet_validation1 <- read.csv(file.path(samplesDir, "211223_samples_vedolizumab_oxford_validation1_PRJ0000008_CDMEDRESP.csv"))
samplesheet_validation1$Basename <- samplesheet_validation1$Path

#prepare rgset
rgset_vedo <- read.metharray.exp(targets = samplesheet_validation1)
```

```{r import ancillary data}
pmprobes <- read.csv(file.path(commondataDir, "IlluminaHumanMethylationBeadChipEPIC", "Non-specific-probes-illuminaEPIC.csv"), stringsAsFactors = F, header = F)[,1]
```

## Quality control

```{r quality control, echo = FALSE}
qcDir <- file.path(outputDir, "qc")
dir.create(qcDir)

methylaid_obj <- summarize(samplesheet_validation1, force = T)
visualize(methylaid_obj)

shinymethyl_obj <- shinySummarize(rgset_vedo)
runShinyMethyl(shinymethyl_obj)

ewastools_obj <- read_idats(samplesheet_validation1$Basename)
ewastools_betas <- dont_normalize(ewastools_obj)
ewastools_snps <- call_genotypes(ewastools_betas[ewastools_obj$manifest[probe_type == 'rs', index], ], learn = F)
ewastools_snpagreement <- check_snp_agreement(genotypes = ewastools_snps, 
                                              donor_ids = samplesheet_validation1$PATIENT_NAME, 
                                              sample_ids = paste0(samplesheet_validation1$SXS, "_", samplesheet_validation1$POS))

saveRDS(ewastools_snps, file.path(rdsDir, "ewastools_snps.Rds"))
#cleanup
rm(ewastools_obj, ewastools_betas, ewastools_snps, ewastools_snpagreement)
gc()
```

MethylAid unfortunately did not work, shinyMethyl worked and suggested no strange outliers besides patient 6975. ewastools suggested no oddities when comparing the genetic variant profile with the sample IDs.

```{r snp quality control}
ewastools_snpbetas_cor <- cor(ewastools_betas[grep("rs", rownames(ewastools_betas)),])

Cairo::Cairo(width = 1700, height = 1700, type = "pdf", file = file.path(qcDir, "snp_pairwisecorrelations.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap::pheatmap(mat = ewastools_snpbetas_cor,
                   annotation_col = data.frame(Patient = samplesheet_validation1$PATIENT_NAME,
                                               row.names = paste0(samplesheet_validation1$SXS, "_", samplesheet_validation1$POS)),
                   annotation_row = data.frame(Patient = samplesheet_validation1$PATIENT_NAME,
                                               row.names = paste0(samplesheet_validation1$SXS, "_", samplesheet_validation1$POS)),
                   labels_row = samplesheet_validation1$PATIENT_NAME)
dev.off()
```

From ewastools it is clear that samples identified as being from the same individual, namely Pat_1816 and Pat_6885 are derived from the same individual.

## Exploratory data analyses

```{r eda setup}
edaDir <- file.path(outputDir, "eda")
dir.create(edaDir)
```

```{r cell distribution}
if(!file.exists(file.path(edaDir, "estimatedCellProportion.csv"))){
  hub <- ExperimentHub()
  FlowSorted.Blood.EPIC <- hub[["EH1136"]]
  
  cellDist <- estimateCellCounts(rgset_vedo, 
                                 cellTypes = c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Neu"),
                                 compositeCellType = "Blood",
                                 processMethod = "auto", 
                                 probeSelect = "auto",
                                 referencePlatform = c("IlluminaHumanMethylationEPIC"))
  cellDist <- data.frame(cellDist, 
                         response = pData(rgset_vedo)$Response,
                         donor = pData(rgset_vedo)$PATIENT_NAME,
                         array = rownames(cellDist))
  write.csv(cellDist, file.path(edaDir, "estimatedCellProportion.csv"))
} else{
  cellDist <- read.csv(file.path(edaDir, "estimatedCellProportion.csv"), row.names = 1, stringsAsFactors = F)
}
```

```{r preprocessing}
gmset_vedo <- preprocessFunnorm(rgSet = rgset_vedo)
#gmset_vedo_processed <- gmset_vedo[with(getSnpInfo(gmset_vedo), which(CpG_maf < 0.01 | is.na(CpG_maf))),]
#gmset_vedo_processed <- gmset_vedo[with(getSnpInfo(gmset_vedo), which(SBE_maf < 0.01 | is.na(SBE_maf))),]
gmset_vedo_processed <- gmset_vedo[-which(minfi::getAnnotation(gmset_vedo)$chr %in% c("chrX", "chrY")),]

gmset_vedo_filtered <- gmset_vedo_processed[!names(gmset_vedo_processed) %in% pmprobes,]
gmset_vedo_filtered <- gmset_vedo_filtered[-which(getM(gmset_vedo_filtered) == -Inf, arr.ind = T)[,1],]

betas_vedo <- getBeta(gmset_vedo_filtered)
mvals_vedo <- getM(gmset_vedo_filtered)
```

```{r sex}
table(pData(gmset_vedo_filtered)$Sex, pData(gmset_vedo_filtered)$predictedSex)
```

```{r pca}
mvals_vedo_dm <- mvals_vedo - rowMeans(mvals_vedo)
mvals_vedo_svd <- svd(t(mvals_vedo_dm))
mvals_vedo_svd_expvar <- round(mvals_vedo_svd$d/sum(mvals_vedo_svd$d), 2)*100

mvals_vedo_svd_df <- data.frame(Sample_ID = pData(gmset_vedo_filtered)$PATIENT_NAME,
                                Plate = pData(gmset_vedo_filtered)$METH_PLATE,
                                Batch = pData(gmset_vedo_filtered)$METH_BATCH,
                                Disease = pData(gmset_vedo_filtered)$Dx,
                                Response = pData(gmset_vedo_filtered)$Response,
                                Sex = pData(gmset_vedo_filtered)$Sex,
                                Age = pData(gmset_vedo_filtered)$Age,
                                PC1 = mvals_vedo_svd$u[,1],
                                PC2 = mvals_vedo_svd$u[,2])

mvals_vedo_svd_plotobj <- ggplot(mvals_vedo_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Response = Response, Sex = Sex, Age = Age)) +
  geom_point(aes(col = Response), size = 3) +
  theme_bw() +
  xlab(paste0("PC1 (", mvals_vedo_svd_expvar[1], "%)")) +
  ylab(paste0("PC2 (", mvals_vedo_svd_expvar[2], "%)")) +
  scale_fill_manual(values = response_cols) +
  scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())

Cairo::Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(edaDir, "pca.pdf"), bg = "white", units = "px", dpi = 120)
print(mvals_vedo_svd_plotobj)
dev.off()
```

# Differential analyses

```{r known responses}
gmset_vedo_filtered_known <- gmset_vedo_filtered[,pData(gmset_vedo_filtered)$Response %in% c("Y", "N")]
```

## All

```{r RvNR all setup}
RvNR_allDir <- file.path(outputDir, "RvNR_all")
dir.create(RvNR_allDir)
```

```{r RvNR all dmp analysis}
design_vedo_all <- model.matrix(~Response + Sex + Age, data = pData(gmset_vedo_filtered_known))
colnames(design_vedo_all) <- c("(Intercept)", "R", "M", "Age")

mvals_lmfit_vedo_all <- lmFit(getM(gmset_vedo_filtered_known), design = design_vedo_all)
betas_lmfit_vedo_all <- lmFit(getBeta(gmset_vedo_filtered_known), design = design_vedo_all)

mvals_ebayes_vedo_all <- eBayes(mvals_lmfit_vedo_all)
betas_ebayes_vedo_all <- eBayes(betas_lmfit_vedo_all)

RvNR_vedo_top_all <- topTable(mvals_ebayes_vedo_all, coef = "R", number = "Inf", adjust.method = "BH")
RvNR_vedo_top_all <- cbind(RvNR_vedo_top_all, 
                           Beta = coefficients(betas_ebayes_vedo_all)[rownames(RvNR_vedo_top_all), "R"],
                           data.frame(getAnnotation(gmset_vedo_filtered_known)[rownames(RvNR_vedo_top_all), ]))
write.csv(RvNR_vedo_top_all, file.path(RvNR_allDir, "RvNR_all.csv"))
```

## CD only

```{r RvNR CD setup}
RvNR_CDDir <- file.path(outputDir, "RvNR_CD")
dir.create(RvNR_CDDir)

gmset_vedo_filtered_known_CD <- gmset_vedo_filtered_known[,pData(gmset_vedo_filtered_known)$Dx == "CD"]
```

```{r RvNR CD dmp analysis}
design_vedo_CD <- model.matrix(~Response + Sex + Age, data = pData(gmset_vedo_filtered_known_CD))
colnames(design_vedo_CD) <- c("(Intercept)", "R", "M", "Age")

mvals_lmfit_vedo_CD <- lmFit(getM(gmset_vedo_filtered_known_CD), design = design_vedo_CD)
betas_lmfit_vedo_CD <- lmFit(getBeta(gmset_vedo_filtered_known_CD), design = design_vedo_CD)

mvals_ebayes_vedo_CD <- eBayes(mvals_lmfit_vedo_CD)
betas_ebayes_vedo_CD <- eBayes(betas_lmfit_vedo_CD)

RvNR_vedo_top_CD <- topTable(mvals_ebayes_vedo_CD, coef = "R", number = "Inf", adjust.method = "BH")
RvNR_vedo_top_CD <- cbind(RvNR_vedo_top_CD, 
                          Beta = coefficients(betas_ebayes_vedo_CD)[rownames(RvNR_vedo_top_CD), "R"], 
                          data.frame(getAnnotation(gmset_vedo_filtered_known_CD)[rownames(RvNR_vedo_top_CD), ]))
write.csv(RvNR_vedo_top_CD, file.path(RvNR_CDDir, "RvNR_CD.csv"))

NDlib::cpg_strip_plot(cpg_num = "cg13458596", 
                      betas = getBeta(gmset_vedo_filtered_known_CD), 
                      pData(gmset_vedo_filtered_known_CD)$Response)
```

# Comparison with discovery data

## Linear comparisons

```{r import discovery data}
RvNR_vedo_discovery <- read.csv(file.path("output", "01_DM_analyses", "vedolizumab", "200323", "dmps", "RvNRDir", "RvNR_vedo.csv"), stringsAsFactors = F, row.names = 1)
```

```{r merge results}
RvNR_comparison <- merge(x = RvNR_vedo_discovery[, c("Name", "logFC", "t", "P.Value", "adj.P.Val", "Beta")], 
                         y = RvNR_vedo_top_CD, 
                         by = "Name")
RvNR_comparison$Discovery_sig <- "Non-significant"
RvNR_comparison$Discovery_sig[RvNR_comparison$adj.P.Val.x<0.05] <- "Significant"

ggplot(RvNR_comparison, aes(x = Beta.x, y = Beta.y, alpha = Discovery_sig)) +
  geom_point_rast() +
  theme_bw()

```

## Elastic net prediction

```{r import elnet predictor cpgs}
elnet_stablor_cpgs <- read.csv(file.path("output", "01_DM_analyses", "vedolizumab", "200323", "classification", "elasticnet", "median_stabilized_lor_df.csv"), stringsAsFactors = F, row.names = 1)
```

```{r validation predictions}
val_odds <- exp(elnet_stablor_cpgs$med_stablor %*% mvals_vedo[elnet_stablor_cpgs$Name,])
val_probs <- val_odds/(1+val_odds)

data.frame(probs = t(val_probs),
           Disease = pData(gmset_vedo_filtered)$Dx,
           Response = pData(gmset_vedo_filtered)$Response,
           Sex = pData(gmset_vedo_filtered)$Sex) %>%
  dplyr::filter(Disease == "CD") %>%
  ggplot(aes(x = Response, y = probs)) +
  geom_boxplot() +
  labs(title = "Probability of Response") +
  ylab("Probability") +
  ylim(0, 1) +
  geom_jitter() +
  theme_bw()
```

## Gradient boosting machine

```{r import gbm predictor cpgs}
gbm_stablor_cpgs <- read.csv(file.path("output", "01_DM_analyses", "vedolizumab", "200323", "classification", "gbm", "noob_nosnp",  "feat_imp_pairs.txt"), stringsAsFactors = F, sep = "\t")
gbm_stablor_cpgs_df <- data.frame(CpG = gsub("(cg[0-9]+)_.+$", "\\1", gbm_stablor_cpgs$FeatName),
                                  Gene = gsub("cg[0-9]+_(.+)$", "\\1", gbm_stablor_cpgs$FeatName))

mvals_vedo_dm_gbmcpgs <- mvals_vedo_dm[gbm_stablor_cpgs_df$CpG,]
mvals_vedo_dm_gbmcpgs_svd <- svd(t(mvals_vedo_dm_gbmcpgs))
mvals_vedo_dm_gbmcpgs_svd_expvar <- round(mvals_vedo_dm_gbmcpgs_svd$d/sum(mvals_vedo_dm_gbmcpgs_svd$d), 2)*100

mvals_vedo_svd_gbm_df <- data.frame(Sample_ID = pData(gmset_vedo_filtered)$PATIENT_NAME,
                                    Plate = pData(gmset_vedo_filtered)$METH_PLATE,
                                    Batch = pData(gmset_vedo_filtered)$METH_BATCH,
                                    Disease = pData(gmset_vedo_filtered)$Dx,
                                    Response = pData(gmset_vedo_filtered)$Response,
                                    Sex = pData(gmset_vedo_filtered)$Sex,
                                    Age = pData(gmset_vedo_filtered)$Age,
                                    PC1 = mvals_vedo_dm_gbmcpgs_svd$u[,1],
                                    PC2 = mvals_vedo_dm_gbmcpgs_svd$u[,2])

mvals_vedo_svd_gbm_plotobj <- ggplot(mvals_vedo_svd_gbm_df, aes(x = PC1, y = PC2, Label = Sample_ID, Response = Response, Sex = Sex, Age = Age)) +
  geom_point(aes(col = Response), size = 3) +
  theme_bw() +
  xlab(paste0("PC1 (", mvals_vedo_dm_gbmcpgs_svd_expvar[1], "%)")) +
  ylab(paste0("PC2 (", mvals_vedo_dm_gbmcpgs_svd_expvar[2], "%)")) +
  scale_fill_manual(values = response_cols) +
  scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())

```

# Prepare data for GBM

```{r processed data setup}
processedDir <- file.path(outputDir, "processed_data")
dir.create(processedDir)
```

```{r with SNPs}
gmset_noob_vedo <- preprocessNoob(rgset_vedo)
gmset_noob_vedo <- mapToGenome(gmset_noob_vedo)

cpgs <- gsub("^(cg[0-9]+|ch.+(R|F))_.+$", "\\1", read.csv("data/cpgs/cpgs_t.csv", stringsAsFactors = F)[-1,1])

gmset_noob_vedo_all <- gmset_noob_vedo[cpgs,]

betas_noob_vedo <- data.frame(Response = pData(gmset_noob_vedo_all)$Response,
                              t(minfi::getBeta(gmset_noob_vedo_all)),
                              row.names = rownames(pData(gmset_noob_vedo_all)))

mod_cpgsids <- paste0(rownames(gmset_noob_vedo_all), "_", unlist(lapply(strsplit(getAnnotation(gmset_noob_vedo_all)[,"UCSC_RefGene_Name"], ";"), function(i) {i <- paste(unique(i), collapse = ";"); ifelse(length(i) == 0, NA, i)})))
mod_cpgsids <- gsub("^(cg[0-9]{8})_$", "\\1", mod_cpgsids)
colnames(betas_noob_vedo) <- c("Response", mod_cpgsids)

data.table::fwrite(betas_noob_vedo, file.path(processedDir, "PRJ0000008_CDMEDRESP_vedolizumab_validation1_betas_noob.csv"))
```

```{r without SNPs}
cpgs_nosnp <- gsub("^(cg[0-9]+|ch.+(R|F))_.+$", "\\1", read.csv("data/cpgs_nosnp_t.csv", stringsAsFactors = F)[-1,1])
cpgs_nosnp <- gsub("ch.1.3642016F_TPR", "ch.1.3642016F", cpgs_nosnp)

gmset_noob_vedo_nosnp <- gmset_noob_vedo[cpgs_nosnp,]

betas_noob_vedo_nosnp <- data.frame(Response = pData(gmset_noob_vedo_nosnp)$Response,
                                    t(minfi::getBeta(gmset_noob_vedo_nosnp)),
                                    row.names = rownames(pData(gmset_noob_vedo_nosnp)))

data.table::fwrite(betas_noob_vedo_nosnp, file.path(processedDir, "PRJ0000008_CDMEDRESP_vedolizumab_validation1_betas_noob_nosnp.csv"))
```

## Gradient boosting machine

```{r import gbm predictor cpgs}
gbm_stablor_cpgs <- read.csv(file.path("output", "01_DM_analyses", "vedolizumab", "200323", "classification", "gbm", "noob_nosnp",  "feat_imp_pairs.txt"), stringsAsFactors = F, sep = "\t")
gbm_stablor_cpgs_df <- data.frame(CpG = gsub("(cg[0-9]+)_.+$", "\\1", gbm_stablor_cpgs$FeatName),
                                  Gene = gsub("cg[0-9]+_(.+)$", "\\1", gbm_stablor_cpgs$FeatName))

mvals_vedo_dm_gbmcpgs <- mvals_vedo_dm[gbm_stablor_cpgs_df$CpG,]
mvals_vedo_dm_gbmcpgs_svd <- svd(t(mvals_vedo_dm_gbmcpgs))
mvals_vedo_dm_gbmcpgs_svd_expvar <- round(mvals_vedo_dm_gbmcpgs_svd$d/sum(mvals_vedo_dm_gbmcpgs_svd$d), 2)*100

mvals_vedo_svd_gbm_df <- data.frame(Sample_ID = pData(gmset_vedo_filtered)$PATIENT_NAME,
                                    Plate = pData(gmset_vedo_filtered)$METH_PLATE,
                                    Batch = pData(gmset_vedo_filtered)$METH_BATCH,
                                    Disease = pData(gmset_vedo_filtered)$Dx,
                                    Response = pData(gmset_vedo_filtered)$Response,
                                    Sex = pData(gmset_vedo_filtered)$Sex,
                                    Age = pData(gmset_vedo_filtered)$Age,
                                    PC1 = mvals_vedo_dm_gbmcpgs_svd$u[,1],
                                    PC2 = mvals_vedo_dm_gbmcpgs_svd$u[,2])

mvals_vedo_svd_gbm_plotobj <- ggplot(mvals_vedo_svd_gbm_df, aes(x = PC1, y = PC2, Label = Sample_ID, Response = Response, Sex = Sex, Age = Age)) +
  geom_point(aes(col = Response), size = 3) +
  theme_bw() +
  xlab(paste0("PC1 (", mvals_vedo_dm_gbmcpgs_svd_expvar[1], "%)")) +
  ylab(paste0("PC2 (", mvals_vedo_dm_gbmcpgs_svd_expvar[2], "%)")) +
  scale_fill_manual(values = response_cols) +
  scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())

```