---
title: "Differential Methylation Analyses Adalimumab Discovery"
author: "Andrew Y.F. Li Yim"
date: "07/6/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r packages import, echo = FALSE}
require(minfi)
require(ggplot2)
require(Cairo)
require(dplyr)
require(ewastools)
require(shinyMethyl)
require(rkgreslib)
```

```{r directories setup, echo = FALSE}
#input
samplesDir <- file.path("data", "samples", "02_validation")

#output
outputDir <- file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906")
dir.create(outputDir)
qcDir <- file.path(outputDir, "qc")
dir.create(qcDir)
rdsDir <- file.path(outputDir, "rds")
dir.create(rdsDir)
```

# Import and normalization

```{r import and preprocessing}
#import samples
samples <- readxl::read_excel(file.path(samplesDir, "210906_samples_vedolizumab_aumc_validation1_PRJ0000008_CDMEDRESP.xlsx"))
samples$Basename <- samples$Path
samples$resptime <- with(samples, paste0(Response, "_", Timepoint))

#prepare rgset
rgset_vedoval <- read.metharray.exp(targets = samples, force = T)
```

# QC

```{r ewastools}
ewastools_combined <- read_idats(samples$Basename)
ewastools_combined_betas <- dont_normalize(ewastools_combined)
ewastools_combined_snps <- call_genotypes(ewastools_combined_betas[ewastools_combined$manifest[probe_type == 'rs', index], ], learn = F)
ewastools_combined_snpagreement <- check_snp_agreement(genotypes = ewastools_combined_snps, 
                                                       donor_ids = samples$Biobank_ID, 
                                                       sample_ids = paste0(samples$Sentrix_ID, "_", samples$Sentrix_Position))
```

F241 sample mismatch.

```{r}
Cairo(width = 2500, height = 2000, type = "pdf", file = file.path(qcDir, "Biobank_mismatch.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap::pheatmap(cor(ewastools_combined_betas[rownames(ewastools_combined_snps$snps),]), 
                   annotation_col = data.frame(row.names = paste0(samples$Sentrix_ID, "_", samples$Sentrix_Position),
                                               BiobankID = samples$Biobank_ID), 
                   labels_col = samples$Biobank_ID)
dev.off()
```


```{r shinyMethyl}
rgset_vedoval_ss <- shinySummarize(rgset_vedoval)
runShinyMethyl(rgset_vedoval_ss)
```

# Preprocess for HORAIZON

```{r horaizon T1}
rgset_vedoval_t1 <- rgset_vedoval[,pData(rgset_vedoval)$Timepoint == "T1"]

rgset_vedoval_t1_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t1,
                                                       colname_of_interest = "Response", 
                                                       variable_of_interest = "R",
                                                       allosome_removal = T, 
                                                       snp_removal = F, 
                                                       outpath = file.path("output", "210607", "horaizon", "PRJ0000008_CDMEDRESP_infliximab_discovery_T1_betas_noob"))
rgset_vedoval_t1_horaizon_nosnp <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t1,
                                                             colname_of_interest = "Response", 
                                                             variable_of_interest = "R",
                                                             allosome_removal = T, 
                                                             snp_removal = T, 
                                                             min_maf = 0, 
                                                             outpath = file.path("output", "210607", "horaizon", "PRJ0000008_CDMEDRESP_infliximab_discovery_T1_betas_noob_nosnp"))
rgset_vedoval_t1_horaizon_nosnp_nogap <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t1,
                                                                   colname_of_interest = "Response", 
                                                                   variable_of_interest = "R", 
                                                                   allosome_removal = T, 
                                                                   snp_removal = T,
                                                                   min_maf = 0, 
                                                                   gap_removal = T,
                                                                   gap_threshold = 0.2,
                                                                   outpath = file.path("output", "210607", "horaizon", "PRJ0000008_CDMEDRESP_infliximab_discovery_T1_betas_noob_nosnp_nogap"))
```

```{r horaizon T2}
rgset_vedoval_t2 <- rgset_vedoval[,pData(rgset_vedoval)$Timepoint == "T2"]

rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t2,
                              colname_of_interest = "Response", 
                              variable_of_interest = "R",
                              allosome_removal = T, 
                              snp_removal = F, 
                              outpath = file.path("output", "210607", "horaizon", "PRJ0000008_CDMEDRESP_infliximab_discovery_T2_betas_noob"))
rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t2,
                              colname_of_interest = "Response", 
                              variable_of_interest = "R",
                              allosome_removal = T, 
                              snp_removal = T, 
                              min_maf = 0, 
                              outpath = file.path("output", "210607", "horaizon", "PRJ0000008_CDMEDRESP_infliximab_discovery_T2_betas_noob_nosnp"))
rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t2,
                              colname_of_interest = "Response", 
                              variable_of_interest = "R", 
                              allosome_removal = T, 
                              snp_removal = T,
                              min_maf = 0, 
                              gap_removal = T,
                              gap_threshold = 0.2,
                              outpath = file.path("output", "210607", "horaizon", "PRJ0000008_CDMEDRESP_infliximab_discovery_T2_betas_noob_nosnp_nogap"))

```
