---
title: "DM analyses response"
author: "Andrew Y.F. Li Yim"
date: "6 November 2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

In this report we will perform the differential methylation analyses on responders and non-responders towards the integrin \alpha4\beta7 agonist vedolizumab. 

```{r packages}
require(pheatmap)
require(gridExtra)
require(grid)
require(cowplot)
require(ggpubr)
require(minfi)
require(reshape2)
require(NDlib)
require(Homo.sapiens)
require(ggbio)
require(Cairo)
require(biomaRt)
require(dplyr)
require(TxDb.Hsapiens.UCSC.hg19.knownGene)
require(SNPlocs.Hsapiens.dbSNP151.GRCh38)
require(liftOver)
require(rsnps)
require(biomaRt)
require(viridis)
require(sva)

source("src/DM_analyses/dmr_plot.R")
source("src/DM_analyses/dmg_plot.R")
source("src/DM_analyses/dmr_effect_plot.R")
```

```{r setup, echo = FALSE}
#setup input directories
samplesDir <- file.path("data","samples", "02_validation")
commondataDir <- file.path("/home", "andrewliyim", "lkgres", "common_data")

#setup output directories
outputDir <- file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211113")
dir.create(outputDir)
dmpDir <- file.path(outputDir, "dmps")
dir.create(dmpDir)
dmrDir <- file.path(outputDir, "dmrs")
dir.create(dmrDir)
rdsDir <- file.path(outputDir, "rds")
dir.create(rdsDir)

#import promiscuous probes
pmprobes <- read.csv(file.path(commondataDir, "IlluminaHumanMethylationBeadChipEPIC", "Non-specific-probes-illuminaEPIC.csv"), stringsAsFactors = F, header = F)[,1]

#import enhancer regions
enhancers <- read.csv("/home/ayliyim/archive/projects/FAC0000001_REGR/output/ActivePromoterEnhancerLinks_hg19_annotated.csv", stringsAsFactors = F)[,-1]
colnames(enhancers)[1:3] <- c("baitChr", "baitStart", "baitEnd")
enhancers_gr <- makeGRangesFromDataFrame(enhancers, keep.extra.columns = T, seqnames.field = "oeChr", start.field = "oeSt", end.field = "oeEnd")
enhancers_gr <- enhancers_gr[enhancers_gr$distanceToTSS == 0,]

#biomart
enssnp_mart <- useMart(biomart = "ENSEMBL_MART_SNP", host = "grch37.ensembl.org", path = "/biomart/martservice", dataset = "hsapiens_snp")

#pheatmap setup
breaksList <- seq(0, 1, by = 0.01)

#ggplot colors
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

#Colors for response
response_cols <- c(R = "#E69F00", NR = "#56B4E9")
```

```{r plotting functions}
#Overrepresentation plotter
overrep_plotter <- function(pvals, pathway, N = 10){
  plot_df <- data.frame(pvalue = pvals,
                        pathway = factor(pathway, levels = rev(pathway)), 
                        stringsAsFactors = F)
  ggplot(plot_df[1:N,], aes(y = -log10(pvalue), x = pathway)) +
    geom_bar(stat = "identity") +
    coord_flip() +
    theme_bw() +
    theme(axis.title.y = element_blank())
}

#Methylation status across time plotter
cpg_timeplot <- function(cgid, anno, betas, response, timepoint, donor, response_cols, enlarged = F){
  #cgid: CpG identifier
  #anno: annotation object as obtained through minfi::getAnnotation()
  #betas: beta object as obtained through minfi::getBeta()
  #response: response variable per sample
  #timepoint: timeooint variable per sample
  #donor: donor variable per sample
  #response_cols: named color vector
  
  beta_df <- data.frame(Beta = betas[cgid,], Response = response, Timepoint = timepoint, Donor = donor)
  
  #Annotation
  hgnc <- paste(unique(strsplit(anno[cgid, "UCSC_RefGene_Name"], ";")[[1]]), collapse = ";")
  coords <- paste0(anno[cgid, "chr"], ":", anno[cgid, "pos"])
  
  if(hgnc != ""){
    stitle <- paste0(coords, " (", hgnc, ")")
  } else{
    stitle <- paste0(coords)
  }
  
  #Visualization
  gplot <- ggplot(beta_df, aes(x = Timepoint, y = Beta, col = Response)) +
    stat_summary(fun.data = mean_se, geom = "crossbar",  aes(color = Response)) +
    geom_point(aes(group = Donor, shape = Response), alpha = 0.25, show.legend = F, position = position_dodge(0.1)) +
    #geom_jitter(aes(group = Donor, shape = Response), width = 0.1, alpha = 0.25, show.legend = F) +
    geom_line(aes(group = Donor, alpha = 0.25), linetype = "dotted", show.legend = F, position = position_dodge(0.1)) +
    #geom_smooth(method = "lm", aes(group = Response), se = F) +
    scale_color_manual(values = response_cols) +
    xlab("Timepoint") +
    ylab("% Methylation") +
    labs(title = cgid,
         subtitle = stitle) +
    theme_bw() +
    scale_x_continuous(breaks=seq(0,3,1)) +
    theme(legend.title = element_blank(),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          legend.position = "bottom",
          text = element_text(size = 12))
  
  if(enlarged == F) gplot <- gplot + coord_cartesian(ylim = c(0, 100))
  
  return(gplot)
}
```

```{r import data}
samples <- readxl::read_excel(file.path(samplesDir, "210906_samples_vedolizumab_aumc_validation1_PRJ0000008_CDMEDRESP.xlsx"))
samples$Basename <- samples$Path
samples$resptime <- with(samples, paste0(Response, "_", Timepoint))

rgset_vedoval <- read.metharray.exp(targets = samples, force = T)
```

```{r import all data}
samples_all <- readxl::read_excel(file.path(samplesDir, "total", "samples_vedolizumab_discval.xlsx"))

samples_all$Basename <- samples_all$Filepath
samples_all$resptime <- with(samples_all, paste0(Response, "_", Timepoint))
samples_all$resptime[samples_all$Cohort == "Validation"] <- samples_all$Timepoint[samples_all$Cohort == "Validation"]

rgset_vedodiscval <- read.metharray.exp(targets = samples_all, force = T)
```

# Exploratory data analysis

```{r Quality control, echo = FALSE}
edaDir <- file.path(outputDir, "eda")
dir.create(edaDir)
```

## Blood cell distribution estimation

The next few steps will include blood cell estimation, normalization and subsequent differential methylation analysis after which we will perform enrichment analysis. The final step is to build a classifier using the raw data.

```{r Blood cell distribution}
require(FlowSorted.Blood.EPIC)
require(ExperimentHub)

if(!file.exists(file.path(edaDir, "estimatedCellProportion.csv"))){
  hub <- ExperimentHub()
  FlowSorted.Blood.EPIC <- hub[["EH1136"]]
  
  cellDist <- estimateCellCounts(rgset_vedodiscval_noF241, 
                                 cellTypes = c("CD8T", "CD4T", "NK", "Bcell", "Mono", "Neu"),
                                 compositeCellType = "Blood",
                                 processMethod = "auto", 
                                 probeSelect = "auto",
                                 referencePlatform = c("IlluminaHumanMethylationEPIC"))
  cellDist <- data.frame(cellDist, 
                         response = pData(rgset_vedodiscval_noF241)$Response,
                         timepoint = pData(rgset_vedodiscval_noF241)$Timepoint,
                         donor = pData(rgset_vedodiscval_noF241)$Biobank_ID,
                         array = rownames(cellDist))
  write.csv(cellDist, file.path(edaDir, "estimatedCellProportion.csv"))
} else{
  cellDist <- read.csv(file.path(edaDir, "estimatedCellProportion.csv"), row.names = 1, stringsAsFactors = F)
}

cellDist_melt <- reshape2::melt(cellDist, id.vars = c("response", "timepoint", "array", "donor"), value.name = "Proportion")

cellDist_plot <- ggplot(cellDist_melt, aes(x = timepoint, y = Proportion, col = response)) +
  geom_point(aes(group = donor), alpha = 0.25, show.legend = F) +
  geom_line(aes(group = donor, alpha = 0.25), linetype = "dotted", show.legend = F) +
  geom_smooth(method = "lm", aes(group = response), se = F) +
  scale_color_manual(values = response_cols) +
  facet_wrap(~variable) +
  xlab("Timepoint") +
  theme_bw() +
  #scale_x_continuous(breaks=seq(0,3,1)) +
  theme(legend.title = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.position = "bottom",
        text = element_text(size = 12))

Cairo(width = 1500, height = 1000, type = "pdf", file = file.path(edaDir, "estimatedCellProportion.pdf"), bg = "white", units = "px", dpi = 120)
print(cellDist_plot)
dev.off()
```

The estimated cell distribution does not show very large differences, suggesting that any changes do not occur in the blood-associated CpGs.

In the next few chunks, we will perform comparative differential methylation analyses. TWe will need to perform some degree of preprocessing. Usually we would remove probes binding to CpGs that lie on genetic variants, but this time we retain them as we are searching for biomarkers representative of response regardless of modality. Subsequent investigation will need to uncover the source of the signal.

```{r vedo preprocessing}
rgset_selected <- rgset_vedodiscval[,pData(rgset_vedodiscval)$Timepoint %in% c("T1", "T2")]
rgset_vedodiscval_noF241 <- rgset_selected[,pData(rgset_selected)$Biobank_ID != "F241"]

gmset_discval <- preprocessFunnorm(rgSet = rgset_vedodiscval_noF241)
#gmset_discval_processed <- gmset_discval[with(getSnpInfo(gmset_discval), which(CpG_maf < 0.01 | is.na(CpG_maf))),]
#gmset_discval_processed <- gmset_discval[with(getSnpInfo(gmset_discval), which(SBE_maf < 0.01 | is.na(SBE_maf))),]
#gmset_discval_processed <- gmset_discval[-which(minfi::getAnnotation(gmset_discval)$chr %in% c("chrX", "chrY")),]

betas_discval <- getBeta(gmset_discval)
mvals_discval <- getM(gmset_discval)

saveRDS(gmset_discval_processed, file.path(rdsDir, "gmset_discval_filtered.Rds"))
```

```{r vedo beta PCA together}
betas_discval_dm <- betas_discval - rowMeans(betas_discval)
betas_discval_svd <- svd(t(betas_discval_dm))
betas_discval_svd_expvar <- round(betas_discval_svd$d/sum(betas_discval_svd$d), 2)*100

betas_discval_svd_df <- data.frame(Sample_ID = pData(gmset_discval)$SXS_POS,
                                   Sentrix_ID = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_discval)$SXS_POS),
                                   Sentrix_Pos = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_discval)$SXS_POS),
                                   Origin = pData(gmset_discval)$Biobank_ID,
                                   Cohort = pData(gmset_discval)$Cohort,
                                   Timepoint = pData(gmset_discval)$Timepoint,
                                   Response = pData(gmset_discval)$Response,
                                   Sex = pData(gmset_discval)$predictedSex,
                                   Age = pData(gmset_discval)$Age_baseline,
                                   PC1 = betas_discval_svd$u[,1],
                                   PC2 = betas_discval_svd$u[,2])

betas_vedo_svd_plotobj <- ggplot(betas_discval_svd_df, aes(x = PC1, y = PC2)) +
  geom_point(aes(col = Cohort, shape = Response), size = 3) +
  theme_bw() +
  #facet_wrap(~Timepoint) +
  labs(title = "PCA: Functional normalization",
       subtitle = "All timepoints",
       x = paste0("PC1 (", betas_discval_svd_expvar[1], "%)"),
       y = paste0("PC2 (", betas_discval_svd_expvar[2], "%)")) +
  #scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())

Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(edaDir, "PCA.pdf"), bg = "white", units = "px", dpi = 120)
print(betas_vedo_svd_plotobj)
dev.off()
```

```{r vedo beta PCA AMC}
gmset_amc <- gmset_discval[,pData(gmset_discval)$Cohort %in% c("Discovery", "Validation")]
betas_amc <- getBeta(gmset_amc)
betas_amc_dm <- betas_amc - rowMeans(betas_amc)

predictor_cpgs <- readxl::read_excel(file.path("output", "DM_analyses", "220113_predictor_cpgs.xlsx"))
predictor_cpgs_proper <- predictor_cpgs %>% 
  dplyr::filter(Comparison == "independent discovery validation")

betas_predcpgs_amc_svd <- svd(t(betas_amc_dm[predictor_cpgs_proper$CGID,]))
betas_predcpgs_amc_svd_expvar <- round(betas_predcpgs_amc_svd$d/sum(betas_predcpgs_amc_svd$d), 2)*100

betas_predcpgs_amc_df <- data.frame(Sample_ID = pData(gmset_amc)$SXS_POS,
                                    Sentrix_ID = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_amc)$SXS_POS),
                                    Sentrix_Pos = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_amc)$SXS_POS),
                                    Origin = pData(gmset_amc)$Biobank_ID,
                                    Cohort = pData(gmset_amc)$Cohort,
                                    Disease = pData(gmset_amc)$Disease,
                                    Response = pData(gmset_amc)$Response,
                                    Sex = pData(gmset_amc)$predictedSex,
                                    Age = pData(gmset_amc)$Age_baseline,
                                    PC1 = betas_predcpgs_amc_svd$u[,1],
                                    PC2 = betas_predcpgs_amc_svd$u[,2])

betas_predcpgs_amc_svd_plotobj <- ggplot(betas_predcpgs_amc_df, aes(x = PC1, y = PC2)) +
  geom_point(aes(col = Response,shape = Cohort), size = 3) +
  theme_bw() +
  facet_wrap(~Cohort) +
  labs(title = "PCA: AMC Discovery + Validation",
       subtitle = "Predictor CpGs",
       x = paste0("PC1 (", betas_predcpgs_amc_svd_expvar[1], "%)"),
       y = paste0("PC2 (", betas_predcpgs_amc_svd_expvar[2], "%)")) +
  #scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())

Cairo(width = 3250, height = 1500, type = "pdf", file = file.path(edaDir, "PCA_AMC_cohortsplit.pdf"), bg = "white", units = "px", dpi = 120)
print(betas_predcpgs_amc_svd_plotobj)
dev.off()
```

```{r vedo beta PCA Oxford}
gmset_oxford <- gmset_discval[,pData(gmset_discval)$Cohort == "Oxford"]
betas_oxford <- getBeta(gmset_oxford)
betas_oxford_dm <- betas_oxford - rowMeans(betas_oxford)

predictor_cpgs <- readxl::read_excel(file.path("output", "DM_analyses", "220113_predictor_cpgs.xlsx"))
predictor_cpgs_proper <- predictor_cpgs %>% 
  dplyr::filter(Comparison == "independent discovery validation")

betas_predcpgs_oxford_svd <- svd(t(betas_oxford_dm[predictor_cpgs_proper$CGID,]))
betas_predcpgs_oxford_svd_expvar <- round(betas_predcpgs_oxford_svd$d/sum(betas_predcpgs_oxford_svd$d), 2)*100

betas_predcpgs_oxford_df <- data.frame(Sample_ID = pData(gmset_oxford)$SXS_POS,
                                       Sentrix_ID = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_oxford)$SXS_POS),
                                       Sentrix_Pos = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_oxford)$SXS_POS),
                                       Origin = pData(gmset_oxford)$Biobank_ID,
                                       Cohort = pData(gmset_oxford)$Cohort,
                                       Disease = pData(gmset_oxford)$Disease,
                                       Response = pData(gmset_oxford)$Response,
                                       Sex = pData(gmset_oxford)$predictedSex,
                                       Age = pData(gmset_oxford)$Age_baseline,
                                       PC1 = betas_predcpgs_oxford_svd$u[,1],
                                       PC2 = betas_predcpgs_oxford_svd$u[,2])

betas_predcpgs_oxford_svd_plotobj <- ggplot(betas_predcpgs_oxford_df, aes(x = PC1, y = PC2)) +
  geom_point(aes(col = Response), size = 3) +
  theme_bw() +
  facet_wrap(~Disease) +
  labs(title = "PCA: Oxford validation",
       subtitle = "Predictor CpGs",
       x = paste0("PC1 (", betas_predcpgs_oxford_svd_expvar[1], "%)"),
       y = paste0("PC2 (", betas_predcpgs_oxford_svd_expvar[2], "%)")) +
  #scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())

Cairo(width = 4750, height = 1500, type = "pdf", file = file.path(edaDir, "PCA_Oxford.pdf"), bg = "white", units = "px", dpi = 120)
print(betas_predcpgs_oxford_svd_plotobj)
dev.off()
```

```{r}
summary(aov(PC2~Cohort+Sentrix_Pos, data = betas_discval_svd_df))
```


There appears to be an observable difference between in the different cohorts. This is understandable, as both cohorts were run separately. We will next try to remove these batch effects using ComBat.

```{r combat normalization}
betas_discval_combat_c <- ComBat(dat = betas_discval, 
                                 batch = pData(gmset_discval)$Cohort, 
                                 mod = NULL)
betas_discval_combat_cs <- ComBat(dat = betas_discval_combat_c, 
                                  batch = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_discval)$SXS_POS), 
                                  mod = NULL)
betas_discval_combat_csp <- ComBat(dat = betas_discval_combat_cs, 
                                   batch = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_discval)$SXS_POS), 
                                   mod = NULL)

betas_discval_combat_csp_bcd <- limma::removeBatchEffect(x = betas_discval_combat_csp, 
                                                         covariates = model.matrix(~CD8T+CD4T+NK+Bcell+Mono+Neu, cellDist))



#betas_discval_combat_s <- ComBat(dat = betas_discval, 
#                                 batch = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_discval)$SXS_POS), 
#                                 mod = NULL)
#betas_discval_combat_sc <- ComBat(dat = betas_discval_combat_s, 
#                                  batch = pData(gmset_discval)$Cohort, 
#                                  mod = NULL)
```

```{r}
S203740920093_R01C01 <- data.frame(cohort_slide = betas_discval_combat_sc[,1],
                                   slide_cohort = betas_discval_combat_cs[,1])

Cairo(width = 1000, height = 1250, type = "pdf", file = file.path("203740920093_R01C01.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(S203740920093_R01C01, aes(x = cohort_slide, y = slide_cohort)) +
  geom_abline(intercept = 0, slope = 1, alpha = 0.5, linetype = "dashed") +
  geom_point_rast(alpha = 0.5) +
  labs(title = "203740920093_R01C01",
       y = "1st: Slide, 2nd: Cohort",
       x = "1st: Cohort, 2nd: Slide") +
  xlim(-0.1, 1.1) +
  ylim(-0.1, 1.1) +
  theme_bw()
dev.off()

for(i in 1:ncol(betas_discval_combat_sc)){print(summary(betas_discval_combat_sc[,i]-betas_discval_combat_cs[,i]))}
```

```{r combat cohort pca}
betas_discval_combat_c_dm <- betas_discval_combat_c - rowMeans(betas_discval_combat_c)
betas_discval_combat_c_svd <- svd(t(betas_discval_combat_c_dm))
betas_discval_combat_c_svd_expvar <- round(betas_discval_combat_c_svd$d/sum(betas_discval_combat_c_svd$d), 2)*100

betas_discval_combat_c_svd_df <- data.frame(Sample_ID = pData(gmset_discval)$SXS_POS,
                                            Sentrix_ID = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_discval)$SXS_POS),
                                            Sentrix_Pos = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_discval)$SXS_POS),
                                            Origin = pData(gmset_discval)$Biobank_ID,
                                            Cohort = pData(gmset_discval)$Cohort,
                                            Timepoint = pData(gmset_discval)$Timepoint,
                                            Response = pData(gmset_discval)$Response,
                                            Sex = pData(gmset_discval)$predictedSex,
                                            Age = pData(gmset_discval)$Age_baseline,
                                            PC1 = betas_discval_combat_c_svd$u[,1],
                                            PC2 = betas_discval_combat_c_svd$u[,2])

betas_discval_combat_c_svd_plotobj <- ggplot(betas_discval_combat_c_svd_df, aes(x = PC1, y = PC2)) +
  geom_point(aes(col = Cohort, shape = Response), size = 3) +
  theme_bw() +
  #facet_wrap(~Timepoint) +
  labs(title = "PCA: Functional normalization + ComBat (discovery/validation)",
       subtitle = "All timepoints",
       x = paste0("PC1 (", betas_discval_combat_c_svd_expvar[1], "%)"),
       y = paste0("PC2 (", betas_discval_combat_c_svd_expvar[2], "%)")) +
  #scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())
```

```{r combat cohort slide pca}
betas_discval_combat_cs_dm <- betas_discval_combat_cs - rowMeans(betas_discval_combat_cs)
betas_discval_combat_cs_svd <- svd(t(betas_discval_combat_cs_dm))
betas_discval_combat_cs_svd_expvar <- round(betas_discval_combat_cs_svd$d/sum(betas_discval_combat_cs_svd$d), 2)*100

betas_discval_combat_cs_svd_df <- data.frame(Sample_ID = pData(gmset_discval)$SXS_POS,
                                             Sentrix_ID = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_discval)$SXS_POS),
                                             Sentrix_Pos = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_discval)$SXS_POS),
                                             Origin = pData(gmset_discval)$Biobank_ID,
                                             Cohort = pData(gmset_discval)$Cohort,
                                             Timepoint = pData(gmset_discval)$Timepoint,
                                             Response = pData(gmset_discval)$Response,
                                             Sex = pData(gmset_discval)$predictedSex,
                                             Age = pData(gmset_discval)$Age_baseline,
                                             PC1 = betas_discval_combat_cs_svd$u[,1],
                                             PC2 = betas_discval_combat_cs_svd$u[,2])

betas_discval_combat_cs_svd_plotobj <- ggplot(betas_discval_combat_cs_svd_df, aes(x = PC1, y = PC2)) +
  geom_point(aes(col = Cohort, shape = Response), size = 3) +
  theme_bw() +
  #facet_wrap(~Timepoint) +
  labs(title = "PCA: Functional normalization + ComBat (discovery/validation, slide)",
       subtitle = "All timepoints",
       x = paste0("PC1 (", betas_discval_combat_cs_svd_expvar[1], "%)"),
       y = paste0("PC2 (", betas_discval_combat_cs_svd_expvar[2], "%)")) +
  #scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())
```

```{r combat cohort slide pos pca}
betas_discval_combat_csp_dm <- betas_discval_combat_csp - rowMeans(betas_discval_combat_csp)
betas_discval_combat_csp_svd <- svd(t(betas_discval_combat_csp_dm))
betas_discval_combat_csp_svd_expvar <- round(betas_discval_combat_csp_svd$d/sum(betas_discval_combat_csp_svd$d), 2)*100

betas_discval_combat_csp_svd_df <- data.frame(Sample_ID = pData(gmset_discval)$SXS_POS,
                                              Sentrix_ID = gsub("([0-9]+)_R[0-9]{2}C[0-9]{2}", "\\1", pData(gmset_discval)$SXS_POS),
                                              Sentrix_Pos = gsub("[0-9]+_(R[0-9]{2}C[0-9]{2})", "\\1", pData(gmset_discval)$SXS_POS),
                                              Origin = pData(gmset_discval)$Biobank_ID,
                                              Cohort = pData(gmset_discval)$Cohort,
                                              Timepoint = pData(gmset_discval)$Timepoint,
                                              Response = pData(gmset_discval)$Response,
                                              Sex = pData(gmset_discval)$predictedSex,
                                              Age = pData(gmset_discval)$Age_baseline,
                                              PC1 = betas_discval_combat_csp_svd$u[,1],
                                              PC2 = betas_discval_combat_csp_svd$u[,2])

betas_discval_combat_csp_svd_plotobj <- ggplot(betas_discval_combat_csp_svd_df, aes(x = PC1, y = PC2)) +
  geom_point(aes(col = Cohort, shape = Response), size = 3) +
  theme_bw() +
  #facet_wrap(~Timepoint) +
  labs(title = "PCA: Functional normalization + ComBat (discovery/validation, slide, position)",
       subtitle = "All timepoints",
       x = paste0("PC1 (", betas_discval_combat_csp_svd_expvar[1], "%)"),
       y = paste0("PC2 (", betas_discval_combat_csp_svd_expvar[2], "%)")) +
  #scale_shape_manual(values = c(T1 = 21, T2 = 22, T3 = 24)) +
  theme(text = element_text(size = 12),
        legend.position = "bottom",
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.title = element_blank())
```

```{r}
Cairo(width = 3000, height = 3250, type = "pdf", file = file.path("pca.pdf"), bg = "white", units = "px", dpi = 120)
ggarrange(betas_vedo_svd_plotobj, betas_discval_combat_c_svd_plotobj, betas_discval_combat_cs_svd_plotobj, betas_discval_combat_csp_svd_plotobj, 
          nrow = 2, 
          ncol = 2, 
          align = "hv", 
          legend = "bottom", 
          common.legend = T)
dev.off()
```

## Annotation data

Even though Illumina has provided a plethora of metadata, some is outdated. Two of the more pressing issues are the annotated genetic variants and the genome build. Here, we aim to annotate the metadata with more recent metadata on genetic variants. 

```{r prepare annotation data}
gmset_vedo_anno <- minfi::getAnnotation(gmset_vedo_filtered)
gmset_vedo_anno_gr <- makeGRangesFromDataFrame(gmset_vedo_anno, keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
```

As the `SNPlocs.Hsapiens.dbSNP151.GRCh38` is annotated in GRCh38, the first step would be to uplift the data to GRCh38.

```{r uplift to grch38}
require(liftOver)

hg19tohg38 <- import.chain("/home/ayliyim/archive/common_data/liftover_chains/hg19ToHg38.over.chain")

gmset_vedo_anno_grch38_coords <- unlist(liftOver(gmset_vedo_anno_gr, hg19tohg38))
mcols(gmset_vedo_anno_grch38_coords) <- data.frame(Name = gmset_vedo_anno_grch38_coords$Name)
```

```{r find gvs}
require(SNPlocs.Hsapiens.dbSNP151.GRCh38)
require(biomaRt)

enssnp_mart <- useDataset(mart = useMart("ENSEMBL_MART_SNP"), dataset = "hsapiens_snp")

seqlevelsStyle(gmset_vedo_anno_grch38_coords) <- "NCBI"

gmset_vedo_anno_grch38_coords_split <- split(gmset_vedo_anno_grch38_coords, ceiling(seq_along(gmset_vedo_anno_grch38_coords)/1000))

gmset_vedo_anno_grch38_snps <- lapply(gmset_vedo_anno_grch38_coords_split, function(splitfold){
  
  #Find SNPs
  snpol <- snpsByOverlaps(x = SNPlocs.Hsapiens.dbSNP151.GRCh38, ranges = splitfold)
  snpcpgol <- findOverlaps(splitfold, snpol)
  snpol_filtered <- snpol[subjectHits(snpcpgol),]
  snpol_filtered$CpG <- as.character(splitfold[queryHits(snpcpgol),]$Name)
  
  #Attach MAFs
  snp_maf <- getBM(attributes = c("refsnp_id", "allele", "minor_allele_freq"), 
                   filters = "snp_filter", 
                   values = snpol_filtered$RefSNP_id, 
                   mart = enssnp_mart)
  snp_maf <- snp_maf[!is.na(snp_maf$minor_allele_freq),]
  
  #Merge
  snpol_maf <- snpol_filtered[match(snp_maf$refsnp_id, snpol_filtered$RefSNP_id),]
  snpol_maf$Alleles <- snpol_maf$alleles_as_ambig
  snpol_maf$MAF <- snp_maf$minor_allele_freq
  
  return(GRanges(snpol_maf))
})
gc()

gmset_vedo_anno_grch38_snps <- GRangesList(gmset_vedo_anno_grch38_snps)
gmset_vedo_anno_grch38_snps <- unlist(gmset_vedo_anno_grch38_snps)
data.frame(gmset_vedo_anno_grch38_snps)
```

```{r merge gvs with grch38}
gmset_vedo_anno_grch38_coords_snps <- merge(as.data.frame(gmset_vedo_anno_grch38_coords), mcols(gmset_vedo_anno_grch38_snps), by.x = "Name", by.y = "CpG", all.x = T)
```

```{r merge gvs with grch38}
require(dplyr)

gmset_vedo_anno_grch38 <- merge(gmset_vedo_anno, gmset_vedo_anno_grch38_coords_snps, by = "Name", all.x = T)
gmset_vedo_anno_grch38 <- gmset_vedo_anno_grch38 %>%
  data.frame() %>%
  rename(chr_grch37 = chr,
         pos_grch37 = pos,
         chr_grch38 = seqnames,
         pos_grch38 = start) %>%
  select(chr_grch37, 
         pos_grch37, 
         chr_grch38, 
         pos_grch38, 
         everything(),
         -end,
         -width,
         -strand.y,
         -alleles_as_ambig)

gmset_vedo_anno_grch38_nona <- gmset_vedo_anno_grch38[!is.na(gmset_vedo_anno_grch38$chr_grch38),]
gmset_vedo_anno_grch38_nona_gr <- makeGRangesFromDataFrame(df = gmset_vedo_anno_grch38_nona, keep.extra.columns = T, seqnames.field = "chr_grch38", start.field = "pos_grch38", end.field = "pos_grch38")

saveRDS(gmset_vedo_anno_grch38_nona, file.path(rdsDir, "gmset_vedo_anno_grch38_annotated.Rds"))
gmset_vedo_anno_grch38_nona <- readRDS(file.path(rdsDir, "gmset_vedo_anno_grch38_annotated.Rds"))
```

```{r filter out the SNPs with MAF > 0.01}
snp_cpgs <- gmset_vedo_anno_grch38_nona$Name[which(gmset_vedo_anno_grch38_nona$MAF>0.01)]

gmset_vedo_filtered_nosnp <- gmset_vedo_filtered[-which(rownames(gmset_vedo_filtered) %in% snp_cpgs),]
```

```{r gaphunting}
gmset_vedo_filtered_nosnp_gh <- gaphunter(object = gmset_vedo_filtered_nosnp, threshold = 0.2)
saveRDS(gmset_vedo_filtered_nosnp_gh, file.path(rdsDir, "gmset_vedo_filtered_nosnp_gh.Rds"))

stripchart(getBeta(gmset_vedo_filtered_nosnp)["cg09845941",], xlim = c(0,1))

gmset_vedo_filtered_nosnp_nogap <- gmset_vedo_filtered_nosnp[!rownames(gmset_vedo_filtered_nosnp) %in% rownames(gmset_vedo_filtered_nosnp_gh$proberesults),]
```

# Differential methylation

```{r vedo DM annotation setup}
gmset_vedo_anno_enh_ol <- findOverlaps(query = gmset_vedo_anno_gr, subject = enhancers_gr)
gmset_vedo_anno_enh_ol <- data.frame(gmset_vedo_anno_enh_ol, enh_gene = enhancers_gr[subjectHits(gmset_vedo_anno_enh_ol),]$SYMBOL)
gmset_vedo_anno_enh_ol_list <- split(gmset_vedo_anno_enh_ol, gmset_vedo_anno_enh_ol$queryHits)
gmset_vedo_anno_enh_genes <- do.call(rbind, lapply(gmset_vedo_anno_enh_ol_list, function(cpg){
  paste(cpg$enh_gene, collapse = ";")
}))

gmset_vedo_anno_gr$enhancer_gene <- ""
gmset_vedo_anno_gr$enhancer_gene[as.numeric(rownames(gmset_vedo_anno_enh_genes))] <- gmset_vedo_anno_enh_genes
```

```{r vedo DM analyses preparation T1 and T2}
require(limma)

gmset_vedo_filtered_t1t2 <- gmset_vedo_filtered[,pData(gmset_vedo_filtered)$Timepoint %in% c(1,2)]
pData(gmset_vedo_filtered_t1t2)$Smoker <- "No"
pData(gmset_vedo_filtered_t1t2)$Smoker[pData(gmset_vedo_filtered_t1t2)$Smoking_status_baseline %in% c("Current smoker")] <- "Yes"

design_vedo <- model.matrix(~0 + resptime + Sex + Age_baseline + Smoker, data = pData(gmset_vedo_filtered_t1t2))
colnames(design_vedo)[1:7] <- c("T1_NR", "T2_NR", "T1_R", "T2_R", "Male", "Age", "Smoker")

# mvals_lmfit_vedo_dupcor <- duplicateCorrelation(getM(gmset_vedo_filtered_t1t2), design = design_vedo, block = pData(gmset_vedo_filtered_t1t2)$Biobank_ID)
# mvals_lmfit_vedo <- lmFit(getM(gmset_vedo_filtered_t1t2), design = design_vedo, block = pData(gmset_vedo_filtered_t1t2)$Biobank_ID, correlation = mvals_lmfit_vedo_dupcor$consensus.correlation)
mvals_lmfit_vedo <- lmFit(getM(gmset_vedo_filtered_t1t2), design = design_vedo)

# betas_lmfit_vedo_dupcor <- duplicateCorrelation(getBeta(gmset_vedo_filtered_t1t2), design = design_vedo, block = pData(gmset_vedo_filtered_t1t2)$Biobank_ID)
# betas_lmfit_vedo <- lmFit(getBeta(gmset_vedo_filtered_t1t2), design = design_vedo, block = pData(gmset_vedo_filtered_t1t2)$Biobank_ID, correlation = betas_lmfit_vedo_dupcor$consensus.correlation)
betas_lmfit_vedo <- lmFit(getBeta(gmset_vedo_filtered_t1t2), design = design_vedo)

contrast_mat_vedo <- makeContrasts(
  RvNRT1 = T1_R-T1_NR,
  RvNRT2 = T2_R-T2_NR,
  RvNR = (T1_R+T2_R)/2-(T1_NR+T2_NR)/2,
  T2vT1R = T2_R-T1_R,
  T2vT1NR = T2_NR-T1_NR,
  T2vT1 = (T2_R+T2_NR)/2-(T1_R+T1_NR)/2,
  RvNRvT2vT1 = (T2_R-T1_R)-(T2_NR-T1_NR),
  levels = design_vedo
)

mvals_contrastfit_vedo <- contrasts.fit(mvals_lmfit_vedo, contrast_mat_vedo)
mvals_ebayes_vedo <- eBayes(mvals_contrastfit_vedo)

betas_contrastfit_vedo <- contrasts.fit(betas_lmfit_vedo, contrast_mat_vedo)
betas_ebayes_vedo <- eBayes(betas_contrastfit_vedo)
```

## T2vT1

### T2vT1 R

```{r vedo DM analyses T2vT1 for R}
T2vT1RDir <- file.path(dmpDir, "T2vT1RDir")
dir.create(T2vT1RDir)

T2vT1R_vedo_top <- topTable(mvals_ebayes_vedo, coef = "T2vT1R", number = "Inf", adjust.method = "BH")
T2vT1R_vedo_top <- cbind(T2vT1R_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(T2vT1R_vedo_top), "T2vT1R"], data.frame(gmset_vedo_anno_gr[rownames(T2vT1R_vedo_top), ]))
write.csv(T2vT1R_vedo_top, file.path(T2vT1RDir, "T2vT1R_vedo.csv"))

T2vT1R_vedo_sig <- T2vT1R_vedo_top[T2vT1R_vedo_top$adj.P.Val < 0.05,]
#write.csv(T2vT1R_vedo_sig, file.path(T2vT1RDir, "T2vT1R_vedo_sig.csv"))
```

### T2vT1 NR

```{r vedo DM analyses T2vT1 for NR}
T2vT1NRDir <- file.path(dmpDir, "T2vT1NRDir")
dir.create(T2vT1NRDir)

T2vT1NR_vedo_top <- topTable(mvals_ebayes_vedo, coef = "T2vT1NR", number = "Inf", adjust.method = "BH")
T2vT1NR_vedo_top <- cbind(T2vT1NR_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(T2vT1NR_vedo_top), "T2vT1NR"], data.frame(gmset_vedo_anno_gr[rownames(T2vT1NR_vedo_top), ]))
write.csv(T2vT1NRDir, file.path(T2vT1NRDir, "T2vT1NR_vedo.csv"))

T2vT1NR_vedo_sig <- T2vT1NR_vedo_top[T2vT1NR_vedo_top$adj.P.Val < 0.05,]
#write.csv(T2vT1NR_vedo_sig, file.path(T2vT1RDir, "T2vT1NR_vedo_sig.csv"))
```

### T2vT1 global

```{r vedo DM analyses T2vT1 for R and NR}
T2vT1Dir <- file.path(dmpDir, "T2vT1Dir")
dir.create(T2vT1Dir)

T2vT1_vedo_top <- topTable(mvals_ebayes_vedo, coef = "T2vT1", number = "Inf", adjust.method = "BH")
T2vT1_vedo_top <- cbind(T2vT1_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(T2vT1_vedo_top), "T2vT1"], data.frame(gmset_vedo_anno_gr[rownames(T2vT1_vedo_top), ]))
write.csv(T2vT1_vedo_top, file.path(T2vT1Dir, "T2vT1_vedo.csv"))

T2vT1_vedo_sig <- T2vT1_vedo_top[T2vT1_vedo_top$adj.P.Val < 0.05,]
#write.csv(T2vT1_vedo_sig, file.path(T2vT1Dir, "T2vT1_vedo_sig.csv"))

cpg_strip_plot(cpg_num = rownames(T2vT1_vedo_top)[1], betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$Response, type = "SE", enlarged = F)
```

## RvNR

### RvNR T1

```{r vedo DM analyses RvNR for T1}
RvNRT1Dir <- file.path(dmpDir, "RvNRT1Dir")
dir.create(RvNRT1Dir)

RvNRT1_vedo_top <- topTable(mvals_ebayes_vedo, coef = "RvNRT1", number = "Inf", adjust.method = "BH")
RvNRT1_vedo_top <- cbind(RvNRT1_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(RvNRT1_vedo_top), "RvNRT1"], data.frame(gmset_vedo_anno_gr[rownames(RvNRT1_vedo_top), ]))
write.csv(RvNRT1_vedo_top, file.path(RvNRT1Dir, "RvNRT1_vedo.csv"))

RvNRT1_vedo_sig <- RvNRT1_vedo_top[RvNRT1_vedo_top$adj.P.Val < 0.05,]
#write.csv(RvNRT1_vedo_sig, file.path(RvNRT1Dir, "RvNRT1_vedo_sig.csv"))

cpg_strip_plot(cpg_num = rownames(RvNRT1_vedo_top)[1], betas = getBeta(gmset_vedo), factor_interest = pData(gmset_vedo)$resptime, type = "SE")
```

### RvNR T2

```{r vedo DM analyses RvNR for T2}
RvNRT2Dir <- file.path(dmpDir, "RvNRT2Dir")
dir.create(RvNRT2Dir)

RvNRT2_vedo_top <- topTable(mvals_ebayes_vedo, coef = "RvNRT2", number = "Inf", adjust.method = "BH")
RvNRT2_vedo_top <- cbind(RvNRT2_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(RvNRT2_vedo_top), "RvNRT2"], data.frame(gmset_vedo_anno_gr[rownames(RvNRT2_vedo_top), ]))
write.csv(RvNRT2_vedo_top, file.path(RvNRT2Dir, "RvNRT2_vedo.csv"))

RvNRT2_vedo_sig <- RvNRT2_vedo_top[RvNRT2_vedo_top$adj.P.Val < 0.05,]
#write.csv(RvNRT2_vedo_sig, file.path(RvNRT2Dir, "RvNRT1_vedo_sig.csv"))
```

### RvNR T1 and T2

```{r vedo DM analyses RvNR for T1 and T2}
RvNRDir <- file.path(dmpDir, "RvNRDir")
dir.create(RvNRDir)

RvNR_vedo_top <- topTable(mvals_ebayes_vedo, coef = "RvNR", number = "Inf", adjust.method = "BH")
RvNR_vedo_top <- cbind(RvNR_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(RvNR_vedo_top), "RvNR"], data.frame(gmset_vedo_anno_gr[rownames(RvNR_vedo_top), ]))
write.csv(RvNR_vedo_top, file.path(RvNRDir, "RvNR_vedo.csv"))

RvNR_vedo_sig <- RvNR_vedo_top[RvNR_vedo_top$adj.P.Val < 0.05,]
write.csv(RvNR_vedo_sig, file.path(RvNRDir, "RvNR_vedo_sig.csv"))

RvNR_vedo_top_gr <- makeGRangesFromDataFrame(RvNR_vedo_top, keep.extra.columns = T)
```

##### Visualization

```{r cross bar plots}
cbRvNRDir <- file.path(RvNRDir, "crossbarplots")
dir.create(cbRvNRDir)

for(i in 1:nrow(RvNR_vedo_sig)){
  Cairo(width = 800, height = 800, type = "pdf", file = file.path(cbRvNRDir, paste0(i, "_", rownames(RvNR_vedo_sig)[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(NDlib::cpg_strip_plot(cpg_num = rownames(RvNR_vedo_sig)[i], 
                              betas = getBeta(gmset_vedo_filtered_t1t2), 
                              factor_interest = pData(gmset_vedo_filtered_t1t2)$resptime, 
                              type = "SE") + 
          labs(title = rownames(RvNR_vedo_sig)[i],
               subtitle = paste0(RvNR_vedo_sig$seqnames[i], ":", RvNR_vedo_sig$start[i], " (", RvNR_vedo_sig$UCSC_RefGene_Name, ")")) +
          theme(panel.grid.major = element_blank(), 
                panel.grid.minor = element_blank()))
  dev.off() 
}
```

```{r time crossbar plot}
tcbRvNRDir <- file.path(RvNRDir, "timecrossbarplots")
dir.create(tcbRvNRDir)

for(i in 1:nrow(RvNR_vedo_sig)){
  Cairo(width = 800, height = 800, type = "pdf", file = file.path(tcbRvNRDir, paste0(i, "_", rownames(RvNR_vedo_sig)[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_timeplot(cgid = rownames(RvNR_vedo_sig)[i], 
                     anno = gmset_vedo_anno,
                     betas = getBeta(gmset_vedo_filtered), 
                     response = pData(gmset_vedo_filtered)$Response,
                     timepoint = pData(gmset_vedo_filtered)$Timepoint,
                     donor = pData(gmset_vedo_filtered)$Biobank_ID,
                     response_cols = response_cols))
  dev.off() 
}
```

#### Time stability

```{r T2vT1 RvNR}
T2vT1_vedo_top[rownames(RvNR_vedo_sig),]
```

```{r vedo DM analyses preparation all timepoints}
require(limma)

pData(gmset_vedo_filtered)$Smoker <- "No"
pData(gmset_vedo_filtered)$Smoker[pData(gmset_vedo_filtered)$Smoking_status_baseline %in% c("Current smoker")] <- "Yes"

design_vedo_all <- model.matrix(~0 + resptime + Sex + Age_baseline + Smoker, data = pData(gmset_vedo_filtered))
colnames(design_vedo_all) <- c("T1_NR", "T2_NR", "T3_NR", "T1_R", "T2_R", "T3_R", "Male", "Age", "Smoker")

mvals_lmfit_vedo_all <- lmFit(getM(gmset_vedo_filtered), design = design_vedo_all)
betas_lmfit_vedo_all <- lmFit(getBeta(gmset_vedo_filtered), design = design_vedo_all)

contrast_mat_vedo_all <- makeContrasts(
  T2vT1 = (T2_R+T2_NR)/2-(T1_R+T1_NR)/2,
  T3vT2 = (T3_R+T3_NR)/2-(T2_R+T2_NR)/2,
  levels = design_vedo_all
)

mvals_contrastfit_vedo_all <- contrasts.fit(mvals_lmfit_vedo_all, contrast_mat_vedo_all)
mvals_ebayes_vedo_all <- eBayes(mvals_contrastfit_vedo_all)

betas_contrastfit_vedo_all <- contrasts.fit(betas_lmfit_vedo_all, contrast_mat_vedo_all)
betas_ebayes_vedo_all <- eBayes(betas_contrastfit_vedo_all)
```

```{r vedo time stability analyses}
T2vT1_vedo_top_all <- topTable(mvals_ebayes_vedo_all, coef = "T2vT1", number = "Inf", adjust.method = "BH")
RvNR_T2vT1_top_all <- T2vT1_vedo_top_all[rownames(RvNR_vedo_sig),]
RvNR_T2vT1_top_all <- cbind(RvNR_T2vT1_top_all, 
                            Beta = coefficients(betas_ebayes_vedo_all)[rownames(RvNR_T2vT1_top_all), "T2vT1"],
                            Beta_SE = sqrt(betas_ebayes_vedo_all$s2.post[rownames(RvNR_T2vT1_top_all)]) * betas_ebayes_vedo_all$stdev.unscaled[rownames(RvNR_T2vT1_top_all),"T2vT1"])

T3vT2_vedo_top_all <- topTable(mvals_ebayes_vedo_all, coef = "T3vT2", number = "Inf", adjust.method = "BH")
RvNR_T3vT2_top_all <- T3vT2_vedo_top_all[rownames(RvNR_vedo_sig),]
RvNR_T3vT2_top_all <- cbind(RvNR_T3vT2_top_all, Beta = coefficients(betas_ebayes_vedo_all)[rownames(RvNR_T3vT2_top_all), "T3vT2"],
                            Beta_SE = sqrt(betas_ebayes_vedo_all$s2.post[rownames(RvNR_T3vT2_top_all)]) * betas_ebayes_vedo_all$stdev.unscaled[rownames(RvNR_T3vT2_top_all),"T3vT2"])

RvNR_T3vT2vT1_top_all <- data.frame(RvNR_T2vT1_top_all[,c("logFC", "t", "P.Value", "Beta", "Beta_SE")], RvNR_T3vT2_top_all[,c("logFC", "t", "P.Value", "Beta", "Beta_SE")])

colnames(RvNR_T3vT2vT1_top_all) <- c("Mdiff_T2vT1", "t_T2vT1", "pvalue_T2vT1", "Betadiff_T2vT1", "Betadiff_SE_T2vT1", "Mdiff_T3vT2", "t_T3vT2", "pvalue_T3vT2", "Betadiff_T3vT2", "Betadiff_SE_T3vT2")

write.csv(RvNR_T3vT2vT1_top_all, file.path(RvNRDir, "RvNR_vedo_sig_T1T2T3.csv"))
```

```{r}
RvNR_T3vT2vT1_top_all %>%
  ggplot(aes(x = Betadiff_T2vT1, y = Betadiff_T3vT2)) +
  geom_vline(xintercept = 0, col = "black") +
  geom_hline(yintercept = 0, col = "black") +
  geom_errorbar(aes(ymin = Betadiff_T3vT2 - Betadiff_SE_T3vT2, 
                    ymax = Betadiff_T3vT2 + Betadiff_SE_T3vT2), col = "grey") +
  geom_errorbarh(aes(xmin = Betadiff_T2vT1 - Betadiff_SE_T2vT1, 
                     xmax = Betadiff_T2vT1 + Betadiff_SE_T2vT1), col = "grey") +
  geom_point() +
  theme_bw() +
  labs(title = "Response-associated DMPs over time",
       subtitle = "Mean difference in methylation Â± SE") +
  xlab("T2 vs T1") +
  ylab("T3 vs T2") +
  coord_cartesian(xlim = c(-0.15,0.15),
                  ylim = c(-0.15,0.15)) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

```{r equivalence testing}
sesoi <- min(abs(RvNR_vedo_sig$logFC))
mdiff_t <- mvals_ebayes_vedo_t1t2t3$coefficients[, "Timepoint"]
mdiff_t1 <- (mdiff_t - sesoi)/mvals_ebayes_vedo_t1t2t3$stdev.unscaled[,"Timepoint"]/sqrt(mvals_ebayes_vedo_t1t2t3$s2.post)
mdiff_tp1 <- pt(mdiff_t1, df=mvals_ebayes_vedo_t1t2t3$df.total, lower.tail=TRUE) # One-sided p-value.

mdiff_t2 <- (mdiff_t + sesoi)/mvals_ebayes_vedo_t1t2t3$stdev.unscaled[,"Timepoint"]/sqrt(mvals_ebayes_vedo_t1t2t3$s2.post)
mdiff_tp2 <- pt(mdiff_t2, df=mvals_ebayes_vedo_t1t2t3$df.total, lower.tail=FALSE)
mval_t_tost_pval <- pmax(mdiff_tp1, mdiff_tp2)

mval_t_tost_pval[rownames(RvNR_vedo_sig)]
```


#### KEGG enrichment analysis

For enrichment analyses we will need to aggregate the methylation signal per gene. For this reason I will utilize the promoters as they tend to have a better interpretability relative to enhancers.

```{r vedo dmprom kegg enrichment}
require(ChIPpeakAnno)
require(dplyr)
require(fgsea)

RvNR_gene_vedo_top_gr <- annotatePeakInBatch(myPeakList = makeGRangesFromDataFrame(RvNR_vedo_top, keep.extra.columns = T), 
                                             AnnotationData = unlist(GenomicRanges::reduce(transcriptsBy(x = TxDb.Hsapiens.UCSC.hg19.knownGene, by = "gene"))), 
                                             output="nearestBiDirectionalPromoters",
                                             bindingRegion=c(-5000, 1000))

RvNR_gene_vedo_top_agg <- RvNR_gene_vedo_top_gr %>% 
  data.frame() %>%
  dplyr::group_by(feature) %>%
  dplyr::summarize(MeanBeta = mean(Beta, na.rm=TRUE))

RvNR_gene_vedo_top_bd <- RvNR_gene_vedo_top_agg$MeanBeta
names(RvNR_gene_vedo_top_bd) <- RvNR_gene_vedo_top_agg$feature

#For visualization
RvNR_gene_vedo_top_bd_df <- data.frame(Entrez = names(RvNR_gene_vedo_top_bd),
                                       kegg = paste0("hsa:", names(RvNR_gene_vedo_top_bd)),
                                       tstat = RvNR_gene_vedo_top_bd)
write.csv(RvNR_gene_vedo_top_bd_df, file.path(RvNRDir, "RvNR_gene_vedo_top_bd.csv"))

kegg_hs <- EnrichmentBrowser::getGenesets("hsa", db = "kegg")
RvNR_gene_vedo_obj <- fgsea(pathways = kegg_hs, 
                            stats = RvNR_gene_vedo_top_bd,
                            minSize = 15,
                            maxSize = 500)
RvNR_gene_vedo_obj <- RvNR_gene_vedo_obj[order(RvNR_gene_vedo_obj$pval),]

RvNR_gene_vedo_obj$ID <- gsub("^[A-Z0-9_]+_MM_(.+)_Ensembl$", "\\1", RvNR_gene_vedo_obj$pathway)
kegg_RvNR_gene_vedo <- data.frame(RvNR_gene_vedo_obj)[-8]

write.csv(kegg_RvNR_gene_vedo, file.path(RvNRDir, "kegg_RvNR_gene_vedo.csv"))
```

### GV

```{r vedo DM analyses RvNR for T1 and T2 gaps}
rvnrgvDir <- file.path(RvNRDir, "genetic_variants")
dir.create(rvnrgvDir)

gmset_vedo_filtered_gh <- gaphunter(object = gmset_vedo_filtered)

RvNR_vedo_sig_gapped <- intersect(rownames(gmset_vedo_filtered_gh$proberesults), rownames(RvNR_vedo_sig))

gmset_vedo_filtered_gh_clusters <- gmset_vedo_filtered_gh$proberesults[RvNR_vedo_sig_gapped,]

RvNR_vedo_sig_gapped_anno_gr <- makeGRangesFromDataFrame(gmset_vedo_anno[RvNR_vedo_sig_gapped,], keep.extra.columns = F, start.field = "pos", end.field = "pos")
```

Identify whether any known SNPs are located within 51 bps of the gapped CpGs (probe length is 50 bps).

```{r vedo DM analyses RvNR for T1 and T2 catalogued variants}
hg19tohg38 <- import.chain("/home/ayliyim/archive/common_data/liftover_chains/hg19ToHg38.over.chain")

RvNR_vedo_sig_gapped_anno_gr_grch38 <- unlist(liftOver(RvNR_vedo_sig_gapped_anno_gr, hg19tohg38))
RvNR_vedo_sig_gapped_anno_gr_grch38 <- RvNR_vedo_sig_gapped_anno_gr_grch38+51
seqlevelsStyle(RvNR_vedo_sig_gapped_anno_gr_grch38) <- seqlevelsStyle(SNPlocs.Hsapiens.dbSNP151.GRCh38)

RvNR_vedo_sig_gapped_snps <- snpsByOverlaps(x = SNPlocs.Hsapiens.dbSNP151.GRCh38, 
                                            ranges = RvNR_vedo_sig_gapped_anno_gr_grch38)
saveRDS(RvNR_vedo_sig_gapped_snps, file.path(rvnrgvDir, "RvNR_vedo_sig_gapped_snps.Rds"))

#MAF
enssnp_mart <- useDataset(mart = useMart("ENSEMBL_MART_SNP"),dataset="hsapiens_snp")

RvNR_vedo_sig_gapped_snps_df <- getBM(attributes = c("refsnp_id", "allele", "minor_allele_freq"), filters = "snp_filter", values = RvNR_vedo_sig_gapped_snps$RefSNP_id, mart = enssnp_mart)
RvNR_vedo_sig_gapped_snps_df <- RvNR_vedo_sig_gapped_snps_df[!is.na(RvNR_vedo_sig_gapped_snps_df$minor_allele_freq),]

RvNR_vedo_sig_gapped_snps <- makeGRangesFromDataFrame(merge(data.frame(RvNR_vedo_sig_gapped_snps), 
                                                            RvNR_vedo_sig_gapped_snps_df, 
                                                            by.x = "RefSNP_id", 
                                                            by.y = "refsnp_id"), 
                                                      keep.extra.columns = T, 
                                                      start.field = "pos", 
                                                      end.field = "pos")

RvNR_vedo_sig_gapped_anno_snps_ol <- findOverlaps(query = RvNR_vedo_sig_gapped_snps, subject = RvNR_vedo_sig_gapped_anno_gr_grch38)

RvNR_vedo_sig_gapped_snps_ol <- RvNR_vedo_sig_gapped_snps[queryHits(RvNR_vedo_sig_gapped_anno_snps_ol)]
RvNR_vedo_sig_gapped_snps_ol$CpG <- names(RvNR_vedo_sig_gapped_anno_gr_grch38)[subjectHits(RvNR_vedo_sig_gapped_anno_snps_ol)]
RvNR_vedo_sig_gapped_snps_ol$CpG_pos <- start(RvNR_vedo_sig_gapped_anno_gr_grch38-51)[subjectHits(RvNR_vedo_sig_gapped_anno_snps_ol)]
RvNR_vedo_sig_gapped_snps_ol$GV_relpos <- RvNR_vedo_sig_gapped_snps_ol$CpG_pos-start(RvNR_vedo_sig_gapped_snps_ol)
RvNR_vedo_sig_gapped_snps_ol$GV_type <- "probe GV"
RvNR_vedo_sig_gapped_snps_ol$GV_type[RvNR_vedo_sig_gapped_snps_ol$GV_relpos %in% c(-1, 0, 1)] <- "CpG GV"
RvNR_vedo_sig_gapped_snps_ol$Gene <- RvNR_vedo_sig$UCSC_RefGene_Name[match(RvNR_vedo_sig_gapped_snps_ol$CpG, rownames(RvNR_vedo_sig))]
RvNR_vedo_sig_gapped_snps_ol$Pvalue <- RvNR_vedo_sig$P.Value[match(RvNR_vedo_sig_gapped_snps_ol$CpG, rownames(RvNR_vedo_sig))]
RvNR_vedo_sig_gapped_snps_ol$Beta <- RvNR_vedo_sig$Beta[match(RvNR_vedo_sig_gapped_snps_ol$CpG, rownames(RvNR_vedo_sig))]

RvNR_vedo_sig_gapped_snps_df <- data.frame(RvNR_vedo_sig_gapped_snps_ol)
RvNR_vedo_sig_gapped_snps_df <- RvNR_vedo_sig_gapped_snps_df[, c("CpG", "seqnames", "CpG_pos", "Beta", "Pvalue", "RefSNP_id", "start", "GV_relpos", "allele", "GV_type", "minor_allele_freq", "Gene")]
colnames(RvNR_vedo_sig_gapped_snps_df) <- c("CpG", "Chromosome", "CpG_position", "Beta", "Pvalue", "RSID", "GV_position", "GV_CpG_relpos", "Alleles", "GV_type", "MAF_all_1000g", "Gene")

write.csv(RvNR_vedo_sig_gapped_snps_df, file = file.path(rvnrgvDir, "RvNR_vedo_sig_gapped_snps_annotated.csv"))

#up
RvNR_clustered_barplot <- data.frame(Type = c("non-clustered", "clustered"),
                                     DMPs = c(nrow(RvNR_vedo_sig)-nrow(gmset_vedo_filtered_gh_clusters), nrow(gmset_vedo_filtered_gh_clusters)),
                                     Group = "DMPs") %>%
  ggplot(aes(x = Group, y = DMPs, fill = Type)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = DMPs), position = position_stack(vjust = 0.5)) +
  #scale_fill_viridis(discrete = T) +
  coord_flip() +
  labs(title = "Putative GVs",
       subtitle = "DMPs presenting clustered pattern") +
  theme_bw() +
  theme(legend.position = "bottom", 
        axis.title.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

#left
RvNR_clustered_typegv_df <- data.frame(table(RvNR_vedo_sig_gapped_snps_df$GV_type)) %>%
  dplyr::rename("Location" = "Var1", 
                "GVs" = "Freq") %>%
  dplyr::mutate(Location = gsub(" GV", "", RvNR_clustered_typegv_df$Var1))

RvNR_clustered_typegv_barplot <- ggplot(RvNR_clustered_typegv_df, aes(x = Location, y = GVs)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = GVs), position = position_stack(vjust = 0.5), col = "white") +
  ggtitle("Location of GVs") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

RvNR_vedo_gv_plot_left <- ggarrange(RvNR_clustered_barplot, RvNR_clustered_typegv_barplot, nrow = 2, align = "v", heights = c(0.5, 1))

#right
RvNR_clustered_gv_maf_xyplot <- RvNR_vedo_sig_gapped_snps_df %>%
  ggplot(aes(x = GV_CpG_relpos, y = MAF_all_1000g, col = MAF_all_1000g, alpha = -log(abs(GV_CpG_relpos+0.001)))) +
  #geom_line(aes(group = CpG)) +
  geom_point() +
  geom_rect(data = data.frame(x0 = -1.5, x1 = 1.5),
            inherit.aes = FALSE,
            aes(xmin = x0, xmax = x1, ymin = -Inf, ymax = Inf), alpha = 0.15) +
  #facet_wrap(~CpG, ncol = 12) +
  xlab("Position relative to CpG of interest") +
  ylab("MAF") +
  theme_bw() +
  theme(legend.pos = "none", 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

RvNR_vedo_gv_plot <- ggarrange(RvNR_vedo_gv_plot_left, RvNR_clustered_gv_maf_xyplot, ncol = 2, widths = c(0.5, 1))

Cairo(width = 2500, height = 1250, type = "pdf", file = file.path(rvnrgvDir, "RvNR_vedo_GVs.pdf"), bg = "white", units = "px", dpi = 120)
#Cairo(width = 10000, height = 3000, type = "pdf", file = file.path(gvDir, "medstablor_df_GVs.pdf"), bg = "white", units = "px", dpi = 120)
print(RvNR_vedo_gv_plot)
dev.off()
```

```{r}
RvNR_vedo_sig_gapped_snps_df %>%
  dplyr::filter(GV_CpG_relpos %in% c(-1, 0, 1)) %>%
  dplyr::select


rsidphen <- getBM(attributes = c("refsnp_id", "clinical_significance", "phenotype_description"), 
                  filters = "snp_filter", 
                  values = unique(RvNR_vedo_sig_gapped_snps_df$RSID), 
                  mart = enssnp_mart)
```


### DMG

```{r RvNR dmg setup}
RvNR_vedo_mdmgDir <- file.path(outputDir, "dmg", "RvNR")
dir.create(RvNR_vedo_mdmgDir, recursive = T)
```

```{r RvNR vedo genes with multiple dmps}
RvNR_sig_genes <- do.call(c, lapply(strsplit(x = paste0(RvNR_vedo_sig$UCSC_RefGene_Name, ";", RvNR_vedo_sig$enhancer_gene), ";"), unique))
RvNR_sig_genes_freq <- sort(table(RvNR_sig_genes))
RvNR_sig_genes_bg <- unlist(lapply(names(RvNR_sig_genes_freq), function(gene){length(grep(gene, paste0(gmset_vedo_anno_gr$UCSC_RefGene_Name, ";", gmset_vedo_anno_gr$enhancer_gene)))}))
names(RvNR_sig_genes_bg) <- names(RvNR_sig_genes_freq)
RvNR_sig_genes_prop <- data.frame(frequency = as.vector(RvNR_sig_genes_freq),
                                  background = as.vector(RvNR_sig_genes_bg),
                                  proportion = as.vector(RvNR_sig_genes_freq)/as.vector(RvNR_sig_genes_bg),
                                  row.names = names(RvNR_sig_genes_freq))

write.csv(RvNR_sig_genes_freq, file.path(RvNR_vedo_mdmgDir, "RvNR_vedo_dmpgene_prop.csv"))
```

```{r RvNR vedo dmgs}
RvNR_vedo_top_genes <- lapply(strsplit(RvNR_vedo_top$UCSC_RefGene_Name, ";"), unique)
names(RvNR_vedo_top_genes) <- rownames(RvNR_vedo_top)
RvNR_vedo_top_genes_nz <- RvNR_vedo_top_genes[which(lapply(RvNR_vedo_top_genes, length) != 0)]

RvNR_vedo_top_genes_nz_df <- data.frame(CpG = gsub("(^cg[0-9]{8}).+$", "\\1", names(unlist(RvNR_vedo_top_genes_nz))), 
                                        Gene = unlist(RvNR_vedo_top_genes_nz))
RvNR_vedo_top_genes_nz_df$pvalue <- RvNR_vedo_top[RvNR_vedo_top_genes_nz_df$CpG, "P.Value"]

RvNR_vedo_top_genes_nz_list <- split(RvNR_vedo_top_genes_nz_df, RvNR_vedo_top_genes_nz_df$Gene)
RvNR_vedo_top_genes_nz_list <- RvNR_vedo_top_genes_nz_list[unlist(lapply(RvNR_vedo_top_genes_nz_list, nrow)) != 1]

RvNR_vedo_top_fisher <- lapply(X = RvNR_vedo_top_genes_nz_list, FUN = function(gene){
  #aggregation::fisher(gene$pvalue)
  unlist(EmpiricalBrownsMethod::empiricalBrownsMethod(data_matrix = getBeta(gmset_vedo_filtered_t1t2)[which(rownames(gmset_vedo_filtered_t1t2) %in% gene$CpG),],
                                                      p_values = gene$pvalue,
                                                      extra_info = T))
})

rm(RvNR_vedo_top_genes_nz_list)

RvNR_vedo_top_ma <- data.frame(Gene = names(RvNR_vedo_top_fisher), do.call(rbind, RvNR_vedo_top_fisher))
colnames(RvNR_vedo_top_ma) <- c("Gene", "Brown_pval", "Fisher_pval", "Scale_Factor_C", "DF")
RvNR_vedo_top_ma$Brown_padj <- p.adjust(RvNR_vedo_top_ma$Brown_pval)
RvNR_vedo_top_ma <- RvNR_vedo_top_ma[order(RvNR_vedo_top_ma$Brown_pval),]
RvNR_vedo_top_ma_sig <- RvNR_vedo_top_ma[which(RvNR_vedo_top_ma$Brown_padj<0.05),]

write.csv(RvNR_vedo_top_ma, file.path(RvNR_vedo_mdmgDir, "RvNR_vedo_dmg.csv"))
```

```{r RvNR vedo dmg visualization}
RvNR_vedo_mdmgplotDir <- file.path(RvNR_vedo_mdmgDir, "dmg_plots")
dir.create(RvNR_vedo_mdmgplotDir)

for(i in 1:nrow(RvNR_vedo_top_ma_sig)){
  Cairo(width = 1000, height = 750, type = "pdf", file = file.path(RvNR_vedo_mdmgplotDir, paste0(i, "_", RvNR_vedo_top_ma_sig$Gene[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(dmg_plot(gene_of_interest = RvNR_vedo_top_ma_sig$Gene[i], tophits_gr = RvNR_vedo_top_gr, smooth = T, significance = T))
  dev.off()
}

Cairo(width = 1000, height = 750, type = "pdf", file = file.path(RvNR_vedo_mdmgplotDir, paste0(11, "_", RvNR_top_ma_sig$Gene[11], ".pdf")), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = RvNR_vedo_top_ma_sig$Gene[i], tophits_gr = RvNR_vedo_top_gr, smooth = T, significance = F, stat = "identity"))
dev.off()

print(dmg_plot_continuous(gene_of_interest = RvNR_vedo_top_ma_sig$Gene[i], tophits_gr = RvNR_vedo_top_gr, smooth = T, pval_colname = "P.Value"))
```

### DMR

```{r vedo dmrs}
require(DMRcate)
require(ChIPpeakAnno)
require(biomaRt)

set.seed(123)
RvNR_dmr_vedo_anno <- cpg.annotate(object = gmset_vedo_filtered_t1t2,
                                   datatype = "array",
                                   arraytype = "EPIC",
                                   what = "M",
                                   analysis.type = "differential",
                                   design = design_vedo,
                                   contrasts = T,
                                   cont.matrix = contrast_mat_vedo,
                                   coef = "RvNR")
RvNR_dmr_vedo <- dmrcate(RvNR_dmr_vedo_anno, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
RvNR_dmr_vedo_gr <- extractRanges(RvNR_dmr_vedo, genome = "hg19")

write.csv(as.data.frame(RvNR_dmr_vedo_gr), file.path(dmrDir, "RvNR_vedo.csv"))

#Annotation
RvNR_dmr_vedo_cpaanno <- annotatePeakInBatch(myPeakList = RvNR_dmr_vedo_gr, 
                                             AnnotationData = genes(TxDb.Hsapiens.UCSC.hg19.knownGene),
                                             output="nearestBiDirectionalPromoters",
                                             bindingRegion=c(-5000, 1000))

RvNR_dmr_vedo_cpaanno$SYMBOL <- mapIds(x = org.Hs.eg.db, 
                                       keys = as.character(RvNR_dmr_vedo_cpaanno$gene_id), 
                                       column = "SYMBOL", 
                                       keytype = "ENTREZID", 
                                       multiVals="first")

names(RvNR_dmr_vedo_cpaanno) <- 1:length(RvNR_dmr_vedo_cpaanno)

write.csv(as.data.frame(RvNR_dmr_vedo_cpaanno), file.path(dmrDir, "RvNR_vedo_annotated.csv"))

vedodmrplot1 <- dmr_plot(dmr_gr = RvNR_dmr_vedo_gr[1], meth_data = getBeta(gmset_vedo_filtered_t1t2), anno_gr = gmset_vedo_anno_gr, meth_groups = pData(gmset_vedo_filtered_t1t2)$Response, title = "(no annotated gene)", highlight = c(start(RvNR_dmr_vedo_gr[1]), end(RvNR_dmr_vedo_gr[1])))
vedodmrplot2 <- dmr_plot(dmr_gr = RvNR_dmr_vedo_gr[2], meth_data = betas_vedo, anno_gr = gmset_vedo_anno_gr, meth_groups = pData(gmset_vedo_filtered)$Response, title = "(no annotated gene)", highlight = c(start(RvNR_dmr_vedo_gr[2]), end(RvNR_dmr_vedo_gr[2])))
vedodmrplot3 <- dmr_plot(dmr_gr = RvNR_dmr_vedo_gr[3], meth_data = betas_vedo, anno_gr = gmset_vedo_anno_gr, meth_groups = pData(gmset_vedo_filtered)$Response, title = "(no annotated gene)", highlight = c(start(RvNR_dmr_vedo_gr[3]), end(RvNR_dmr_vedo_gr[3])))
```

## RvNRvT2vT1

```{r vedo DM analyses, echo = F}
RvNRvT2vT1Dir <- file.path(dmpDir, "RvNRvT2vT1")
dir.create(RvNRvT2vT1Dir)

RvNRvT2vT1_vedo_top <- topTable(mvals_ebayes_vedo, coef = "RvNRvT2vT1", number = "Inf", adjust.method = "BH")
RvNRvT2vT1_vedo_top <- cbind(RvNRvT2vT1_vedo_top, Beta = coefficients(betas_ebayes_vedo)[rownames(RvNRvT2vT1_vedo_top), "RvNRvT2vT1"], data.frame(gmset_vedo_anno_gr[rownames(RvNRvT2vT1_vedo_top), ]))
write.csv(RvNRvT2vT1_vedo_top, file.path(RvNRvT2vT1Dir, "RvNRvT2vT1_vedo.csv"))

RvNRvT2vT1_vedo_sig <- RvNRvT2vT1_vedo_top[RvNRvT2vT1_vedo_top$adj.P.Val < 0.05,]
#write.csv(RvNRvT2vT1_vedo_sig, file.path(RvNRvT2vT1Dir, "RvNRvT2vT1_vedo_sig.csv"))

Cairo(width = 1000, height = 1000, type = "png", file = file.path(RvNRvT2vT1Dir, "volcano_vedo_RvNR.png"), bg = "white", units = "px", dpi = 120)
print(with(RvNRvT2vT1_vedo_top, volcano_plot(effect_sizes = Beta, 
                                             pvals = P.Value, 
                                             int_effect_threshold = 0.1, 
                                             significance = adj.P.Val < 0.05, 
                                             identifiers = rownames(RvNR_vedo_top), 
                                             title = "Vedolizumab: Responders vs non-responders")))
dev.off()

cpg_timeplot(cgid = rownames(RvNRvT2vT1_vedo_top)[1], 
             anno = gmset_vedo_anno,
             betas = getBeta(gmset_vedo_filtered)*100, 
             response = pData(gmset_vedo_filtered)$Response,
             timepoint = pData(gmset_vedo_filtered)$Timepoint,
             donor = pData(gmset_vedo_filtered)$Biobank_ID,
             response_cols = response_cols,
             enlarged = T
)
```

### DMG

```{r RvNRvT2vT1 dmg setup}
RvNRvT2vT1_vedo_mdmgDir <- file.path(outputDir, "dmg", "RvNRvT2vT1")
dir.create(RvNRvT2vT1_vedo_mdmgDir, recursive = T)
```

```{r RvNRvT2vT1 vedo dmgs}
RvNRvT2vT1_vedo_top_genes <- lapply(strsplit(RvNRvT2vT1_vedo_top$UCSC_RefGene_Name, ";"), unique)
names(RvNRvT2vT1_vedo_top_genes) <- rownames(RvNRvT2vT1_vedo_top)
RvNRvT2vT1_vedo_top_genes_nz <- RvNRvT2vT1_vedo_top_genes[which(lapply(RvNRvT2vT1_vedo_top_genes, length) != 0)]

RvNRvT2vT1_vedo_top_genes_nz_df <- data.frame(CpG = gsub("(^cg[0-9]{8}).+$", "\\1", names(unlist(RvNRvT2vT1_vedo_top_genes_nz))), 
                                              Gene = unlist(RvNRvT2vT1_vedo_top_genes_nz))
RvNRvT2vT1_vedo_top_genes_nz_df$pvalue <- RvNRvT2vT1_vedo_top[RvNRvT2vT1_vedo_top_genes_nz_df$CpG, "P.Value"]

RvNRvT2vT1_vedo_top_genes_nz_list <- split(RvNRvT2vT1_vedo_top_genes_nz_df, RvNRvT2vT1_vedo_top_genes_nz_df$Gene)
RvNRvT2vT1_vedo_top_genes_nz_list <- RvNRvT2vT1_vedo_top_genes_nz_list[unlist(lapply(RvNRvT2vT1_vedo_top_genes_nz_list, nrow)) != 1]

RvNRvT2vT1_vedo_top_fisher <- lapply(X = RvNRvT2vT1_vedo_top_genes_nz_list, FUN = function(gene){
  #aggregation::fisher(gene$pvalue)
  unlist(EmpiricalBrownsMethod::empiricalBrownsMethod(data_matrix = getBeta(gmset_vedo_filtered_t1t2)[which(rownames(gmset_vedo_filtered_t1t2) %in% gene$CpG),],
                                                      p_values = gene$pvalue,
                                                      extra_info = T))
})

rm(RvNRvT2vT1_vedo_top_genes_nz_list)

RvNRvT2vT1_vedo_top_ma <- data.frame(Gene = names(RvNRvT2vT1_vedo_top_fisher), do.call(rbind, RvNRvT2vT1_vedo_top_fisher))
colnames(RvNRvT2vT1_vedo_top_ma) <- c("Gene", "Brown_pval", "Fisher_pval", "Scale_Factor_C", "DF")
RvNRvT2vT1_vedo_top_ma$Brown_padj <- p.adjust(RvNRvT2vT1_vedo_top_ma$Brown_pval)
RvNRvT2vT1_vedo_top_ma <- RvNRvT2vT1_vedo_top_ma[order(RvNRvT2vT1_vedo_top_ma$Brown_pval),]
RvNRvT2vT1_vedo_top_ma_sig <- RvNRvT2vT1_vedo_top_ma[which(RvNRvT2vT1_vedo_top_ma$Brown_padj<0.05),]

write.csv(RvNRvT2vT1_vedo_top_ma, file.path(RvNRvT2vT1_vedo_mdmgDir, "RvNRvT2vT1_vedo_dmg.csv"))
```

# RvNR classification analysis

## glmnet

Unlike in the previous, we would now seek to find markers that are capable of generalization. To this end, we should perform no normalization as we seek to split the data into a training and test set. We can therefore utilize the noob preprocessing function, which performs background correction only.

```{r background correction using noob}
gmset_noob_vedo <- preprocessNoob(rgset_vedo)
gmset_noob_vedo <- mapToGenome(gmset_noob_vedo)

gmset_noob_vedo_filtered <- gmset_noob_vedo[-which(minfi::getAnnotation(gmset_noob_vedo)$chr %in% c("chrX", "chrY")),]
gmset_noob_vedo_filtered <- gmset_noob_vedo_filtered[!names(gmset_noob_vedo_filtered) %in% pmprobes,]
gmset_noob_vedo_t1t2 <- gmset_noob_vedo_filtered[, pData(gmset_noob_vedo_filtered)$Timepoint %in% c(1,2)]
```

```{r noob corrected without snps}
gmset_noob_vedo_nosnp <- gmset_noob_vedo[rownames(gmset_noob_vedo) %in% rownames(gmset_vedo_filtered_nosnp)]

gmset_noob_vedo_nosnp_nogap <- gmset_noob_vedo[rownames(gmset_noob_vedo) %in% rownames(gmset_vedo_filtered_nosnp_nogap)]
```

### All

```{r single cross-validation}
set.seed(918743)
RvNRT1T2_cvfit <- cv.glmnet(x = t(getM(gmset_noob_vedo_t1t2)), 
                            y = onehot_response, family = "binomial", 
                            alpha = 0.5,
                            type.measure = "class")

predict(RvNRT1T2_cvfit, newx = t(getM(gmset_noob_vedo_t1t2))[1:20,], s = "lambda.1se", type = "class")

RvNRT1T2_predictor_cpgs <- coefficients(RvNRT1T2_cvfit, s = "lambda.1se")[,1][coefficients(RvNRT1T2_cvfit)[,1] != 0]
RvNRT1T2_predictor_cpgs <- RvNRT1T2_predictor_cpgs[order(abs(RvNRT1T2_predictor_cpgs), decreasing = T)]

data.frame(RvNR_vedo_top[rownames(RvNR_vedo_top) %in% names(RvNRT1T2_predictor_cpgs)[-1],])

table(rownames(RvNR_vedo_sig) %in% names(RvNRT1T2_predictor_cpgs)[-1])
table(names(RvNRT1T2_predictor_cpgs)[-1] %in% rownames(RvNR_vedo_sig))

NDlib::cpg_strip_plot(cpg_num = "cg07281747", 
                      betas = getBeta(gmset_noob_vedo_t1t2), 
                      factor_interest = pData(gmset_noob_vedo_t1t2)$resptime)
```

```{r repeated cross-validation}
require(glmnet)
require(foreach)

elnetDir <- file.path(outputDir, "elasticnet")
dir.create(elnetDir)

onehot_response <- as.numeric(as.factor(pData(gmset_noob_vedo_t1t2)$Response))-1

reps <- 100

set.seed(243098)
repcv_elnetfit <- foreach(i=1:reps) %do%
  {
    cvfit <- cv.glmnet(x = t(getM(gmset_noob_vedo_t1t2)), 
                       y = onehot_response, family = "binomial", 
                       alpha = 0.5,
                       type.measure = "class")
    rep_predcpgs <- coefficients(cvfit, s = "lambda.1se")[,1][coefficients(cvfit)[,1] != 0]
    return(rep_predcpgs)
  }
saveRDS(repcv_elnetfit, file.path(rdsDir, "repcv_elnetfit.Rds"))

repcv_features <- unique(unlist(lapply(repcv_elnetfit, names)))

replor_df <- Reduce(function(x, y) merge(x, y, by = "CpG", all = T), lapply(repcv_elnetfit, function(fold){
  data.frame(CpG = names(fold),
             lOR = as.numeric(fold))
}))
replor_df <- data.frame(replor_df, row.names = replor_df$CpG)
replor_df <- replor_df[,-1]
replor_df[is.na(replor_df)] <- 0

#Median stabilization
medstablor <- rowMedians(as.matrix(replor_df))
names(medstablor) <- rownames(replor_df)
medstablor <- medstablor[medstablor != 0]
medstablor_df <- data.frame(RvNR_vedo_top[rownames(RvNR_vedo_top) %in% names(medstablor)[-1],],
                            med_stablor = medstablor[-1])

write.csv(medstablor, file.path(elnetDir, "median_stabilized_lor.csv"))
write.csv(medstablor_df, file.path(elnetDir, "median_stabilized_lor_df.csv"))
```

Time stability of the predictor CpGs across the three timepoints.

```{r time stability predictor CpGs}
time_vedo_t1t2t3_respdmps <- time_vedo_t1t2t3_top[rownames(medstablor_df),]

betas_vedo_t1t2t3_respdmps <- getBeta(gmset_vedo_filtered_t1t2t3)[rownames(medstablor_df),]
betas_vedo_t1t2t3_respdmps_melt <- melt(betas_vedo_t1t2t3_respdmps, varnames = c("CpG_ID", "SXS_POS"), value.name = "Beta")
betas_vedo_t1t2t3_respdmps_melt <- cbind(betas_vedo_t1t2t3_respdmps_melt, pData(gmset_vedo_filtered_t1t2t3)[betas_vedo_t1t2t3_respdmps_melt$SXS_POS, c("Timepoint", "Response", "Biobank_ID")])

```

```{r vedo genes with multiple dmps}
RvNR_vedo_mdmpDir <- file.path(dmpDir, "gene_mdmps")
dir.create(RvNR_vedo_mdmpDir)

medstablor_df_genes <- do.call(c, lapply(strsplit(x = paste0(medstablor_df$UCSC_RefGene_Name, ";", medstablor_df$enhancer_gene), ";"), unique))
medstablor_df_genes_freq <- sort(table(medstablor_df_genes))
medstablor_df_genes_bg <- unlist(lapply(names(medstablor_df_genes_freq), function(gene){length(grep(gene, paste0(gmset_vedo_anno_gr$UCSC_RefGene_Name, ";", gmset_vedo_anno_gr$enhancer_gene)))}))
names(medstablor_df_genes_bg) <- names(medstablor_df_genes_freq)
medstablor_df_genes_prop <- data.frame(frequency = as.vector(medstablor_df_genes_freq),
                                       background = as.vector(medstablor_df_genes_bg),
                                       proportion = as.vector(medstablor_df_genes_freq)/as.vector(medstablor_df_genes_bg),
                                       row.names = names(medstablor_df_genes_freq))

medstablor_df_genes_prop_mdmp <- medstablor_df_genes_prop[medstablor_df_genes_prop$frequency>1,]
medstablor_df_genes_prop_mdmp <- medstablor_df_genes_prop_mdmp[order(medstablor_df_genes_prop_mdmp$proportion, decreasing = T),]

write.csv(medstablor_df_genes_prop, file.path(dmpDir, "RvNR_vedo_DMP_prop.csv"))
```

```{r vedo dmeg}
#All DMEGs
medstablor_df_gene_mdmp <- rownames(medstablor_df_genes_prop_mdmp)
medstablor_df_gene_mdmp <- medstablor_df_gene_mdmp[-which(medstablor_df_gene_mdmp == "")]

medstablor_df_gene_mdmp_gr <- makeGRangesFromDataFrame(do.call(rbind, lapply(medstablor_df_gene_mdmp, function(i) {
  df <- medstablor_df[grep(i, paste0(medstablor_df$UCSC_RefGene_Name, ";", medstablor_df$enhancer_gene)),]
  df_min <- with(df, data.frame(seqnames = unique(seqnames),
                                start = min(start),
                                end = max(end),
                                gene = i))
  
  return(df_min)
})), keep.extra.columns = T)

for(i in 1:length(medstablor_df_gene_mdmp_gr)){
  Cairo(width = 2000, height = 1000, type = "pdf", file = file.path(RvNR_vedo_mdmpDir, paste0(medstablor_df_gene_mdmp_gr$gene[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(dmr_effect_plot(region_of_interest = medstablor_df_gene_mdmp_gr[i,], tophits = RvNR_vedo_top_gr, significance = F, smooth = F))
  dev.off()
}
```

#### Visualization

```{r density plot}
replor_melt <- melt(cbind(feature = rownames(replor_df), replor_df))
replor_melt <- replor_melt[replor_melt$feature != "(Intercept)",]
replor_melt$Stability <- "Unstable"
replor_melt$Stability[replor_melt$feature %in% names(medstablor)] <- "Stable"
replor_melt <- replor_melt[replor_melt$Stability == "Stable",]
replor_melt$feature <- factor(replor_melt$feature, levels = names(sort(medstablor[-1])))
replor_melt$Direction <- "Hypermethylated"
replor_melt$Direction[replor_melt$feature %in% names(medstablor[-1])[which(medstablor[-1]<0)]] <- "Hypomethylated"

stabreplor_plot <- ggplot(replor_melt, aes(x = feature, y = value, col = Direction)) +
  geom_jitter(alpha = 0.10) +
  geom_boxplot(alpha = 0.50) +
  theme_bw() +
  ylab("log(OR)") +
  labs(title = "100 repetitions of 10-fold cross-validations",
       subtitle = "Median-stabilized") +
  theme(axis.title.y = element_blank()) +
  coord_flip()
```

```{r heatmap}
RvNR_vedo_hm_anno_col <- data.frame(row.names = rownames(pData(gmset_vedo_filtered_t1t2)),
                                    Response = pData(gmset_vedo_filtered_t1t2)$Response,
                                    #Donor = pData(gmset_vedo_filtered_t1t2)$Biobank_ID,
                                    Timepoint = paste0("T", pData(gmset_vedo_filtered_t1t2)$Timepoint))

RvNR_vedo_hm_anno_row <- data.frame(row.names = rownames(medstablor_df),
                                    logOR = medstablor_df$med_stablor)

RvNR_vedo_hm_anno_colors <- list(
  logOR = c("blue", "white", "red")
)

RvNR_vedo_hm <- grid.grabExpr(pheatmap(mat = getBeta(gmset_vedo_filtered_t1t2)[names(medstablor)[-1],],
                                       annotation_col = RvNR_vedo_hm_anno_col,
                                       annotation_row = RvNR_vedo_hm_anno_row,
                                       annotation_colors = RvNR_vedo_hm_anno_colors,
                                       #breaks = breaksList, 
                                       show_colnames = F))

Cairo(width = 1000, height = 1000, type = "png", file = file.path(vedodmpDir, "heatmap_vedo_RvNR_sig.png"), bg = "white", units = "px", dpi = 120)
plot_grid(RvNR_vedo_hm)
dev.off()
```

## Response-associated DMPs across time

```{r vedo response dmps across time points 2 and 1}
RvNR_vedo_resp_time_df <- data.frame(CpG = rownames(medstablor_df),
                                     RvNR_beta = medstablor_df$Beta,
                                     RvNR_pval = medstablor_df$P.Value,
                                     RvNR_padj = medstablor_df$adj.P.Val,
                                     T2vT1_beta = T2vT1_vedo_top[rownames(medstablor_df),"Beta"],
                                     T2vT1_pval = T2vT1_vedo_top[rownames(medstablor_df),"P.Value"],
                                     T2vT1_padj = T2vT1_vedo_top[rownames(medstablor_df),"adj.P.Val"])
```

```{r vedo response differential methylation associated with time}
t1t2t3_samples <- pData(gmset_vedo_filtered)$Biobank_ID[pData(gmset_vedo_filtered)$Timepoint %in% c(3)]

gmset_vedo_filtered_t1t2t3 <- gmset_vedo_filtered[, pData(gmset_vedo_filtered)$Biobank_ID %in% t1t2t3_samples]
pData(gmset_vedo_filtered_t1t2t3)$Smoker <- "No"
pData(gmset_vedo_filtered_t1t2t3)$Smoker[pData(gmset_vedo_filtered_t1t2t3)$Smoking_status_baseline %in% c("Current smoker")] <- "Yes"

design_vedo_t1t2t3 <- model.matrix(~Timepoint + Response + Sex + Age_baseline + Smoker, data = pData(gmset_vedo_filtered_t1t2t3))
colnames(design_vedo_t1t2t3) <- c("(Intercept)", "Timepoint", "Response", "Male", "Age", "Smoker")

mvals_lmfit_vedo_t1t2t3 <- lmFit(getM(gmset_vedo_filtered_t1t2t3), design = design_vedo_t1t2t3)
mvals_ebayes_vedo_t1t2t3 <- eBayes(mvals_lmfit_vedo_t1t2t3)
time_vedo_t1t2t3_top <- topTable(mvals_ebayes_vedo_t1t2t3, coef = "Timepoint", number = "Inf", adjust.method = "BH")
time_vedo_t1t2t3_top <- cbind(time_vedo_t1t2t3_top, Beta = coefficients(mvals_ebayes_vedo_t1t2t3)[rownames(mvals_ebayes_vedo_t1t2t3), "Timepoint"], data.frame(gmset_vedo_anno_gr[rownames(time_vedo_t1t2t3_top), ]))
```

## Genetic variants

```{r genetic variants}
gvDir <- file.path(dmpDir, "genetic_variants")
dir.create(gvDir)

require(SNPlocs.Hsapiens.dbSNP151.GRCh38)
require(liftOver)
require(rsnps)
require(biomaRt)
require(viridis)

vedo_gh <- gaphunter(object = gmset_vedo_filtered_t1t2)
saveRDS(vedo_gh, file.path(rdsDir, "vedo_gh.Rds"))

medstablor_df_cpgs_gh <- intersect(rownames(vedo_gh$proberesults), rownames(medstablor_df))
medstablor_df_cpgs_gh <- vedo_gh$proberesults[medstablor_df_cpgs_gh,]

medstablor_df_cpgs_gh_anno_gr <- makeGRangesFromDataFrame(gmset_vedo_anno[rownames(medstablor_df_cpgs_gh),], keep.extra.columns = F, start.field = "pos", end.field = "pos")

hg19tohg38 <- import.chain("/home/ayliyim/archive/common_data/liftover_chains/hg19ToHg38.over.chain")

medstablor_df_cpgs_gh_anno_hg38_gr <- unlist(liftOver(medstablor_df_cpgs_gh_anno_gr, hg19tohg38))
#medstablor_df_cpgs_gh_anno_hg38_gr <- split(medstablor_df_cpgs_gh_anno_hg38_gr, seqnames(medstablor_df_cpgs_gh_anno_hg38_gr))
medstablor_df_cpgs_gh_anno_hg38_gr <- medstablor_df_cpgs_gh_anno_hg38_gr+50
seqlevelsStyle(medstablor_df_cpgs_gh_anno_hg38_gr) <- seqlevelsStyle(SNPlocs.Hsapiens.dbSNP151.GRCh38)

medstablor_df_cpgs_gh_snps <- snpsByOverlaps(x = SNPlocs.Hsapiens.dbSNP151.GRCh38, ranges = medstablor_df_cpgs_gh_anno_hg38_gr)
saveRDS(medstablor_df_cpgs_gh_snps, file.path(gvDir, "medstablor_df_GVs.Rds"))

#MAF
enssnp_mart <- useDataset(mart = useMart("ENSEMBL_MART_SNP"),dataset="hsapiens_snp")
medstablor_df_cpgs_gh_snps_df <- getBM(attributes = c("refsnp_id", "allele", "minor_allele_freq"), filters = "snp_filter", values = medstablor_df_cpgs_gh_snps$RefSNP_id, mart = enssnp_mart)
medstablor_df_cpgs_gh_snps_df <- medstablor_df_cpgs_gh_snps_df[!is.na(medstablor_df_cpgs_gh_snps_df$minor_allele_freq),]

medstablor_df_cpgs_gh_snps <- makeGRangesFromDataFrame(merge(data.frame(medstablor_df_cpgs_gh_snps), medstablor_df_cpgs_gh_snps_df, by.x = "RefSNP_id", by.y = "refsnp_id"), keep.extra.columns = T, start.field = "pos", end.field = "pos")

medstablor_df_cpgs_gh_snps_pcgv_ol <- findOverlaps(medstablor_df_cpgs_gh_snps, medstablor_df_cpgs_gh_anno_hg38_gr)

medstablor_df_cpgs_gh_snps_ol <- medstablor_df_cpgs_gh_snps[queryHits(medstablor_df_cpgs_gh_snps_pcgv_ol)]
medstablor_df_cpgs_gh_snps_ol$CpG <- names(medstablor_df_cpgs_gh_anno_hg38_gr)[subjectHits(medstablor_df_cpgs_gh_snps_pcgv_ol)]
medstablor_df_cpgs_gh_snps_ol$CpG_pos <- start(medstablor_df_cpgs_gh_anno_hg38_gr-50)[subjectHits(medstablor_df_cpgs_gh_snps_pcgv_ol)]
medstablor_df_cpgs_gh_snps_ol$GV_relpos <- medstablor_df_cpgs_gh_snps_ol$CpG_pos-start(medstablor_df_cpgs_gh_snps_ol)
medstablor_df_cpgs_gh_snps_ol$GV_type <- "probe GV"
medstablor_df_cpgs_gh_snps_ol$GV_type[medstablor_df_cpgs_gh_snps_ol$GV_relpos %in% c(-1, 0, 1)] <- "CpG GV"
medstablor_df_cpgs_gh_snps_ol$Gene <- medstablor_df$UCSC_RefGene_Name[match(medstablor_df_cpgs_gh_snps_ol$CpG, medstablor_df$Name)]
medstablor_df_cpgs_gh_snps_ol$Beta <- medstablor_df$P.Value[match(medstablor_df_cpgs_gh_snps_ol$CpG, medstablor_df$Name)]
medstablor_df_cpgs_gh_snps_ol$Pvalue <- medstablor_df$Beta[match(medstablor_df_cpgs_gh_snps_ol$CpG, medstablor_df$Name)]
medstablor_df_cpgs_gh_snps_df <- data.frame(medstablor_df_cpgs_gh_snps_ol)
medstablor_df_cpgs_gh_snps_df <- medstablor_df_cpgs_gh_snps_df[, c("CpG", "seqnames", "CpG_pos", "Beta", "Pvalue", "RefSNP_id", "start", "GV_relpos", "allele", "GV_type", "minor_allele_freq", "Gene")]
colnames(medstablor_df_cpgs_gh_snps_df) <- c("CpG", "Chromosome", "CpG_position", "Beta", "Pvalue", "RSID", "GV_position", "GV_CpG_relpos", "Alleles", "GV_type", "MAF_all_1000g", "Gene")

write.csv(medstablor_df_cpgs_gh_snps_df, file = file.path(gvDir, "medstablor_df_GVs.csv"))

#up
medstablor_df_cl_plot <- data.frame(Type = c("non-clustered", "clustered"),
                                    DMPs = c(nrow(medstablor_df)-nrow(medstablor_df_cpgs_gh), nrow(medstablor_df_cpgs_gh)),
                                    Group = "DMPs") %>%
  ggplot(aes(x = Group, y = DMPs, fill = Type)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = DMPs), position = position_stack(vjust = 0.5)) +
  scale_fill_viridis(discrete = T) +
  coord_flip() +
  ggtitle("DMPs with catalogued GVs") +
  theme_bw() +
  theme(legend.position = "bottom", 
        axis.title.y = element_blank(), 
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

#left
medstablor_df_typegv_df <- data.frame(table(medstablor_df_cpgs_gh_snps_df$GV_type))
medstablor_df_typegv_df$Var1 <- gsub(" GV", "", medstablor_df_typegv_df$Var1)
colnames(medstablor_df_typegv_df) <- c("Location", "GVs")

medstablor_df_typegv_plot <- ggplot(medstablor_df_typegv_df, aes(x = Location, y = GVs)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = GVs), position = position_stack(vjust = 0.5)) +
  ggtitle("Location of GVs") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

RvNR_vedo_gc_plot_tr <- ggarrange(medstablor_df_cl_plot, medstablor_df_typegv_plot, nrow = 2, align = "v", heights = c(0.5, 1))

#right
medstablor_df_cpgs_gv_plot <- data.frame(medstablor_df_cpgs_gh_snps_ol) %>%
  ggplot(aes(x = GV_relpos, y = minor_allele_freq, col = minor_allele_freq, alpha = -log(abs(GV_relpos+0.001)))) +
  #geom_line(aes(group = CpG)) +
  geom_point() +
  geom_rect(data = data.frame(x0 = -1.5, x1 = 1.5),
            inherit.aes = FALSE,
            aes(xmin = x0, xmax = x1, ymin = -Inf, ymax = Inf), alpha = 0.15) +
  #facet_wrap(~CpG, ncol = 12) +
  xlab("Position relative to CpG of interest") +
  ylab("MAF") +
  theme_bw() +
  theme(legend.pos = "none", 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

RvNR_vedo_gc_plot <- ggarrange(RvNR_vedo_gc_plot_tr, medstablor_df_cpgs_gv_plot, ncol = 2, widths = c(0.5, 1))

Cairo(width = 2500, height = 1250, type = "pdf", file = file.path(gvDir, "medstablor_df_GVs.pdf"), bg = "white", units = "px", dpi = 120)
#Cairo(width = 10000, height = 3000, type = "pdf", file = file.path(gvDir, "medstablor_df_GVs.pdf"), bg = "white", units = "px", dpi = 120)
print(RvNR_vedo_gc_plot)
dev.off()
```

## Comparison with pilot data

```{r import pilot results}
samplesheet_pilot <- read.csv(file.path("data", "samples", "samplesheet_PRJ0000008_CDMEDRESP_fixed.csv"), stringsAsFactors = F)
samplesheet_vedo_pilot <- samplesheet_pilot[samplesheet_pilot$Drug == "Vedolizumab",]

samplesheet_vedo_combined <- data.frame(Path = c(samplesheet_vedo_pilot$Path, samplesheet$Filepath),
                                        Slide = c(samplesheet_vedo_pilot$Slide, gsub("([0-9]{12})_R.+", "\\1", samplesheet$SXS_POS)),
                                        Pos = c(samplesheet_vedo_pilot$Array, gsub("[0-9]{12}_(R.+)", "\\1", samplesheet$SXS_POS)),
                                        Biobank_ID = c(samplesheet_vedo_pilot$Donor_ID, samplesheet$Biobank_ID),
                                        Time = c(samplesheet_vedo_pilot$Timepoint, paste0("T", samplesheet$Timepoint)),
                                        Response = c(samplesheet_vedo_pilot$Response, samplesheet$Response),
                                        Age = c(samplesheet_vedo_pilot$Age, samplesheet$Age_baseline),
                                        Sex = c(samplesheet_vedo_pilot$Sex, gsub("([A-Z]).+", "\\1", samplesheet$Sex)),
                                        Cohort = c(rep("Pilot", nrow(samplesheet_vedo_pilot)), rep("Discovery", nrow(samplesheet))),
                                        stringsAsFactors = F)

require(ewastools)
ewastools_combined <- read_idats(samplesheet_vedo_combined$Path)
ewastools_combined_betas <- dont_normalize(ewastools_combined)
ewastools_combined_snps <- call_genotypes(ewastools_combined_betas[ewastools_combined$manifest[probe_type == 'rs', index], ], learn = F)
ewastools_combined_snpagreement <- check_snp_agreement(genotypes = ewastools_combined_snps, 
                                                       donor_ids = samplesheet_vedo_combined$Biobank_ID, 
                                                       sample_ids = paste0(samplesheet_vedo_combined$Slide, "_", samplesheet_vedo_combined$Pos))
```

```{r gmset}
gmset_vedo_filtered_pilot <- readRDS(file.path("output", "01_DM_analyses", "191009", "rds", "gmset_vedo_filtered.Rds"))
```

```{r comparison statistics}
lmp <- function (modelobject) {
  if (class(modelobject) != "lm") stop("Not an object of class 'lm' ")
  f <- summary(modelobject)$fstatistic
  p <- pf(f[1],f[2],f[3],lower.tail=F)
  attributes(p) <- NULL
  return(p)
}

RvNR_vedo_top_pilot <- read.csv(file.path("output", "01_DM_analyses", "191009", "dmps", "vedolizumab", "RvNR_vedo.csv"), stringsAsFactors = F, row.names = 1)
RvNR_vedo_sig_pilot <- RvNR_vedo_top_pilot[RvNR_vedo_top_pilot$adj.P.Val<0.05,]

overlap_pilot_pilotsig_plot_df <- data.frame(CpG = RvNR_vedo_sig_pilot$X,
                                             Discovery_beta = RvNR_vedo_top[rownames(RvNR_vedo_sig_pilot), "Beta"],
                                             Discovery_pval = RvNR_vedo_top[rownames(RvNR_vedo_sig_pilot), "P.Value"],
                                             Pilot_beta = RvNR_vedo_sig_pilot$Beta,
                                             Pilot_pval = RvNR_vedo_sig_pilot$P.Value)

overlap_pilot_pilotsig_plot_lfit <- lm(Discovery_beta ~ Pilot_beta, data = overlap_pilot_pilotsig_plot_df)

overlap_pilot_pilotsig_plot <- ggplot(overlap_pilot_pilotsig_plot_df, aes(x = Pilot_beta, y = Discovery_beta, size = -log10(Discovery_pval))) +
  geom_hline(yintercept = 0, col = "grey") +
  geom_vline(xintercept = 0, col = "grey") +
  labs(title = "Pilot response-associated DMPs",
       subtitle = paste0("R^2 = ", round(summary(overlap_pilot_pilotsig_plot_lfit)$adj.r.squared, 2), 
                         "; p-value = ", formatC(lmp(overlap_pilot_pilotsig_plot_lfit), format = "e", digits = 2))) +
  geom_point() +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

overlap_pilot_discoverysig_plot_df <- data.frame(CpG = rownames(medstablor_df),
                                                 Discovery_beta = medstablor_df$Beta,
                                                 Discovery_pval = medstablor_df$P.Value,
                                                 Pilot_beta = RvNR_vedo_top_pilot[rownames(medstablor_df), "Beta"],
                                                 Pilot_pval = RvNR_vedo_top_pilot[rownames(medstablor_df), "P.Value"])

overlap_pilot_discoverysig_plot_lfit <- lm(Discovery_beta ~ Pilot_beta, data = overlap_pilot_discoverysig_plot_df)
overlap_pilot_discoverysig_plot_lfit_corr <- lm(Discovery_beta ~ Pilot_beta, data = overlap_pilot_discoverysig_plot_df[!overlap_pilot_discoverysig_plot_df$CpG %in% c("cg26181139", "cg13872627"),])

overlap_pilot_discoverysig_plot <- ggplot(overlap_pilot_discoverysig_plot_df, aes(x = Pilot_beta, y = Discovery_beta, size = -log10(Pilot_pval))) +
  geom_hline(yintercept = 0, col = "grey") +
  geom_vline(xintercept = 0, col = "grey") +
  labs(title = "Discovery response-associated DMPs",
       subtitle = paste0("R^2 = ", round(summary(overlap_pilot_discoverysig_plot_lfit)$adj.r.squared, 2), 
                         "; p-value = ", formatC(lmp(overlap_pilot_discoverysig_plot_lfit), format = "e", digits = 2))) +
  geom_point() +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

```{r Discovery DMPs all}
#Pilot
mval_vedo_discoverysig_pilot <- getM(gmset_vedo_filtered_pilot)[rownames(medstablor_df),]
pData(gmset_vedo_filtered_pilot)$Response[pData(gmset_vedo_filtered_pilot)$Response == "Nonresponder"] <- "NR"
pData(gmset_vedo_filtered_pilot)$Response[pData(gmset_vedo_filtered_pilot)$Response == "Responder"] <- "R"
mval_vedo_discoverysig_pilot_svd <- svd(t(mval_vedo_discoverysig_pilot-rowMeans(mval_vedo_discoverysig_pilot)))
mval_vedo_discoverysig_pilot_svd_df <- data.frame(PC1 = mval_vedo_discoverysig_pilot_svd$u[,1],
                                                  PC2 = mval_vedo_discoverysig_pilot_svd$u[,2],
                                                  Response = pData(gmset_vedo_filtered_pilot)$Response)

mval_vedo_discoverysig_pilot_svd_plot <- ggplot(mval_vedo_discoverysig_pilot_svd_df, aes(x = PC1, y = PC2, col = Response)) +
  geom_point() +
  labs(title = "PCA: Discovery DMPs in pilot data") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

#Discovery
mval_vedo_discoverysig_discovery <- getM(gmset_vedo_filtered_t1t2)[rownames(medstablor_df),]
mval_vedo_discoverysig_discovery_svd <- svd(t(mval_vedo_discoverysig_discovery-rowMeans(mval_vedo_discoverysig_discovery)))
mval_vedo_discoverysig_discovery_svd_df <- data.frame(PC1 = mval_vedo_discoverysig_discovery_svd$u[,1],
                                                      PC2 = mval_vedo_discoverysig_discovery_svd$u[,2],
                                                      Response = pData(gmset_vedo_filtered_t1t2)$Response,
                                                      ID = pData(gmset_vedo_filtered_t1t2)$Biobank_ID)

mval_vedo_discoverysig_discovery_svd_plot <- ggplot(mval_vedo_discoverysig_discovery_svd_df, aes(x = PC1, y = PC2, col = Response)) +
  geom_point() +
  labs(title = "PCA: Discovery DMPs in discovery data") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

```{r Pilot DMPs all}
#Pilot
mval_vedo_pilotsig_pilot <- getM(gmset_vedo_filtered_pilot)[rownames(RvNR_vedo_sig_pilot),]
mval_vedo_pilotsig_pilot_svd <- svd(t(mval_vedo_pilotsig_pilot-rowMeans(mval_vedo_pilotsig_pilot)))
mval_vedo_pilotsig_pilot_svd_df <- data.frame(PC1 = mval_vedo_pilotsig_pilot_svd$u[,1],
                                              PC2 = mval_vedo_pilotsig_pilot_svd$u[,2],
                                              Response = pData(gmset_vedo_filtered_pilot)$Response)

mval_vedo_pilotsig_pilot_svd_plot <- ggplot(mval_vedo_pilotsig_pilot_svd_df, aes(x = PC1, y = PC2, col = Response)) +
  geom_point() +
  labs(title = "PCA: Pilot DMPs in pilot data") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

#Discovery
mval_vedo_pilotsig_discovery <- getM(gmset_vedo_filtered_t1t2)[rownames(RvNR_vedo_sig_pilot),]
mval_vedo_pilotsig_discovery_svd <- svd(t(mval_vedo_pilotsig_discovery-rowMeans(mval_vedo_pilotsig_discovery)))
mval_vedo_pilotsig_discovery_svd_df <- data.frame(PC1 = mval_vedo_pilotsig_discovery_svd$u[,1],
                                                  PC2 = mval_vedo_pilotsig_discovery_svd$u[,2],
                                                  Response = pData(gmset_vedo_filtered_t1t2)$Response)

mval_vedo_pilotsig_discovery_svd_plot <- ggplot(mval_vedo_pilotsig_discovery_svd_df, aes(x = PC1, y = PC2, col = Response)) +
  geom_point() +
  labs(title = "PCA: Pilot DMPs in discovery data") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

```{r Overlap pilot discovery}
overlapping_pilot_discovery <- intersect(pData(gmset_vedo_filtered_pilot)$Donor_ID, pData(gmset_vedo_filtered_t1t2)$Biobank_ID)

gmset_vedo_filtered_pilot_ol <- gmset_vedo_filtered_pilot[,pData(gmset_vedo_filtered_pilot)$Donor_ID %in% overlapping_pilot_discovery]
gmset_vedo_filtered_discovery_ol <- gmset_vedo_filtered_t1t2[,pData(gmset_vedo_filtered_t1t2)$Biobank_ID %in% overlapping_pilot_discovery]
```

All the pilot patients were in the discovery cohort, so we just need to calculate the PCA for the discovery cohort.

```{r Pilot DMPs overlapping}
#Discovery
mval_vedo_pilotsig_discovery_ol <- getM(gmset_vedo_filtered_discovery_ol)[rownames(RvNR_vedo_sig_pilot),]
mval_vedo_pilotsig_discovery_ol_svd <- svd(t(mval_vedo_pilotsig_discovery_ol-rowMeans(mval_vedo_pilotsig_discovery_ol)))
mval_vedo_pilotsig_discovery_ol_svd_df <- data.frame(PC1 = mval_vedo_pilotsig_discovery_ol_svd$u[,1],
                                                     PC2 = mval_vedo_pilotsig_discovery_ol_svd$u[,2],
                                                     Response = pData(gmset_vedo_filtered_discovery_ol)$Response,
                                                     Biobank_ID = pData(gmset_vedo_filtered_discovery_ol)$Biobank_ID)

mval_vedo_pilotsig_discovery_ol_svd_plot <- ggplot(mval_vedo_pilotsig_discovery_ol_svd_df, aes(x = PC1, y = PC2, col = Response)) +
  geom_point() +
  labs(title = "PCA: Pilot DMPs in discovery data",
       subtitle = "Overlapping samples only") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

```{r strange sample F096 and F097}
pData(gmset_vedo_filtered_pilot_ol)[pData(gmset_vedo_filtered_pilot_ol)$Biobank_ID %in% c("F096", "F097"),]
pData(gmset_vedo_filtered_pilot_ol)[pData(gmset_vedo_filtered_pilot_ol)$Donor_ID %in% c("F096", "F097"),]
```

```{r vedo dmeg location}
GPR146 <- RvNR_vedo_top_gr[grep("GPR146", paste0(RvNR_vedo_top_gr$GencodeBasicV12_NAME)),]
GPR146_gr <- range(GPR146, ignore.strand = T)
GPR146_gr$gene <- "GPR146"

Cairo(width = 2000, height = 1000, type = "pdf", file = file.path(outputDir, "GPR146_difference.pdf"), bg = "white", units = "px", dpi = 120)
print(dmr_effect_plot(region_of_interest = GPR146_gr, tophits = GPR146_vedo_gr, smooth = F, enhancers = F))
dev.off()

pData(gmset_vedo_filtered_t1t2)$Pilot <- "Discovery"
pData(gmset_vedo_filtered_t1t2)$Pilot[pData(gmset_vedo_filtered_t1t2)$Biobank_ID %in% pData(gmset_vedo_filtered_pilot)$Donor_ID] <- "Pilot"

Cairo(width = 2000, height = 1000, type = "pdf", file = file.path(outputDir, "GPR146_methylation_studygroup.pdf"), bg = "white", units = "px", dpi = 120)
dmr_plot(dmr_gr = GPR146_gr, 
         meth_data = getBeta(gmset_vedo_filtered_t1t2),
         anno_gr = gmset_vedo_anno_gr,
         meth_groups = paste0(pData(gmset_vedo_filtered_t1t2)$Response, "_", pData(gmset_vedo_filtered_t1t2)$Pilot),
         title = "GPR146") +
  ylab("% Methylation") +
  theme(legend.pos = "bottom")
dev.off()

#Stratified by response time

Cairo(width = 800, height = 800, type = "pdf", file = file.path(outputDir, "GPR146_cg20603222.pdf"), bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg20603222", betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$resptime, type = "SE") + 
  labs(title = "cg20603222",
       subtitle = paste0(RvNR_vedo_top["cg20603222",]$seqnames, ":", RvNR_vedo_top["cg20603222",]$start)) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
dev.off()

Cairo(width = 800, height = 800, type = "pdf", file = file.path(outputDir, "GPR146_cg26224354.pdf"), bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg26224354", betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$resptime, type = "SE") + 
  labs(title = "cg26224354",
       subtitle = paste0(RvNR_vedo_top["cg26224354",]$seqnames, ":", RvNR_vedo_top["cg26224354",]$start)) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
dev.off()

Cairo(width = 800, height = 800, type = "pdf", file = file.path(outputDir, "GPR146_cg01414857.pdf"), bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg01414857", betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$resptime, type = "SE") + 
  labs(title = "cg01414857",
       subtitle = paste0(RvNR_vedo_top["cg01414857",]$seqnames, ":", RvNR_vedo_top["cg01414857",]$start)) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
dev.off()

#Stratified by response time and study

Cairo(width = 1000, height = 800, type = "pdf", file = file.path(outputDir, "GPR146_cg20603222_study.pdf"), bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg20603222", betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = paste0(pData(gmset_vedo_filtered_t1t2)$resptime, "_", pData(gmset_vedo_filtered_t1t2)$Pilot), type = "SE") + 
  labs(title = "cg20603222",
       subtitle = paste0(RvNR_vedo_top["cg20603222",]$seqnames, ":", RvNR_vedo_top["cg20603222",]$start)) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
dev.off()

Cairo(width = 1000, height = 800, type = "pdf", file = file.path(outputDir, "GPR146_cg26224354_study.pdf"), bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg26224354", betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = paste0(pData(gmset_vedo_filtered_t1t2)$resptime, "_", pData(gmset_vedo_filtered_t1t2)$Pilot), type = "SE") + 
  labs(title = "cg26224354",
       subtitle = paste0(RvNR_vedo_top["cg26224354",]$seqnames, ":", RvNR_vedo_top["cg26224354",]$start)) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
dev.off()

Cairo(width = 1000, height = 800, type = "pdf", file = file.path(outputDir, "GPR146_cg01414857_study.pdf"), bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg01414857", betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = paste0(pData(gmset_vedo_filtered_t1t2)$resptime, "_", pData(gmset_vedo_filtered_t1t2)$Pilot), type = "SE") + 
  labs(title = "cg01414857",
       subtitle = paste0(RvNR_vedo_top["cg01414857",]$seqnames, ":", RvNR_vedo_top["cg01414857",]$start)) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
dev.off()

```

## Enrichment analyses

```{r vedo sig dmp KEGG overrepresentation}
require(missMethyl)

RvNR_vedo_dmp_enrichment <- gometh(sig.cpg = rownames(medstablor_df), all.cpg = rownames(gmset_vedo_anno), collection = "KEGG", array.type = "EPIC", plot.bias = FALSE, prior.prob = TRUE)
RvNR_vedo_dmp_enrichment <- RvNR_vedo_dmp_enrichment[order(RvNR_vedo_dmp_enrichment$P.DE),]
write.csv(RvNR_vedo_dmp_enrichment, file.path(vedodmpDir, "vedo_dmp_enrichment.csv"))
```

## Prediction analyses

```{r prediction elastic net}
#Training accuracy
train_x <- rbind(1, getM(gmset_noob_vedo_t1t2)[names(medstablor[-1]),])
train_lodds <- c(medstablor %*% train_x)
train_lodds_df <- data.frame(lodds = train_lodds,
                             Predicted = ifelse(train_lodds>0, "R", "NR"),
                             Response = pData(gmset_noob_vedo_t1t2)$Response,
                             ID = pData(gmset_noob_vedo_t1t2)$Biobank_ID,
                             Time = pData(gmset_noob_vedo_t1t2)$Timepoint)
train_lodds_df$Correct <- train_lodds_df$Predicted == train_lodds_df$Response

#T3 accuracy
gmset_noob_vedo_t3 <- gmset_noob_vedo[names(medstablor[-1]), gmset_noob_vedo$Timepoint == 3]
t3_x <- rbind(1, getM(gmset_noob_vedo_t3))
t3_lodds <- c(medstablor %*% t3_x)
t3_lodds_df <- data.frame(lodds = t3_lodds,
                          Predicted = ifelse(t3_lodds>0, "R", "NR"),
                          Response = pData(gmset_noob_vedo_t3)$Response,
                          ID = pData(gmset_noob_vedo_t3)$Biobank_ID)
t3_lodds_df$Correct <- t3_lodds_df$Predicted == t3_lodds_df$Response

#Pilot accuracy
rgset_vedo_pilot <- read.metharray.exp(targets = samplesheet_vedo_pilot)
gmset_noob_vedo_pilot <- preprocessNoob(rgset_vedo_pilot)
gmset_noob_vedo_pilot <- mapToGenome(gmset_noob_vedo_pilot)

pilot_x <- rbind(1, getM(gmset_noob_vedo_pilot[names(medstablor[-1]),]))
pilot_lodds <- c(medstablor %*% pilot_x)
pilot_lodds_df <- data.frame(lodds = pilot_lodds,
                             Predicted = ifelse(pilot_lodds>0, "R", "NR"),
                             Response = pData(gmset_noob_vedo_pilot)$Response,
                             ID = pData(gmset_noob_vedo_pilot)$Donor_ID,
                             Time = pData(gmset_noob_vedo_pilot)$Timepoint,
                             stringsAsFactors = F)
pilot_lodds_df$Response[pilot_lodds_df$Response == "Responder"] <- "R"
pilot_lodds_df$Response[pilot_lodds_df$Response == "Nonresponder"] <- "NR"
pilot_lodds_df$Correct <- pilot_lodds_df$Predicted == pilot_lodds_df$Response

#ROC
require(pROC)
pilot_roc <- roc(as.numeric(as.factor(pData(gmset_noob_vedo_pilot)$Response))-1, 
                 exp(pilot_lodds)/(1+exp(pilot_lodds)),
                 ci = T)
plot_pilot_roc <- ggroc(pilot_roc, legacy.axes = T) +
  geom_point(aes(x = 1-pilot_roc$specificities, y = pilot_roc$sensitivities)) +
  geom_abline(intercept = 0, slope = 1, alpha = 0.5, linetype = "dashed") +
  xlab("1-Specificity") +
  ylab("Sensitivity") +
  theme_bw() +
  labs(title = "Receiver operator characteristic",
       subtitle = "AUC = 0.75; 95% CI [0.51, 0.99]") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(dmpDir, "pilot_roc.pdf"), bg = "white", units = "px", dpi = 120)
plot_pilot_roc
dev.off()

```

# HORAIZON

```{r horaizon discval selection}
rgset_selected <- rgset_vedodiscval[,pData(rgset_vedodiscval)$Timepoint %in% c("T1", "T2")]

rgset_vedodiscval_noF241 <- rgset_selected[,pData(rgset_selected)$Biobank_ID != "F241"]
```

## Discovery (old method; noob no SNP no gap)

```{r old features}
features_old <- read.csv("output/DM_analyses/01_discovery/191106/processed_data/features_noSNP_nogap.csv") %>%
  dplyr::mutate(cgid = gsub("^(.+?)_.+$", "\\1", Response))
```

```{r horaizon discovery old method}
rgset_vedodisc_old <- rgset_vedodiscval_noF241[,pData(rgset_vedodiscval_noF241)$Cohort == "Discovery"]

vedodisc_old_dir <- file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "discovery_old")
dir.create(vedodisc_old_dir)

vedodisc_old <- rkgreslib::preprocessHoraizon(rgset = rgset_vedodisc_old,
                                              colname_of_interest = "Response", 
                                              variable_of_interest = "R",
                                              allosome_removal = F, 
                                              snp_removal = F,
                                              gap_removal = F)

vedodisc_old_noSNP_nogap <- vedodisc_old$X[,c(1, which(gsub("^(.+?)_.+$", "\\1", colnames(vedodisc_old$X)) %in% features_old$cgid))]

data.table::fwrite(vedodisc_old_noSNP_nogap, 
                   file.path(vedodisc_old_dir, "All_disc_noob_noSNP_nogap_X.csv"))

vedodisc_old_noSNP_nogap_sample_metadata <- data.frame(DonorID = pData(rgset_vedodisc_old)$Biobank_ID,
                                                       Timepoint = pData(rgset_vedodisc_old)$Timepoint,
                                                       Response = pData(rgset_vedodisc_old)$Response,
                                                       Cohort = "Discovery",
                                                       row.names = colnames(rgset_vedodisc_old))

write.csv(vedodisc_old_noSNP_nogap_sample_metadata, 
          file.path(vedodisc_old_dir, "All_disc_noob_noSNP_nogap_sample_metadata.csv"))

```

```{r horaizon validation T1 old method}
rgset_vedodisc_t1 <- rgset_vedodisc_old[,pData(rgset_vedodisc_old)$Timepoint == "T1"]

vedodisc_t1_old_noSNP_nogap <- vedodisc_old_noSNP_nogap[colnames(rgset_vedodisc_t1),]

data.table::fwrite(vedodisc_t1_old_noSNP_nogap, 
                   file.path(vedodisc_old_dir, "T1_disc_noob_noSNP_nogap_X.csv"))

vedodisc_t1_old_noSNP_nogap_sample_metadata <- vedodisc_old_noSNP_nogap_sample_metadata[rownames(vedodisc_t1_old_noSNP_nogap),]

write.csv(vedodisc_t1_old_noSNP_nogap_sample_metadata, 
          file.path(vedodisc_old_dir, "T1_disc_noob_noSNP_nogap_sample_metadata.csv"))
```

```{r horaizon validation T2 old method}
rgset_vedodisc_t2 <- rgset_vedodisc_old[,pData(rgset_vedodisc_old)$Timepoint == "T2"]

vedodisc_t2_old_noSNP_nogap <- vedodisc_old_noSNP_nogap[colnames(rgset_vedodisc_t2),]

data.table::fwrite(vedodisc_t2_old_noSNP_nogap, 
                   file.path(vedodisc_old_dir, "T2_disc_noob_noSNP_nogap_X.csv"))

vedodisc_t2_old_noSNP_nogap_sample_metadata <- vedodisc_old_noSNP_nogap_sample_metadata[rownames(vedodisc_t2_old_noSNP_nogap),]

write.csv(vedodisc_t2_old_noSNP_nogap_sample_metadata, 
          file.path(vedodisc_old_dir, "T2_disc_noob_noSNP_nogap_sample_metadata.csv"))
```

## Validation (old method; noob no SNP no gap)

```{r horaizon validation old method}
rgset_vedoval_old <- rgset_vedodiscval_noF241[,pData(rgset_vedodiscval_noF241)$Cohort == "Validation"]

vedoval_old_dir <- file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "validation_old")
dir.create(vedoval_old_dir)

vedoval_old <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_old,
                                             colname_of_interest = "Response", 
                                             variable_of_interest = "R",
                                             allosome_removal = F, 
                                             snp_removal = F,
                                             gap_removal = F)

vedoval_old_noSNP_nogap <- vedoval_old$X[,c(1, which(gsub("^(.+?)_.+$", "\\1", colnames(vedoval_old$X)) %in% features_old$cgid))]

data.table::fwrite(vedoval_old_noSNP_nogap, 
                   file.path(vedoval_old_dir, "All_val_noob_noSNP_nogap_X.csv"))

vedoval_old_noSNP_nogap_sample_metadata <- data.frame(DonorID = pData(rgset_vedoval_old)$Biobank_ID,
                                                      Timepoint = pData(rgset_vedoval_old)$Timepoint,
                                                      Response = NA,
                                                      Cohort = "Validation",
                                                      row.names = colnames(rgset_vedoval_old))

write.csv(vedoval_old_noSNP_nogap_sample_metadata, 
          file.path(vedoval_old_dir, "All_val_noob_noSNP_nogap_sample_metadata.csv"))

```

```{r horaizon validation T1 old method}
rgset_vedoval_t1 <- rgset_vedoval_old[,pData(rgset_vedoval_old)$Timepoint == "T1"]

vedoval_t1_old_noSNP_nogap <- vedoval_old_noSNP_nogap[colnames(rgset_vedoval_t1),]

data.table::fwrite(vedoval_t1_old_noSNP_nogap, 
                   file.path(vedoval_old_dir, "T1_val_noob_noSNP_nogap_X.csv"))

vedoval_t1_old_noSNP_nogap_sample_metadata <- vedoval_old_noSNP_nogap_sample_metadata[rownames(vedoval_t1_old_noSNP_nogap),]

write.csv(vedoval_t1_old_noSNP_nogap_sample_metadata, 
          file.path(vedoval_old_dir, "T1_val_noob_noSNP_nogap_sample_metadata.csv"))
```

```{r horaizon validation T2 old method}
rgset_vedoval_t2 <- rgset_vedoval_old[,pData(rgset_vedoval_old)$Timepoint == "T2"]

vedoval_t2_old_noSNP_nogap <- vedoval_old_noSNP_nogap[colnames(rgset_vedoval_t2),]

data.table::fwrite(vedoval_t2_old_noSNP_nogap, 
                   file.path(vedoval_old_dir, "T2_val_noob_noSNP_nogap_X.csv"))

vedoval_t2_old_noSNP_nogap_sample_metadata <- vedoval_old_noSNP_nogap_sample_metadata[rownames(vedoval_t2_old_noSNP_nogap),]

write.csv(vedoval_t2_old_noSNP_nogap_sample_metadata, 
          file.path(vedoval_old_dir, "T2_val_noob_noSNP_nogap_sample_metadata.csv"))
```

## Validation (new method; noob no SNP no gap)

```{r horaizon validation}
rgset_vedoval_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedodiscval_noF241,
                                                        colname_of_interest = "Response", 
                                                        variable_of_interest = "R",
                                                        allosome_removal = T, 
                                                        snp_removal = F, 
                                                        outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_betas_noob"))

write.csv(data.frame(DonorID = pData(rgset_vedoval)$Biobank_ID,
                     Timepoint = pData(rgset_vedoval)$Timepoint,
                     row.names = colnames(rgset_vedoval)), 
          file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_betas_noob", "sample_metadata.csv"))
```

```{r horaizon validation T1}
rgset_vedoval_t1 <- rgset_vedoval_noF241[,pData(rgset_vedoval_noF241)$Timepoint == "T1"]

rgset_vedoval_t1_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t1,
                                                           colname_of_interest = "Response", 
                                                           variable_of_interest = "R",
                                                           allosome_removal = T, 
                                                           snp_removal = F, 
                                                           outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob"))

rgset_vedoval_t1_horaizon$X[,1] <- rgset_vedoval_t1$Biobank_ID

data.table::fwrite(rgset_vedoval_t1_horaizon$X, 
                   file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob", "X.csv"))

write.csv(data.frame(DonorID = pData(rgset_vedoval_t1)$Biobank_ID,
                     Timepoint = pData(rgset_vedoval_t1)$Timepoint,
                     row.names = colnames(rgset_vedoval_t1)), 
          file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob", "T1_sample_metadata.csv"))
```

```{r horaizon validation T2}
rgset_vedoval_t2 <- rgset_vedoval_noF241[,pData(rgset_vedoval_noF241)$Timepoint == "T2"]

rgset_vedoval_t2_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_t2,
                                                           colname_of_interest = "Response", 
                                                           variable_of_interest = "R",
                                                           allosome_removal = T, 
                                                           snp_removal = F, 
                                                           outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob"))

rgset_vedoval_t2_horaizon$X[,1] <- rgset_vedoval_t2$Biobank_ID

data.table::fwrite(rgset_vedoval_t2_horaizon$X, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob", "X.csv"))

write.csv(data.frame(DonorID = pData(rgset_vedoval_t2)$Biobank_ID,
                     Timepoint = pData(rgset_vedoval_t2)$Timepoint,
                     row.names = colnames(rgset_vedoval_t2)), 
          file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob", "T2_sample_metadata.csv"))
```

```{r horaizon performance binary}
vedoval_predictions_t1 <- as.data.frame(readxl::read_excel(file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "predictions.xlsx")))

predstats <- function(predicted, actual){
  confusion_matrix <- data.frame(table(predicted = predicted, actual = actual))
  
  prediction_stats <- data.frame(TP = confusion_matrix$Freq[4],
                                 TN = confusion_matrix$Freq[1],
                                 FP = confusion_matrix$Freq[2],
                                 FN = confusion_matrix$Freq[3]) %>%
    dplyr::mutate(Precision = TP/(TP+FP),
                  Recall = TP/(TP+FN),
                  Accuracy = (TP+TN)/(TP+FP+FN+TN),
                  F1score = 2*(Recall*Precision)/(Recall+Precision))
  
  return(prediction_stats)
}

vedoval_predictionstats <- foreach(i = 3:ncol(vedoval_predictions_t1), .combine = rbind) %do%
  predstats(predicted = vedoval_predictions_t1[,i], 
            actual = as.character(vedoval_predictions_t1$Actual))

rownames(vedoval_predictionstats) <- paste0("model", 1:10)

write.csv(vedoval_predictionstats, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "predictions_stats.csv"))
```

```{r horaizon performance probabilities}
require(plotROC)

vedoval_prediction_probabilities_t1 <- as.data.frame(readxl::read_excel(file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "210906", "horaizon", "predictions_probabilities.xlsx")))

vedoval_prediction_probabilities_t1 <- merge(data.frame(pData(rgset_vedoval_t1)[,c("Biobank_ID", "Response")]), vedoval_prediction_probabilities_t1, by.x = "Biobank_ID", by.y = "SampleID")
vedoval_prediction_probabilities_t1$Response_bin <- (vedoval_prediction_probabilities_t1$Response == "R")*1

Cairo(width = 1250, height = 1250, type = "pdf", file = file.path("model2_ROC.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(vedoval_prediction_probabilities_t1, aes(m = Model2, d = Response_bin)) +
  geom_roc(labels = FALSE, n.cuts = 0) +
  geom_abline(intercept = 0, col = "grey", linetype = "dashed") +
  theme_bw() +
  labs(title = "Model 2",
       y = "TPR",
       x = "FPR")
dev.off()

Cairo(width = 1250, height = 1250, type = "pdf", file = file.path("model10_ROC.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(vedoval_prediction_probabilities_t1, aes(m = Model10, d = Response_bin)) +
  geom_roc(labels = FALSE, n.cuts = 0) +
  geom_abline(intercept = 0, col = "grey", linetype = "dashed") +
  theme_bw() +
  labs(title = "Model 10",
       y = "TPR",
       x = "FPR")
dev.off()

Cairo(width = 1500, height = 1250, type = "pdf", file = file.path("model_both_ROC.pdf"), bg = "white", units = "px", dpi = 120)
vedoval_prediction_probabilities_t1 %>%
  tidyr::pivot_longer(-c(Biobank_ID, Response, Response_bin), names_to = "Model", values_to = "Probabilities") %>%
  ggplot(aes(m = Probabilities, d = Response, col = Model)) +
  geom_roc(labels = FALSE, n.cuts = 0) +
  geom_abline(intercept = 0, col = "grey", linetype = "dashed") +
  theme_bw() +
  labs(y = "TPR",
       x = "FPR")
dev.off()

```

## Discovery and validation (normalized together)

```{r horaizon all}
vedodiscval_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedodiscval_noF241,
                                                      colname_of_interest = "Response", 
                                                      variable_of_interest = "R",
                                                      allosome_removal = T, 
                                                      snp_removal = F, 
                                                      outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discval"))

vedodiscval_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodiscval_noF241)$Biobank_ID,
                                           Timepoint = pData(rgset_vedodiscval_noF241)$Timepoint,
                                           Response = pData(rgset_vedodiscval_noF241)$Response,
                                           Cohort = pData(rgset_vedodiscval_noF241)$Cohort,
                                           row.names = colnames(rgset_vedodiscval_noF241)) %>%
  dplyr::mutate(Response = ifelse(Cohort == "Validation", NA, Response))

write.csv(vedodiscval_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discval", "samples_metadata.csv"))
```

```{r horaizon T1}
vedodiscval_horaizon_samples_T1 <- vedodiscval_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_horaizon_samples, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discval", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob_discval_sample_metadata.csv"))

data.table::fwrite(vedodiscval_horaizon$X[rownames(vedodiscval_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discval", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob_discval.csv"))
```

```{r horaizon T2}
vedodiscval_horaizon_samples_T2 <- vedodiscval_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discval", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob_discval_sample_metadata.csv"))

data.table::fwrite(vedodiscval_horaizon$X[rownames(vedodiscval_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discval", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob_discval.csv"))
```

## Discovery and validation (normalized together; stricter SNP removal)

```{r horaizon discval selection}
rgset_selected <- rgset_vedodiscval[,pData(rgset_vedodiscval)$Timepoint %in% c("T1", "T2")]

rgset_vedodiscval_noF241 <- rgset_selected[,pData(rgset_selected)$Biobank_ID != "F241"]
```

```{r horaizon all}
vedodiscval_gh0.1_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedodiscval_noF241,
                                                            colname_of_interest = "Response", 
                                                            variable_of_interest = "R",
                                                            allosome_removal = T, 
                                                            snp_removal = T, 
                                                            min_maf = 0,
                                                            gap_removal = T,
                                                            gap_threshold = 0.1,
                                                            outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_betas_noob_discval_gh010"))

vedodiscval_gh0.1_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodiscval_noF241)$Biobank_ID,
                                                 Timepoint = pData(rgset_vedodiscval_noF241)$Timepoint,
                                                 Response = pData(rgset_vedodiscval_noF241)$Response,
                                                 Cohort = pData(rgset_vedodiscval_noF241)$Cohort,
                                                 row.names = colnames(rgset_vedodiscval_noF241))

vedodiscval_gh0.1_horaizon_samples <- vedodiscval_gh0.1_horaizon_samples %>%
  dplyr::mutate(Response = ifelse(Cohort == "Validation", NA, Response))

write.csv(vedodiscval_gh0.1_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_betas_noob_discval_gh010", "samples_metadata.csv"))
```

```{r horaizon T1}
vedodiscval_gh0.1_horaizon_samples_T1 <- vedodiscval_gh0.1_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_gh0.1_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob_discval_gh010_sample_metadata.csv"))

data.table::fwrite(vedodiscval_gh0.1_horaizon$X[rownames(vedodiscval_gh0.1_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T1_betas_noob_discval_gh010.csv"))
```

```{r horaizon T2}
vedodiscval_gh0.1_horaizon_samples_T2 <- vedodiscval_gh0.1_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_gh0.1_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob_discval_gh010_sample_metadata.csv"))

data.table::fwrite(vedodiscval_gh0.1_horaizon$X[rownames(vedodiscval_gh0.1_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_validation_T2_betas_noob_discval_gh010.csv"))
```

## Discovery and validation (normalized together with funnorm; stricter SNP removal)

```{r horaizon funnorm gh010}
vedodiscval_gh0.1_funnorm_horaizon <- preprocessHoraizon(rgset = rgset_vedodiscval_noF241,
                                                         method = "funnorm",
                                                         colname_of_interest = "Response", 
                                                         variable_of_interest = "R",
                                                         allosome_removal = T, 
                                                         snp_removal = T, 
                                                         min_maf = 0,
                                                         gap_removal = T,
                                                         gap_threshold = 0.1,
                                                         outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh010"))

vedodiscval_gh0.1_funnorm_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodiscval_noF241)$Biobank_ID,
                                                         Timepoint = pData(rgset_vedodiscval_noF241)$Timepoint,
                                                         Response = pData(rgset_vedodiscval_noF241)$Response,
                                                         Cohort = pData(rgset_vedodiscval_noF241)$Cohort,
                                                         row.names = colnames(rgset_vedodiscval_noF241))

vedodiscval_gh0.1_funnorm_horaizon_samples <- vedodiscval_gh0.1_funnorm_horaizon_samples %>%
  dplyr::mutate(Response = ifelse(Cohort == "Validation", NA, Response))

write.csv(vedodiscval_gh0.1_funnorm_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh010", "samples_metadata.csv"))
```

```{r horaizon T1 funnorm gh010}
vedodiscval_gh0.1_funnorm_horaizon_samples_T1 <- vedodiscval_gh0.1_funnorm_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_gh0.1_funnorm_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh010",  "T1_sample_metadata.csv"))

data.table::fwrite(vedodiscval_gh0.1_funnorm_horaizon$X[rownames(vedodiscval_gh0.1_funnorm_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh010", "T1_X.csv"))
```

```{r horaizon T2 funnorm gh010}
vedodiscval_gh0.1_funnorm_horaizon_samples_T2 <- vedodiscval_gh0.1_funnorm_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_gh0.1_funnorm_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh010",  "T2_sample_metadata.csv"))

data.table::fwrite(vedodiscval_gh0.1_funnorm_horaizon$X[rownames(vedodiscval_gh0.1_funnorm_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh010", "T2_X.csv"))
```

## Discovery and validation (normalized together with funnorm; no SNP removal)

```{r horaizon funnorm all}
vedodiscval_funnorm_horaizon <- preprocessHoraizon(rgset = rgset_vedodiscval_noF241,
                                                   method = "funnorm",
                                                   colname_of_interest = "Response", 
                                                   variable_of_interest = "R",
                                                   allosome_removal = T, 
                                                   snp_removal = F, 
                                                   gap_removal = F,
                                                   outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm"))

vedodiscval_funnorm_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodiscval_noF241)$Biobank_ID,
                                                   Timepoint = pData(rgset_vedodiscval_noF241)$Timepoint,
                                                   Response = pData(rgset_vedodiscval_noF241)$Response,
                                                   Cohort = pData(rgset_vedodiscval_noF241)$Cohort,
                                                   row.names = colnames(rgset_vedodiscval_noF241))

vedodiscval_funnorm_horaizon_samples <- vedodiscval_funnorm_horaizon_samples %>%
  dplyr::mutate(Response = ifelse(Cohort == "Validation", NA, Response))

write.csv(vedodiscval_funnorm_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm", "samples_metadata.csv"))
```

```{r horaizon T1 funnorm}
vedodiscval_funnorm_horaizon_samples_T1 <- vedodiscval_funnorm_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_funnorm_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm",  "T1_sample_metadata.csv"))

data.table::fwrite(vedodiscval_funnorm_horaizon$X[rownames(vedodiscval_funnorm_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm", "T1_X.csv"))
```

```{r horaizon T2 funnorm}
vedodiscval_funnorm_horaizon_samples_T2 <- vedodiscval_funnorm_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_funnorm_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm",  "T2_sample_metadata.csv"))

data.table::fwrite(vedodiscval_funnorm_horaizon$X[rownames(vedodiscval_funnorm_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm", "T2_X.csv"))
```

## Discovery and validation (normalized together with funnorm noSNP no gap; old method)

```{r horaizon funnorm old method all}
vedodiscval_funnorm_horaizon <- preprocessHoraizon(rgset = rgset_vedodiscval_noF241,
                                                   method = "funnorm",
                                                   colname_of_interest = "Response", 
                                                   variable_of_interest = "R",
                                                   allosome_removal = T, 
                                                   snp_removal = F, 
                                                   gap_removal = F)

vedodiscval_funnorm_old_noSNP_nogap <- vedodiscval_funnorm_horaizon$X[,c(1, which(gsub("^(.+?)_.+$", "\\1", colnames(vedodiscval_funnorm_horaizon$X)) %in% features_old$cgid))]

data.table::fwrite(vedodiscval_funnorm_old_noSNP_nogap, 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh020_(old)", "All_discval_funnorm_gh020_old_X.csv"))

vedodiscval_funnorm_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodiscval_noF241)$Biobank_ID,
                                                   Timepoint = pData(rgset_vedodiscval_noF241)$Timepoint,
                                                   Response = pData(rgset_vedodiscval_noF241)$Response,
                                                   Cohort = pData(rgset_vedodiscval_noF241)$Cohort,
                                                   row.names = colnames(rgset_vedodiscval_noF241))

vedodiscval_funnorm_horaizon_samples <- vedodiscval_funnorm_horaizon_samples %>%
  dplyr::mutate(Response = ifelse(Cohort == "Validation", NA, Response))

write.csv(vedodiscval_funnorm_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh020_(old)", "All_discval_funnorm_gh020_old_sample_metadata.csv"))
```

```{r horaizon funnorm old method T1}
vedodiscval_funnorm_horaizon_samples_T1 <- vedodiscval_funnorm_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_funnorm_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh020_(old)",  "T1_discval_funnorm_gh020_old_sample_metadata.csv"))

data.table::fwrite(vedodiscval_funnorm_old_noSNP_nogap[rownames(vedodiscval_funnorm_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh020_(old)", "T1_discval_funnorm_gh020_old_X.csv"))
```

```{r horaizon funnorm old method T2}
vedodiscval_funnorm_horaizon_samples_T2 <- vedodiscval_funnorm_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_funnorm_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh020_(old)",  "T2_discval_funnorm_gh020_old_sample_metadata.csv"))

data.table::fwrite(vedodiscval_funnorm_old_noSNP_nogap[rownames(vedodiscval_funnorm_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "vedolizumab_discval_funnorm_gh020_(old)", "T2_discval_funnorm_gh020_old_X.csv"))
```

## Discovery and validation (normalized together with funnorm and combat; stricter SNP removal)

```{r preparation}
vedodiscval_gh0.1_funnorm_combat_horaizon_metadata <- pData(gmset_discval) %>%
  data.frame() %>%
  dplyr::select(Biobank_ID, SXS_POS, Timepoint, Response)

betas_discval_combat_csp_horaizon_betas <- betas_discval_combat_csp[,vedodiscval_gh0.1_funnorm_combat_horaizon_metadata$SXS_POS]

snp_cpgs <- minfi::getSnpInfo(gmset_discval)[which(minfi::getSnpInfo(gmset_discval)$CpG_maf > 0 | minfi::getSnpInfo(gmset_discval)$SBE_maf > 0),]
gaps <- minfi::gaphunter(gmset_discval, threshold = 0.1)

gapped_cpgs <- rownames(gaps$proberesults)

cpgs_for_removal <- unique(c(rownames(snp_cpgs), gapped_cpgs))

gmset_discval_combat_horaizon <- gmset_discval[!rownames(betas_discval_combat_csp_horaizon_betas) %in% cpgs_for_removal,]
betas_discval_combat_csp_horaizon_betas_horaizon <- betas_discval_combat_csp_horaizon_betas[!rownames(betas_discval_combat_csp_horaizon_betas) %in% cpgs_for_removal,]

annotation_discval_combat_horaizon <- minfi::getAnnotation(gmset_discval_combat_horaizon)
allosome_probes <- rownames(annotation_discval_combat_horaizon)[annotation_discval_combat_horaizon$chr %in% c("chrX", "chrY")]

gmset_discval_combat_horaizon <- gmset_discval_combat_horaizon[!rownames(gmset_discval_combat_horaizon) %in% allosome_probes,]
betas_discval_combat_csp_horaizon_betas_horaizon <- betas_discval_combat_csp_horaizon_betas_horaizon[!rownames(betas_discval_combat_csp_horaizon_betas_horaizon) %in% allosome_probes,]

genenames <- unlist(lapply(lapply(strsplit(minfi::getAnnotation(gmset_discval_combat_horaizon)$UCSC_RefGene_Name, ";"), unique),
                           function(genes){
                             paste(genes, collapse = ".")
                           })
)

featurenames <- unlist(lapply(lapply(strsplit(minfi::getAnnotation(gmset_discval_combat_horaizon)$UCSC_RefGene_Group, ";"), unique),
                              function(gfeats){
                                paste(gfeats, collapse = ".")
                              })
)

coordinates <- with(getAnnotation(gmset_discval_combat_horaizon), paste0(chr, ".", pos))

# CpG annotations
cpg_annotation <- paste0(rownames(gmset_discval_combat_horaizon), "_", coordinates, "_", genenames, "_", featurenames)
cpg_annotation <- gsub("(^.+?)_+$", "\\1", cpg_annotation)
```

```{r horaizon funnorm combat gh010 All}
vedodiscval_gh0.1_funnorm_combat_horaizon <- t(betas_discval_combat_csp_horaizon_betas_horaizon)
colnames(vedodiscval_gh0.1_funnorm_combat_horaizon) <- cpg_annotation
data.table::fwrite(vedodiscval_gh0.1_funnorm_combat_horaizon, "All_discval_funnorm_combat_X.csv")

vedodiscval_gh0.1_funnorm_combat_horaizon_samples <- data.frame(DonorID = pData(gmset_discval_combat_horaizon)$Biobank_ID,
                                                                Timepoint = pData(gmset_discval_combat_horaizon)$Timepoint,
                                                                Response = pData(gmset_discval_combat_horaizon)$Response,
                                                                Cohort = pData(gmset_discval_combat_horaizon)$Cohort,
                                                                row.names = colnames(gmset_discval_combat_horaizon))

write.csv(vedodiscval_gh0.1_funnorm_combat_horaizon_samples, file.path("All_discval_funnorm_combat_sample_metadata.csv"))
```

```{r horaizon funnorm combat gh010 T1}
vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T1 <- vedodiscval_gh0.1_funnorm_combat_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T1, file.path("T1_discval_funnorm_combat_sample_metadata.csv"))

vedodiscval_gh0.1_funnorm_combat_horaizon_T1 <- vedodiscval_gh0.1_funnorm_combat_horaizon[rownames(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T1),]

data.table::fwrite(vedodiscval_gh0.1_funnorm_combat_horaizon_T1, "T1_discval_funnorm_combat_X.csv")
```

```{r horaizon funnorm combat gh010 T2}
vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T2 <- vedodiscval_gh0.1_funnorm_combat_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T2, file.path("T2_discval_funnorm_combat_sample_metadata.csv"))

vedodiscval_gh0.1_funnorm_combat_horaizon_T2 <- vedodiscval_gh0.1_funnorm_combat_horaizon[rownames(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T2),]

data.table::fwrite(vedodiscval_gh0.1_funnorm_combat_horaizon_T2, "T2_discval_funnorm_combat_X.csv")
```

## Discovery and validation (normalized together with funnorm and combat with blood cell correction; stricter SNP removal)

```{r preparation}
vedodiscval_gh0.1_funnorm_combat_horaizon_metadata <- pData(gmset_discval) %>%
  data.frame() %>%
  dplyr::select(Biobank_ID, SXS_POS, Timepoint, Response)

betas_discval_combat_csp_bcd_horaizon_betas <- betas_discval_combat_csp_bcd[,vedodiscval_gh0.1_funnorm_combat_horaizon_metadata$SXS_POS]

snp_cpgs <- minfi::getSnpInfo(gmset_discval)[which(minfi::getSnpInfo(gmset_discval)$CpG_maf > 0 | minfi::getSnpInfo(gmset_discval)$SBE_maf > 0),]
gaps <- minfi::gaphunter(gmset_discval, threshold = 0.1)

gapped_cpgs <- rownames(gaps$proberesults)

cpgs_for_removal <- unique(c(rownames(snp_cpgs), gapped_cpgs))

gmset_discval_combat_horaizon <- gmset_discval[!rownames(betas_discval_combat_csp_bcd_horaizon_betas) %in% cpgs_for_removal,]
betas_discval_combat_csp_bcd_horaizon_betas_horaizon <- betas_discval_combat_csp_bcd_horaizon_betas[!rownames(betas_discval_combat_csp_bcd_horaizon_betas) %in% cpgs_for_removal,]

annotation_discval_combat_horaizon <- minfi::getAnnotation(gmset_discval_combat_horaizon)
allosome_probes <- rownames(annotation_discval_combat_horaizon)[annotation_discval_combat_horaizon$chr %in% c("chrX", "chrY")]

gmset_discval_combat_horaizon <- gmset_discval_combat_horaizon[!rownames(gmset_discval_combat_horaizon) %in% allosome_probes,]
betas_discval_combat_csp_bcd_horaizon_betas_horaizon <- betas_discval_combat_csp_bcd_horaizon_betas_horaizon[!rownames(betas_discval_combat_csp_bcd_horaizon_betas_horaizon) %in% allosome_probes,]

genenames <- unlist(lapply(lapply(strsplit(minfi::getAnnotation(gmset_discval_combat_horaizon)$UCSC_RefGene_Name, ";"), unique),
                           function(genes){
                             paste(genes, collapse = ".")
                           })
)

featurenames <- unlist(lapply(lapply(strsplit(minfi::getAnnotation(gmset_discval_combat_horaizon)$UCSC_RefGene_Group, ";"), unique),
                              function(gfeats){
                                paste(gfeats, collapse = ".")
                              })
)

coordinates <- with(getAnnotation(gmset_discval_combat_horaizon), paste0(chr, ".", pos))

# CpG annotations
cpg_annotation <- paste0(rownames(gmset_discval_combat_horaizon), "_", coordinates, "_", genenames, "_", featurenames)
cpg_annotation <- gsub("(^.+?)_+$", "\\1", cpg_annotation)
```

```{r horaizon funnorm combat gh010 All}
vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon <- t(betas_discval_combat_csp_bcd_horaizon_betas_horaizon)
colnames(vedodiscval_gh0.1_funnorm_combat_horaizon) <- cpg_annotation
data.table::fwrite(vedodiscval_gh0.1_funnorm_combat_horaizon, "All_discval_funnorm_combat_bcd_X.csv")

vedodiscval_gh0.1_funnorm_combat_horaizon_samples <- data.frame(DonorID = pData(gmset_discval_combat_horaizon)$Biobank_ID,
                                                                Timepoint = pData(gmset_discval_combat_horaizon)$Timepoint,
                                                                Response = pData(gmset_discval_combat_horaizon)$Response,
                                                                Cohort = pData(gmset_discval_combat_horaizon)$Cohort,
                                                                row.names = colnames(gmset_discval_combat_horaizon))

write.csv(vedodiscval_gh0.1_funnorm_combat_horaizon_samples, file.path("All_discval_funnorm_combat_sample_metadata.csv"))
```

```{r horaizon funnorm combat gh010 T1}
vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T1 <- vedodiscval_gh0.1_funnorm_combat_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T1, file.path("T1_discval_funnorm_combat_sample_metadata.csv"))

vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon_T1 <- vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon[rownames(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T1),]

data.table::fwrite(vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon_T1, "T1_discval_funnorm_combat_bcd_X.csv")
```

```{r horaizon funnorm combat gh010 T2}
vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T2 <- vedodiscval_gh0.1_funnorm_combat_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T2, file.path("T2_discval_funnorm_combat_sample_metadata.csv"))

vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon_T2 <- vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon[rownames(vedodiscval_gh0.1_funnorm_combat_horaizon_samples_T2),]

data.table::fwrite(vedodiscval_gh0.1_funnorm_combat_csp_bcd_horaizon_T2, "T2_discval_funnorm_combat_bcd_X.csv")
```

## Discovery

```{r horaizon discovery selection}
rgset_vedodisc_noF241 <- rgset_vedodiscval_noF241[,pData(rgset_vedodiscval_noF241)$Cohort == "Discovery"]
```

### gaphunter 0.2

```{r horaizon discovery}
vedodisc_gh0.2_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedodisc_noF241,
                                                         colname_of_interest = "Response", 
                                                         variable_of_interest = "R",
                                                         allosome_removal = T, 
                                                         snp_removal = T, 
                                                         min_maf = 0,
                                                         gap_removal = T,
                                                         gap_threshold = 0.2,
                                                         outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discovery_gh020"))

vedodisc_gh0.2_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodisc_noF241)$Biobank_ID,
                                              Timepoint = pData(rgset_vedodisc_noF241)$Timepoint,
                                              Response = pData(rgset_vedodisc_noF241)$Response,
                                              Cohort = pData(rgset_vedodisc_noF241)$Cohort,
                                              row.names = colnames(rgset_vedodisc_noF241))

write.csv(vedodisc_gh0.2_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "All_disc_gh020_sample_metadata.csv"))
```

```{r horaizon discovery T1}
vedodisc_gh0.2_horaizon_samples_T1 <- vedodisc_gh0.2_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodisc_gh0.2_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T1_disc_gh020_sample_metadata.csv"))

data.table::fwrite(vedodisc_gh0.2_horaizon$X[rownames(vedodisc_gh0.2_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T1_disc_gh020_sample_X.csv"))
```

```{r horaizon discovery T2}
vedodisc_gh0.2_horaizon_samples_T2 <- vedodisc_gh0.2_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodisc_gh0.2_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T2_disc_gh020_sample_metadata.csv"))

data.table::fwrite(vedodisc_gh0.2_horaizon$X[rownames(vedodisc_gh0.2_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T2_disc_gh020_sample_X.csv"))
```

### gaphunter 0.1

```{r horaizon discovery}
vedodisc_gh0.1_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedodisc_noF241,
                                                         colname_of_interest = "Response", 
                                                         variable_of_interest = "R",
                                                         allosome_removal = T, 
                                                         snp_removal = T, 
                                                         min_maf = 0,
                                                         gap_removal = T,
                                                         gap_threshold = 0.1,
                                                         outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discovery_gh010"))

vedodisc_gh0.1_horaizon_samples <- data.frame(DonorID = pData(rgset_vedodisc_noF241)$Biobank_ID,
                                              Timepoint = pData(rgset_vedodisc_noF241)$Timepoint,
                                              Response = pData(rgset_vedodisc_noF241)$Response,
                                              Cohort = pData(rgset_vedodisc_noF241)$Cohort,
                                              row.names = colnames(rgset_vedodisc_noF241))

write.csv(vedodisc_gh0.1_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "All_disc_gh010_sample_metadata.csv"))
```

```{r horaizon discovery T1}
vedodisc_gh0.1_horaizon_samples_T1 <- vedodisc_gh0.1_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedodisc_gh0.1_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T1_disc_gh010_sample_metadata.csv"))

data.table::fwrite(vedodisc_gh0.1_horaizon$X[rownames(vedodisc_gh0.1_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T1_disc_gh010_sample_X.csv"))
```

```{r horaizon discovery T2}
vedodisc_gh0.1_horaizon_samples_T2 <- vedodisc_gh0.1_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedodisc_gh0.1_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T2_disc_gh010_sample_metadata.csv"))

data.table::fwrite(vedodisc_gh0.1_horaizon$X[rownames(vedodisc_gh0.1_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T2_disc_gh010_sample_X.csv"))
```

## Validation

```{r horaizon validation selection}
rgset_vedoval_noF241 <- rgset_vedodiscval_noF241[,pData(rgset_vedodiscval_noF241)$Cohort == "Validation"]
```

```{r horaizon validation}
vedoval_gh0.1_horaizon <- rkgreslib::preprocessHoraizon(rgset = rgset_vedoval_noF241,
                                                        colname_of_interest = "Response", 
                                                        variable_of_interest = "R",
                                                        allosome_removal = T, 
                                                        snp_removal = T, 
                                                        min_maf = 0,
                                                        gap_removal = T,
                                                        gap_threshold = 0.1,
                                                        outpath = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "PRJ0000008_CDMEDRESP_vedolizumab_betas_noob_discovery_gh010"))

vedoval_gh0.1_horaizon_samples <- data.frame(DonorID = pData(rgset_vedoval_noF241)$Biobank_ID,
                                             Timepoint = pData(rgset_vedoval_noF241)$Timepoint,
                                             Response = NA,
                                             Cohort = pData(rgset_vedoval_noF241)$Cohort,
                                             row.names = colnames(rgset_vedoval_noF241))

write.csv(vedoval_gh0.1_horaizon_samples, file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "All_val_gh010_sample_metadata.csv"))
```

```{r horaizon validation T1}
vedoval_gh0.1_horaizon_samples_T1 <- vedoval_gh0.1_horaizon_samples %>%
  dplyr::filter(Timepoint == "T1")

write.csv(vedoval_gh0.1_horaizon_samples_T1, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T1_val_gh010_sample_metadata.csv"))

data.table::fwrite(vedoval_gh0.1_horaizon$X[rownames(vedoval_gh0.1_horaizon_samples_T1),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T1_val_gh010_sample_X.csv"))
```

```{r horaizon validation T2}
vedoval_gh0.1_horaizon_samples_T2 <- vedoval_gh0.1_horaizon_samples %>%
  dplyr::filter(Timepoint == "T2")

write.csv(vedoval_gh0.1_horaizon_samples_T2, file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T2_val_gh010_sample_metadata.csv"))

data.table::fwrite(vedoval_gh0.1_horaizon$X[rownames(vedoval_gh0.1_horaizon_samples_T2),], 
                   file = file.path("output", "DM_analyses", "03_biological_validation", "aumc1", "211005", "horaizon", "T2_val_gh010_sample_X.csv"))
```

#Figures

## Report October 2019

```{r 1019 figures setup}
figDir <- file.path("docs", "reports", "vedolizumab_discovery", "figures")
dir.create(figDir)
```

### Figure 1

```{r 1019 fig1}
fig1Dir <- file.path(figDir, "fig1")
dir.create(fig1Dir)
```

```{r 1019 fig1a}
fig1a <- stabreplor_plot
```

```{r 1019 fig1b}
fig1b <- RvNR_vedo_hm
```

```{r 1019 fig1b}
fig1c <- ggplot(RvNR_vedo_resp_time_df, aes(x = RvNR_beta, y = T2vT1_beta)) +
  geom_hline(yintercept = 0, col = "grey") +
  geom_vline(xintercept = 0, col = "grey") +
  geom_point() +
  xlim(-0.5,0.5) +
  ylim(-0.5,0.5) +
  labs(title = "Difference in % methylation",
       subtitle = "Response-associated DMPs") +
  xlab("R vs NR") +
  ylab("T2 vs T1") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

```{r 1019 fig1c}
fig1d <- RvNR_vedo_gc_plot_tr
```

```{r 1019 fig1d}
fig1e <- RvNR_vedo_sig_cpgs_gv_plot
```

```{r 1019 fig1}
fig1ab <- plot_grid(fig1a, fig1b, labels = c("a", "b"), nrow = 1, ncol = 2, rel_widths = c(0.5, 1))
fig1cde <- plot_grid(fig1c, fig1d, fig1e, labels = c("c", "d", "e"), nrow = 3, ncol = 1)

fig1 <- plot_grid(fig1ab, fig1cde, nrow = 1, ncol = 2, rel_widths = c(1, 0.3))

Cairo(width = 4000, height = 3000, type = "pdf", file = file.path(fig1Dir, "fig1.pdf"), bg = "white", units = "px", dpi = 120)
print(fig1)
dev.off()
```

### Fig 2

```{r 1019 fig2 setup}
fig2Dir <- file.path(figDir, "fig2")
dir.create(fig2Dir)
```

```{r 1019 fig2a}
# fig2atop <- dmr_plot(dmr_gr = range(gmset_vedo_anno_gr[grep("HLA-C", gmset_vedo_anno_gr$UCSC_RefGene_Name)], ignore.strand = T), 
#                      meth_data = getBeta(gmset_vedo_filtered_t1t2), 
#                      anno_gr = gmset_vedo_anno_gr, 
#                      meth_groups = pData(gmset_vedo_filtered_t1t2)$Response, 
#                      title = "HLA-C") + 
#   ylab("% Methylation") +
#   theme(legend.pos = "none")

fig2atop <- dmr_effect_plot(region_of_interest = medstablor_df_gene_mdmp_gr[1,], tophits = RvNR_vedo_top_gr, significance = F, smooth = F, ylim = c(-0.5, 0.5))

fig2abot_df <- medstablor_df[grep("HLA-C", medstablor_df$UCSC_RefGene_Name),]
fig2abot_df <- fig2abot_df[order(fig2abot_df$start),]

fig2abot1 <- cpg_strip_plot(cpg_num = rownames(fig2abot_df)[1], betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$Response, type = "SE") + 
  labs(title = rownames(fig2abot_df)[1],
       subtitle = paste0(fig2abot_df$seqnames[1], ":", fig2abot_df$start[1])) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
fig2abot2 <- cpg_strip_plot(cpg_num = rownames(fig2abot_df)[2], betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$Response, type = "SE") + 
  labs(title = rownames(fig2abot_df)[2],
       subtitle = paste0(fig2abot_df$seqnames[2], ":", fig2abot_df$start[2])) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
fig2abot3 <- cpg_strip_plot(cpg_num = rownames(fig2abot_df)[3], betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$Response, type = "SE") + 
  labs(title = rownames(fig2abot_df)[3],
       subtitle = paste0(fig2abot_df$seqnames[3], ":", fig2abot_df$start[3])) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
```

```{r 1019 fig2b}
# fig2btop <- dmr_plot(dmr_gr = range(gmset_vedo_anno_gr[grep("FAM163A", gmset_vedo_anno_gr$UCSC_RefGene_Name)], ignore.strand = T), 
#                      meth_data = getBeta(gmset_vedo_filtered_t1t2), 
#                      anno_gr = gmset_vedo_anno_gr, 
#                      meth_groups = pData(gmset_vedo_filtered_t1t2)$Response, 
#                      title = "FAM163A") + 
#   ylab("% Methylation") +
#   theme(legend.pos = "none")

fig2btop <- dmr_effect_plot(region_of_interest = medstablor_df_gene_mdmp_gr[2,], tophits = RvNR_vedo_top_gr, significance = F, smooth = F)

fig2bbot_df <- medstablor_df[grep("FAM163A", medstablor_df$UCSC_RefGene_Name),]
fig2bbot_df <- fig2bbot_df[order(fig2bbot_df$start),]

fig2bbot1 <- cpg_strip_plot(cpg_num = rownames(fig2bbot_df)[1], betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$Response, type = "SE") + 
  labs(title = rownames(fig2bbot_df)[1],
       subtitle = paste0(fig2bbot_df$seqnames[1], ":", fig2bbot_df$start[1])) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())

fig2bbot2 <- cpg_strip_plot(cpg_num = rownames(fig2bbot_df)[2], betas = getBeta(gmset_vedo_filtered_t1t2), factor_interest = pData(gmset_vedo_filtered_t1t2)$Response, type = "SE") + 
  labs(title = rownames(fig2bbot_df)[2],
       subtitle = paste0(fig2bbot_df$seqnames[2], ":", fig2bbot_df$start[2])) + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.title.y = element_blank())
```

```{r 1019 fig2c}
fig2c <- dmr_plot(dmr_gr = RvNR_dmr_vedo_gr[1], 
                  meth_data = getBeta(gmset_vedo_filtered_t1t2), 
                  anno_gr = gmset_vedo_anno_gr, 
                  meth_groups = pData(gmset_vedo_filtered_t1t2)$Response, 
                  highlight = c(start(RvNR_dmr_vedo_gr[1]), end(RvNR_dmr_vedo_gr[1]))) + 
  ylab("% Methylation")
```

```{r 1019 fig2}
Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(fig2Dir, "fig2atop.pdf"), bg = "white", units = "px", dpi = 120)
print(fig2atop)
dev.off()

Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(fig2Dir, "fig2btop.pdf"), bg = "white", units = "px", dpi = 120)
print(fig2btop)
dev.off()

fig2abbot <- ggarrange(fig2abot1, fig2abot2, fig2abot3, fig2bbot1, fig2bbot2, nrow = 1, ncol = 5, common.legend = T, legend.pos = "bottom") 

Cairo(width = 2000, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2abbot.pdf"), bg = "white", units = "px", dpi = 120)
print(fig2abbot$`1`)
dev.off()

Cairo(width = 2000, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2c.pdf"), bg = "white", units = "px", dpi = 120)
print(fig2c)
dev.off()
```

### Fig 3

```{r 1019 fig3 setup}
fig3Dir <- file.path(figDir, "fig3")
dir.create(fig3Dir)
```

```{r 1019 fig3a}
fig3a <- overlap_pilot_pilotsig_plot
```

```{r 1019 fig3b}
fig3b <- overlap_pilot_discoverysig_plot
```

```{r 1019 fig3c}
fig3c <- mval_vedo_pilotsig_discovery_svd_plot
```

```{r 1019 fig3d}
fig3d <- mval_vedo_discoverysig_pilot_svd_plot
```

```{r 1019 fig3e}
fig3e <- mval_vedo_pilotsig_discovery_ol_svd_plot
```

```{r 1019 fig3}
fig3ab <- ggarrange(fig3a + theme(legend.pos = "bottom"), fig3b + theme(legend.pos = "bottom"), nrow = 1, ncol = 2, labels = c("a", "b"))
fig3cd <- ggarrange(fig3c, fig3d, nrow = 1, ncol = 2, common.legend = T, legend = "bottom", labels = c("c", "d"))
fig3e <- ggarrange(fig3e + theme(legend.pos = "none"), nrow = 1, ncol = 2, labels = c("e", ""))

fig3 <- ggarrange(fig3ab, fig3cd, fig3e, nrow = 3, ncol = 1)

Cairo(width = 2000, height = 3000, type = "pdf", file = file.path(fig3Dir, "fig3.pdf"), bg = "white", units = "px", dpi = 120)
print(fig3)
dev.off()
```

### Fig S1

```{r 1019 figS1 setup}
figS1Dir <- file.path(figDir, "figS1")
dir.create(figS1Dir)
```

```{r 1019 figS1}
betas_vedo_t1t2_respdmps <- getBeta(gmset_vedo_filtered_t1t2)[rownames(medstablor_df),]
betas_vedo_t1t2_respdmps_melt <- melt(betas_vedo_t1t2_respdmps, varnames = c("CpG_ID", "SXS_POS"), value.name = "Beta")
betas_vedo_t1t2_respdmps_melt <- cbind(betas_vedo_t1t2_respdmps_melt, pData(gmset_vedo_filtered_t1t2)[betas_vedo_t1t2_respdmps_melt$SXS_POS, c("Timepoint", "Response", "Biobank_ID")])

figS1 <- ggplot(data.frame(betas_vedo_t1t2_respdmps_melt), aes(x = Timepoint, y = Beta, col = Response)) +
  geom_point(aes(group = Biobank_ID), alpha = 0.25, show.legend = F) +
  geom_line(aes(group = Biobank_ID, alpha = 0.25), linetype = "dotted", show.legend = F) +
  geom_smooth(method = "lm", aes(group = Response), se = F) +
  facet_wrap(~CpG_ID, nrow = 10, ncol = 4) +
  ylab("Difference % methylation") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

Cairo(width = 2000, height = 2500, type = "pdf", file = file.path(figS1Dir, "figS1.pdf"), bg = "white", units = "px", dpi = 120)
print(figS1)
dev.off()
```

### Fig S2

```{r 1019 figS2 setup}
figS2Dir <- file.path(figDir, "figS2")
dir.create(figS2Dir)
```

```{r 1019 figS2}
figS2 <- ggplot(data.frame(betas_vedo_t1t2t3_respdmps_melt), aes(x = Timepoint, y = Beta, col = Response)) +
  geom_point(aes(group = Biobank_ID), alpha = 0.25, show.legend = F) +
  geom_line(aes(group = Biobank_ID, alpha = 0.25), linetype = "dotted", show.legend = F) +
  geom_smooth(method = "lm", aes(group = Response), se = F) +
  facet_wrap(~CpG_ID, nrow = 10, ncol = 4) +
  ylab("Difference % methylation") +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

Cairo(width = 2000, height = 2500, type = "pdf", file = file.path(figS2Dir, "figS2.pdf"), bg = "white", units = "px", dpi = 120)
print(figS2)
dev.off()
```

### Fig S3

```{r 1019 figS3 setup}
figS3Dir <- file.path(figDir, "figS3")
dir.create(figS3Dir)
```

```{r 1019 figS3a}
figS3a <- mval_vedo_pilotsig_pilot_svd_plot
```

```{r 1019 figS3b}
figS3b <- mval_vedo_discoverysig_discovery_svd_plot
```

```{r 1019 figS3}
figS3 <- ggarrange(figS3a, figS3b, nrow = 1, ncol = 2, label = c("a", "b"), legend = "bottom", common.legend = T)

Cairo(width = 2000, height = 1000, type = "pdf", file = file.path(figS3Dir, "figS3.pdf"), bg = "white", units = "px", dpi = 120)
print(figS3$`1`)
dev.off()
```

### Visualization predictor genes

```{r visualization predictor CpGs}
cpg_strat_plot <- function(cpg, betas, facet_factor, x_factor, title, subtitle){
  beta_melt <- reshape2::melt(betas[cpg,])
  beta_df <- cbind(beta_melt, 
                   facet_factor, 
                   x_factor)
  
  ggplot(data = beta_df, aes(x = x_factor, y = value)) +
    stat_summary(fun.data = mean_se, geom = "crossbar", alpha = 0.5) +
    geom_jitter() +
    theme_bw() +
    facet_wrap(~facet_factor) +
    xlab("") +
    labs(title = title, subtitle = subtitle) +
    ylab("% Methylation") +
    ylim(0,1) +
    scale_shape_manual(values=(1:nlevels(beta_df$Cohort))%%10) +
    theme(axis.text = element_text(size = 12),
          axis.title = element_text(size = 14, face = "bold"),
          axis.title.x = element_blank(),
          legend.title = element_text(size = 14, face = "bold"),
          legend.text = element_text(size = 12),
          plot.title = element_text(face = "bold"),
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          strip.text.x = element_text(size = 14))
}

visDir <- file.path(figDir, "predictor CpGs")
dir.create(visDir)

pdfvisDir <- file.path(visDir, "pdf")
dir.create(pdfvisDir)

pngvisDir <- file.path(visDir, "png")
dir.create(pngvisDir)

for(i in 1:nrow(medstablor_df)){
  Cairo(width = 500, height = 500, type = "png", file = file.path(pngvisDir, paste0(medstablor_df[i,]$Name, "_vedo.png")) , bg = "white", units = "px", dpi = 120)
  print(cpg_strat_plot(cpg = medstablor_df[i,]$Name, betas = getBeta(gmset_vedo_filtered), facet_factor = pData(gmset_vedo_filtered)$Response, x_factor = paste0("T", pData(gmset_vedo_filtered)$Timepoint), title = paste0("Vedo: ", medstablor_df[i,]$Name), subtitle = paste0(medstablor_df[i,]$seqnames, ":", medstablor_df[i,]$start)))
  dev.off()
  
  Cairo(width = 1000, height = 750, type = "pdf", file = file.path(pdfvisDir, paste0(medstablor_df[i,]$Name, "_vedo.pdf")) , bg = "white", units = "px", dpi = 120)
  print(cpg_strat_plot(cpg = medstablor_df[i,]$Name, betas = getBeta(gmset_vedo_filtered), facet_factor = pData(gmset_vedo_filtered)$Response, x_factor = paste0("T", pData(gmset_vedo_filtered)$Timepoint), title = paste0("Vedo: ", medstablor_df[i,]$Name), subtitle = paste0(medstablor_df[i,]$seqnames, ":", medstablor_df[i,]$start)))
  dev.off()
}

```

## Abstract ECCO 2020

```{r}
eccoDir <- file.path("docs", "ecco")
dir.create(eccoDir)

eccofigDir <- file.path("docs", "ecco", "figures")
dir.create(eccofigDir)
```


```{r}
replor_df_abs <- replor_df
rownames(replor_df_abs) <- c("(Intercept)", paste0("CpG", 1:(nrow(replor_df_abs)-1)))
replor_df_abs_med <- rowMedians(as.matrix(replor_df_abs))
names(replor_df_abs_med) <- rownames(replor_df_abs)
replor_df_abs_med <- replor_df_abs_med[replor_df_abs_med != 0]

replor_melt_abs <- melt(cbind(feature = rownames(replor_df_abs), replor_df_abs))
replor_melt_abs <- replor_melt_abs[replor_melt_abs$feature != "(Intercept)",]
replor_melt_abs$Stability <- "Unstable"
replor_melt_abs$Stability[replor_melt_abs$feature %in% names(replor_df_abs_med)] <- "Stable"
replor_melt_abs <- replor_melt_abs[replor_melt_abs$Stability == "Stable",]
replor_melt_abs$feature <- factor(replor_melt_abs$feature, levels = names(sort(replor_df_abs_med[-1])))
replor_melt_abs$Direction <- "Positive"
replor_melt_abs$Direction[replor_melt_abs$feature %in% names(replor_df_abs_med[-1])[which(replor_df_abs_med[-1]<0)]] <- "Negative"

fig_abs_l <- ggplot(replor_melt_abs, aes(x = feature, y = value, col = Direction)) +
  geom_jitter(alpha = 0.10) +
  geom_boxplot(alpha = 0.50) +
  theme_bw() +
  ylab("log(OR)") +
  labs(title = "100 repetitions of 10-fold cross-validations",
       subtitle = "Median-stabilized") +
  theme(axis.title.y = element_blank()) +
  coord_flip()

fig_abs_r <- plot_pilot_roc

fig_abs <- ggarrange(fig_abs_l, fig_abs_r, nrow = 1, ncol = 2)

Cairo(width = 2200, height = 1100, type = "pdf", file = file.path(eccofigDir, "fig_abstract.pdf"), bg = "white", units = "px", dpi = 120)
fig_abs
dev.off()
```

# Save the data

```{r raw data}
require(data.table)

rgset_raw <- mapToGenome(preprocessRaw(rgset_vedo))

write.csv2(minfi::getAnnotation(rgset_raw), file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_annotation.csv"))
write.csv2(minfi::getBeta(rgset_raw), file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_semi_raw.csv"))

rgset_raw_nosex <- rgset_raw[-which(minfi::getAnnotation(rgset_raw)$chr %in% c("chrX", "chrY")),]
rgset_raw_nosex_noprom <- rgset_raw_nosex[!names(rgset_raw_nosex) %in% pmprobes,]
write.csv2(minfi::getBeta(rgset_raw_nosex_noprom), file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_semi_raw_filtered.csv"))

excluded_probes <- rbind(data.frame(minfi::getAnnotation(rgset_raw)[which(minfi::getAnnotation(rgset_raw)$chr %in% c("chrX", "chrY")), c("chr", "pos", "Name")], excluded = "Allosomal"),
                         data.frame(minfi::getAnnotation(rgset_raw)[rownames(rgset_raw) %in% pmprobes, c("chr", "pos", "Name")], excluded = "Promiscuous"))

write.csv(excluded_probes, file.path(outputDir, "processed_data", "excluded_probes.csv"))
```

```{r raw data t1}
rgset_raw_nosex_noprom_t1 <- rgset_raw_nosex_noprom[,pData(rgset_raw_nosex_noprom)$Timepoint == 1]
beta_vedo_raw_filtered_t1 <- data.frame(Response = pData(rgset_raw_nosex_noprom_t1)[,c("Response")],
                                        t(minfi::getBeta(rgset_raw_nosex_noprom_t1)),
                                        row.names = pData(rgset_raw_nosex_noprom_t1)$Biobank_ID)

fwrite(beta_vedo_raw_filtered_t1, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_semi_raw_filtered_t1_prepared.csv"))
```

```{r raw data t2}
rgset_raw_nosex_noprom_t2 <- rgset_raw_nosex_noprom[,pData(rgset_raw_nosex_noprom)$Timepoint == 2]
beta_vedo_raw_filtered_t2 <- data.frame(Response = pData(rgset_raw_nosex_noprom_t2)$Response,
                                        t(minfi::getBeta(rgset_raw_nosex_noprom_t2)),
                                        row.names = pData(rgset_raw_nosex_noprom_t2)$Biobank_ID)

fwrite(beta_vedo_raw_filtered_t2, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_semi_raw_filtered_t2_prepared.csv"))
```

```{r raw data across t1 and t2}
beta_vedo_raw_filtered_dt1t2 <- minfi::getBeta(rgset_raw_nosex_noprom_t2)-minfi::getBeta(rgset_raw_nosex_noprom_t1)
beta_vedo_raw_filtered_dt1t2_anno <- data.frame(Response = pData(rgset_raw_nosex_noprom_t2)$Response,
                                                t(beta_vedo_raw_filtered_dt1t2),
                                                row.names = pData(rgset_raw_nosex_noprom_t2)$Biobank_ID)

fwrite(beta_vedo_raw_filtered_dt1t2_anno, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_semi_raw_filtered_dt1t2_prepared.csv"))
```

```{r background corrected and normalized data t1}
gmset_vedo_filtered_t1 <- gmset_vedo_filtered[,pData(gmset_vedo_filtered)$Timepoint == 1]

snp_cpgs <- getSnpInfo(gmset_vedo_filtered_t1)
cpgs_keep <- rownames(snp_cpgs)[-which(snp_cpgs$Probe_maf>0.01 | snp_cpgs$CpG_maf>0.01 | snp_cpgs$SBE_maf>0.01)]

gmset_vedo_filtered_t1 <- gmset_vedo_filtered_t1[cpgs_keep,]

cpg_annotation_filtered_t1 <- paste0(rownames(gmset_vedo_filtered_t1), "_", unlist(lapply(lapply(strsplit(getAnnotation(gmset_vedo_filtered_t1)$UCSC_RefGene_Name, ";"), unique), function(genes){paste(genes, collapse = "_")})))
cpg_annotation_filtered_t1 <- gsub("(^.+)_$", "\\1", cpg_annotation_filtered_t1)

beta_vedo_filtered_t1 <- data.frame(Response = pData(gmset_vedo_filtered_t1)[,c("Response")],
                                    t(minfi::getBeta(gmset_vedo_filtered_t1)),
                                    row.names = pData(gmset_vedo_filtered_t1)$Biobank_ID)
colnames(beta_vedo_filtered_t1)[2:ncol(beta_vedo_filtered_t1)] <- cpg_annotation_filtered_t1

fwrite(beta_vedo_filtered_t1, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_funnorm_filtered_nosnp_t1_prepared.csv"))

fwrite(beta_vedo_filtered_t1, "/home/ayliyim/surfdrive/Shared/191211/PRJ0000008_CDMEDRESP_vedolizumab_funnorm_filtered_nosnp_t1_prepared.csv")
```

```{r background corrected and normalized data t2}
gmset_vedo_filtered_t2 <- gmset_vedo_filtered[,pData(gmset_vedo_filtered)$Timepoint == 2]
beta_vedo_filtered_t2 <- data.frame(Response = pData(gmset_vedo_filtered_t2)[,c("Response")],
                                    t(minfi::getBeta(gmset_vedo_filtered_t2)),
                                    row.names = pData(gmset_vedo_filtered_t2)$Biobank_ID)

fwrite(beta_vedo_filtered_t2, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_funnorm_filtered_t2_prepared.csv"))
```

```{r background corrected and normalized data across t1 and t2}
beta_vedo_filtered_dt1t2 <- minfi::getBeta(gmset_vedo_filtered_t2)-minfi::getBeta(gmset_vedo_filtered_t1)

beta_vedo_filtered_dt1t2_anno <- data.frame(Response = pData(gmset_vedo_filtered_t2)[,c("Response")], 
                                            t(beta_vedo_filtered_dt1t2),
                                            row.names = pData(gmset_vedo_filtered_t2)$Biobank_ID)
colnames(beta_vedo_filtered_dt1t2_anno)[2:ncol(beta_vedo_filtered_dt1t2_anno)] <- cpg_annotation_filtered_t1

fwrite(beta_vedo_filtered_dt1t2_anno, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_funnorm_filtered_dt2t1_prepared.csv"))
```

### Background correction only

```{r background corrected}
cpg_annotation_vedo <- paste0(rownames(gmset_noob_vedo), "_", unlist(lapply(lapply(strsplit(getAnnotation(gmset_noob_vedo)$UCSC_RefGene_Name, ";"), unique), function(genes){paste(genes, collapse = "_")})))
cpg_annotation_vedo <- gsub("(^.+)_$", "\\1", cpg_annotation_vedo)

beta_vedo_noob <- data.frame(Response = pData(gmset_noob_vedo)[,c("Response")],
                             t(minfi::getBeta(gmset_noob_vedo)),
                             row.names = paste0(pData(gmset_noob_vedo)$Biobank_ID, "_", pData(gmset_noob_vedo)$Timepoint))
colnames(beta_vedo_noob)[2:ncol(beta_vedo_noob)] <- cpg_annotation_vedo

beta_vedo_noob_t1 <- beta_vedo_noob[grep("_1", rownames(beta_vedo_noob)),]
beta_vedo_noob_t2 <- beta_vedo_noob[grep("_2", rownames(beta_vedo_noob)),]
beta_vedo_noob_t3 <- beta_vedo_noob[grep("_3", rownames(beta_vedo_noob)),]

data.table::fwrite(beta_vedo_noob_t1, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_t1.csv"))
data.table::fwrite(beta_vedo_noob_t2, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_t2.csv"))
data.table::fwrite(beta_vedo_noob_t3, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_t3.csv"))
```

```{r background corrected without snps}
cpg_annotation_vedo_nosnp <- paste0(rownames(gmset_noob_vedo_nosnp), "_", unlist(lapply(lapply(strsplit(getAnnotation(gmset_noob_vedo_nosnp)$UCSC_RefGene_Name, ";"), unique), function(genes){paste(genes, collapse = "_")})))
cpg_annotation_vedo_nosnp <- gsub("(^.+)_$", "\\1", cpg_annotation_vedo_nosnp)

beta_vedo_noob_nosnp <- data.frame(Response = pData(gmset_noob_vedo_nosnp)[,c("Response")],
                                   t(minfi::getBeta(gmset_noob_vedo_nosnp)),
                                   row.names = paste0(pData(gmset_noob_vedo_nosnp)$Biobank_ID, "_", pData(gmset_noob_vedo_nosnp)$Timepoint))
colnames(beta_vedo_noob_nosnp)[2:ncol(beta_vedo_noob_nosnp)] <- cpg_annotation_vedo_nosnp

beta_vedo_noob_nosnp_t1 <- beta_vedo_noob_nosnp[grep("_1", rownames(beta_vedo_noob_nosnp)),]
beta_vedo_noob_nosnp_t2 <- beta_vedo_noob_nosnp[grep("_2", rownames(beta_vedo_noob_nosnp)),]
beta_vedo_noob_nosnp_t3 <- beta_vedo_noob_nosnp[grep("_3", rownames(beta_vedo_noob_nosnp)),]

data.table::fwrite(beta_vedo_noob_nosnp_t1, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_nosnp_t1.csv"))
data.table::fwrite(beta_vedo_noob_nosnp_t2, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_nosnp_t2.csv"))
data.table::fwrite(beta_vedo_noob_nosnp_t3, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_nosnp_t3.csv"))
```

```{r background corrected without snps without gaps}
cpg_annotation_vedo_nosnp_nogap <- paste0(rownames(gmset_noob_vedo_nosnp_nogap), "_", unlist(lapply(lapply(strsplit(getAnnotation(gmset_noob_vedo_nosnp_nogap)$UCSC_RefGene_Name, ";"), unique), function(genes){paste(genes, collapse = "_")})))
cpg_annotation_vedo_nosnp_nogap <- gsub("(^.+)_$", "\\1", cpg_annotation_vedo_nosnp_nogap)

beta_vedo_noob_nosnp_nogap <- data.frame(Response = pData(gmset_noob_vedo_nosnp_nogap)[,c("Response")],
                                         t(minfi::getBeta(gmset_noob_vedo_nosnp_nogap)),
                                         row.names = paste0(pData(gmset_noob_vedo_nosnp_nogap)$Biobank_ID, "_", pData(gmset_noob_vedo_nosnp_nogap)$Timepoint))
colnames(beta_vedo_noob_nosnp_nogap)[2:ncol(beta_vedo_noob_nosnp_nogap)] <- cpg_annotation_vedo_nosnp_nogap

beta_vedo_noob_nosnp_nogap_t1 <- beta_vedo_noob_nosnp_nogap[grep("_1", rownames(beta_vedo_noob_nosnp_nogap)),]
beta_vedo_noob_nosnp_nogap_t2 <- beta_vedo_noob_nosnp_nogap[grep("_2", rownames(beta_vedo_noob_nosnp_nogap)),]
beta_vedo_noob_nosnp_nogap_t3 <- beta_vedo_noob_nosnp_nogap[grep("_3", rownames(beta_vedo_noob_nosnp_nogap)),]

data.table::fwrite(beta_vedo_noob_nosnp_nogap_t1, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_nosnp_nogap_t1.csv"))
data.table::fwrite(beta_vedo_noob_nosnp_nogap_t2, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_nosnp_nogap_t2.csv"))
data.table::fwrite(beta_vedo_noob_nosnp_nogap_t3, file.path(outputDir, "processed_data", "PRJ0000008_CDMEDRESP_vedolizumab_noob_nosnp_nogap_t3.csv"))
```


```{r save vedo gmset}
saveRDS(gmset_vedo_filtered, file.path(rdsDir, "gmset_vedo_filtered.Rds"))
```

# T1 T2 T3 R v NR

```{r}
gmset_vedo_filtered_t1t2t3

table(pData(gmset_vedo_filtered_t1t2t3)$Biobank_ID)
```

```{r vedo DM analyses preparation}
design_rvnr_t1t2t3 <- model.matrix(~0 + resptime + Sex + Age_baseline + Smoker, data = pData(gmset_vedo_filtered_t1t2t3))
colnames(design_rvnr_t1t2t3) <- c("T1_NR", "T2_NR", "T3_NR", "T1_R", "T2_R", "T3_R", "Male", "Age", "Smoker")

mvals_lmfit_vedo_t1t2t3 <- lmFit(getM(gmset_vedo_filtered_t1t2t3), design = design_rvnr_t1t2t3)
betas_lmfit_vedo_t1t2t3 <- lmFit(getBeta(gmset_vedo_filtered_t1t2t3), design = design_rvnr_t1t2t3)

contrast_mat_vedot1t2t3 <- makeContrasts(
  RvNRT1 = T1_R-T1_NR,
  RvNRT2 = T2_R-T2_NR,
  RvNRT3 = T3_R-T3_NR,
  levels = design_rvnr_t1t2t3
)

mvals_contrastfit_vedo_t1t2t3 <- contrasts.fit(mvals_lmfit_vedo_t1t2t3, contrast_mat_vedot1t2t3)
mvals_ebayes_vedo_t1t2t3 <- eBayes(mvals_contrastfit_vedo_t1t2t3)

betas_contrastfit_vedo_t1t2t3 <- contrasts.fit(betas_lmfit_vedo_t1t2t3, contrast_mat_vedot1t2t3)
betas_ebayes_vedo_t1t2t3 <- eBayes(betas_contrastfit_vedo_t1t2t3)
```

## T1_RvNR

```{r vedo DM analyses RvNR for T1 subsets}
RvNR_T1_vedot1t2t3_top <- topTable(mvals_ebayes_vedo_t1t2t3, coef = "RvNRT1", number = "Inf", adjust.method = "BH")
RvNR_T1_vedot1t2t3_top <- cbind(RvNR_T1_vedot1t2t3_top, Beta = coefficients(betas_ebayes_vedo_t1t2t3)[rownames(T2vT1R_vedo_top), "RvNRT1"], data.frame(gmset_vedo_anno_gr[rownames(RvNR_T1_vedot1t2t3_top), ]))
```

```{r vedo DM analyses RvNR for T2 subsets}
RvNR_T2_vedot1t2t3_top <- topTable(mvals_ebayes_vedo_t1t2t3, coef = "RvNRT2", number = "Inf", adjust.method = "BH")
RvNR_T2_vedot1t2t3_top <- cbind(RvNR_T2_vedot1t2t3_top, Beta = coefficients(betas_ebayes_vedo_t1t2t3)[rownames(T2vT1R_vedo_top), "RvNRT2"], data.frame(gmset_vedo_anno_gr[rownames(RvNR_T2_vedot1t2t3_top), ]))
```

```{r vedo DM analyses RvNR for T3 subsets}
RvNR_T3_vedot1t2t3_top <- topTable(mvals_ebayes_vedo_t1t2t3, coef = "RvNRT3", number = "Inf", adjust.method = "BH")
RvNR_T3_vedot1t2t3_top <- cbind(RvNR_T3_vedot1t2t3_top, Beta = coefficients(betas_ebayes_vedo_t1t2t3)[rownames(T2vT1R_vedo_top), "RvNRT3"], data.frame(gmset_vedo_anno_gr[rownames(RvNR_T3_vedot1t2t3_top), ]))
```


